// @arcgis/map-components/dist/components/arcgis-print-preview-input/customElement.js
import { c as t } from "../../chunks/runtime.js";
import { html as i } from "lit-html";
import { watch as r } from "@arcgis/core/core/reactiveUtils.js";
import { LitElement as s, noShadowRoot as a } from "@arcgis/lumina";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
class n extends s {
  constructor() {
    super(...arguments), this._showPrintAreaEnabled = !1;
  }
  static {
    this.properties = { _showPrintAreaEnabled: 16, viewModel: 0, messages: 0 };
  }
  static {
    this.shadowRootOptions = a;
  }
  loaded() {
    this.manager.onLifecycle(() => [
      r(() => [this.viewModel.showPrintAreaEnabled], () => {
        const { showPrintAreaEnabled: e } = this.viewModel;
        this._showPrintAreaEnabled = e;
      }, { initial: !0 })
    ]);
  }
  render() {
    return i`<calcite-label layout=inline><calcite-switch .checked=${this._showPrintAreaEnabled} @calciteSwitchChange=${(e) => {
      this.viewModel.showPrintAreaEnabled = !!e.currentTarget.checked;
    }}></calcite-switch>${this.messages.printPreview}</calcite-label>`;
  }
}
t("arcgis-print-preview-input", n);
export {
  n as PrintPreviewInput
};
