// @arcgis/charts-components/dist/components/arcgis-charts-config-bar-chart/customElement.js
import { c as n } from "../../chunks/runtime.js";
import "@arcgis/charts-spec";
import { isEqual as r } from "lodash-es";
import "d3-array";
import { g as l, a_ as h, b2 as c } from "../../chunks/interfaces.js";
import "@arcgis/core/geometry/support/jsonUtils.js";
import "@arcgis/core/rest/support/AttributeBinsQuery.js";
import "@arcgis/core/rest/support/Query.js";
import "@arcgis/core/rest/support/StatisticDefinition.js";
import "@arcgis/core/time/TimeExtent.js";
import "@arcgis/core/core/promiseUtils.js";
import "@arcgis/core/request.js";
import { b as d, h as p } from "../../chunks/index4.js";
import { LitElement as g, createEvent as m } from "@arcgis/lumina";
import { html as a } from "lit";
import { C as o } from "../../chunks/interfaces2.js";
import { css as f } from "@lit/reactive-element/css-tag.js";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const C = f`:host{margin-bottom:var(--arcgis-charts-intra-section-margin);margin-top:var(--arcgis-charts-intra-section-margin);height:100%}`;
class u extends g {
  constructor() {
    super(...arguments), this.popoverElement = null, this.onModelConfigChange = async (e) => {
      const i = e?.detail.config ?? this.model.getConfig();
      this.layer === void 0 && await this.updateFetchLayerFieldsInfo(i);
    }, this.activeColorMatch = !1, this.popoverPlacement = "leading", this.arcgisChartsConfigNoSeriesUpdate = m();
  }
  static {
    this.properties = { layerFieldsInfo: 16, activeColorMatch: 16, model: 0, timeZone: 3, popoverPlacement: 3, seriesColor: 2, calculatedMinMaxBounds: 2, page: 3, layer: 0, usePopupTemplateFieldsInfo: 5 };
  }
  static {
    this.styles = C;
  }
  async load() {
    this.layer !== void 0 && await this.layerChange(this.layer, void 0), this.seriesColor !== void 0 && this.seriesColorChange(this.seriesColor), this.usePopupTemplateFieldsInfo !== void 0 && await this.usePopupTemplateFieldsInfoChange(this.usePopupTemplateFieldsInfo, void 0), this.modelChange(this.model), this.onModelConfigChange(), this.updateDataFilters();
  }
  async willUpdate(e) {
    e.has("model") && this.modelChange(this.model, e.get("model")), e.has("layer") && await this.layerChange(this.layer, e.get("layer")), e.has("usePopupTemplateFieldsInfo") && this.usePopupTemplateFieldsInfo !== void 0 && await this.usePopupTemplateFieldsInfoChange(this.usePopupTemplateFieldsInfo, e.get("usePopupTemplateFieldsInfo")), e.has("seriesColor") && this.seriesColor !== void 0 && this.seriesColorChange(this.seriesColor), this.layerFieldsInfo === void 0 && await this.updateFetchLayerFieldsInfo();
  }
  disconnectedCallback() {
    super.disconnectedCallback(), this.model.removeEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange);
  }
  modelChange(e, t) {
    t?.removeEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange), e?.addEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange);
  }
  async layerChange(e, t) {
    e !== void 0 && !r(t, e) && await this.updateFetchLayerFieldsInfo();
  }
  async usePopupTemplateFieldsInfoChange(e, t) {
    e !== void 0 && e !== t && await this.updateFetchLayerFieldsInfo();
  }
  seriesColorChange(e) {
    this.activeColorMatch = e?.get(0) === void 0, this.model.getColorMatch() && e?.forEach((t, i) => {
      d(t) && e?.get(0) !== void 0 && t?.color !== void 0 && !r(this.model.getSeriesColor(i), t?.color) && this.model.setSeriesColor(t?.color, i);
    });
  }
  async updateDataFilters() {
    const e = this.model.getDataFilter(), t = e?.where, i = this.layer?.definitionExpression ?? void 0;
    r(t, i) || await this.model.setDataFilter({
      ...e,
      where: i
    });
  }
  async updateFetchLayerFieldsInfo(e) {
    const t = e ?? this.model.getConfig();
    let i = this.layer;
    if (!i) {
      const s = p(t) ? t.iLayer : void 0;
      s && (i = await l(s));
    }
    i && (this.layerFieldsInfo = await h(i, this.usePopupTemplateFieldsInfo));
  }
  popoverOpen(e) {
    const { open: t = !1, element: i = null } = { ...e.detail };
    t === !0 && this.popoverElement !== i ? (this.popoverElement !== null && (this.popoverElement.open = !1), this.popoverElement = i) : t === !1 && this.popoverElement === i && (this.popoverElement = null);
  }
  render() {
    const e = this.layerFieldsInfo?.find((s) => s.name === this.model.getXAxisField()), t = e !== void 0 && c(e);
    let i;
    switch (this.page) {
      case o.Data:
        i = a`<arcgis-charts-config-bar-chart-data .model=${this.model} .layerFieldsInfo=${this.layerFieldsInfo}></arcgis-charts-config-bar-chart-data>`;
        break;
      case o.Series:
        i = a`<arcgis-charts-config-bar-chart-series .model=${this.model} .activeColorMatch=${this.activeColorMatch} @arcgisChartsConfigBarChartSeriesPopoverChange=${this.popoverOpen} .popoverPlacement=${this.popoverPlacement}></arcgis-charts-config-bar-chart-series>`;
        break;
      case o.Axes:
        i = a`<arcgis-charts-config-axes .model=${this.model} show-max-y-bound-option .calculatedMinMaxBounds=${this.calculatedMinMaxBounds}></arcgis-charts-config-axes>`;
        break;
      case o.Guides:
        i = a`<arcgis-charts-config-guides .model=${this.model} .timeZone=${this.timeZone} .enableXAxisGuide=${t} enable-y-axis-guide .calculatedMinMaxBounds=${this.calculatedMinMaxBounds} @arcgisChartsConfigGuidePopoverChange=${this.popoverOpen}></arcgis-charts-config-guides>`;
        break;
      case o.Format:
        i = a`<arcgis-charts-config-format .model=${this.model} @arcgisChartsConfigFormatPopoverChange=${this.popoverOpen}></arcgis-charts-config-format>`;
        break;
      case o.General:
        i = a`<arcgis-charts-config-general .model=${this.model}></arcgis-charts-config-general>`;
        break;
      default:
        i = "";
        break;
    }
    return i;
  }
}
n("arcgis-charts-config-bar-chart", u);
export {
  u as ArcgisChartsConfigBarChart
};
