// @arcgis/charts-components/dist/model.js
import { B as i, C as e, F as o, G as t, L as r, M as l, a as d, P as M, b as h, S as n, T as C } from "./chunks/common.js";
import { a as p, e as W, i as x, g as S, b as g, r as y, h as K, l as P, m as c, p as B, k as D, q as f, n as F, j as G, o as L, d as T, f as u, c as b } from "./chunks/type-guards.js";
import { c as A } from "./chunks/setup-utils.js";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
export {
  i as BoxPlotDataSortingKinds,
  e as ColorFormatKeys,
  o as FormatKeys,
  t as GuideOrientation,
  r as LineFormatKeys,
  l as ModelTypes,
  d as PieChartDataModes,
  M as PieChartDataSortingKinds,
  h as PieChartSlicesDisplay,
  n as SerialChartDataSortingKinds,
  C as TextFormatKeys,
  A as createModel,
  p as isBarChartModel,
  W as isBoxPlotModel,
  x as isChartModel,
  S as isHistogramModel,
  g as isLineChartModel,
  y as isModelOfType,
  K as isModelWithBasicXAxis,
  P as isModelWithDataLabels,
  c as isModelWithGuides,
  B as isModelWithLegend,
  D as isModelWithRotation,
  f as isModelWithTemporalBinning,
  F as isModelWithXGuides,
  G as isModelWithXYAxes,
  L as isModelWithYGuides,
  T as isPieChartModel,
  u as isScatterplotModel,
  b as isSerialChartModel
};
