// @arcgis/map-components/dist/components/arcgis-daylight/customElement.js
import { c as e } from "../../chunks/runtime.js";
import i from "@arcgis/core/widgets/Daylight.js";
import { LitElement as t, createEvent as s, noShadowRoot as a } from "@arcgis/lumina";
import { m as h } from "../../chunks/useWidget.js";
import { css as l } from "@lit/reactive-element/css-tag.js";
import { reEmitEvent as o } from "@arcgis/lumina/controllers/accessor";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const d = l`@layer{arcgis-daylight{display:block}}`, r = h(i);
class n extends t {
  constructor() {
    super(...arguments), this.widget = r(this), this.viewModel = this.widget.viewModel, this.autoDestroyDisabled = !1, this.currentSeason = this.viewModel.currentSeason, this.dateOrSeason = this.widget.dateOrSeason, this.dayPlaying = this.viewModel.dayPlaying, this.headingLevel = this.widget.headingLevel, this.hideDatePicker = this.widget.visibleElements.datePicker, this.hideHeader = this.widget.visibleElements.header, this.hidePlayButtons = this.widget.visibleElements.playButtons, this.hideShadowsToggle = this.widget.visibleElements.shadowsToggle, this.hideSunLightingToggle = this.widget.visibleElements.sunLightingToggle, this.hideTimezone = this.widget.visibleElements.timezone, this.icon = this.widget.icon, this.label = this.widget.label, this.playSpeedMultiplier = this.widget.playSpeedMultiplier, this.position = "bottom-left", this.timeSliderSteps = this.widget.timeSliderSteps, this.yearPlaying = this.viewModel.yearPlaying, this.arcgisReady = s(), this.arcgisUserDateTimeChange = o(() => this.viewModel, "user-date-time-change");
  }
  static {
    this.properties = { autoDestroyDisabled: 5, currentSeason: 1, dateOrSeason: 1, dayPlaying: 5, headingLevel: 9, hideDatePicker: 5, hideHeader: 5, hidePlayButtons: 5, hideShadowsToggle: 5, hideSunLightingToggle: 5, hideTimezone: 5, icon: 1, label: 1, playSpeedMultiplier: 9, position: 1, referenceElement: 1, timeSliderSteps: 9, yearPlaying: 5 };
  }
  static {
    this.shadowRootOptions = a;
  }
  static {
    this.styles = d;
  }
  async destroy() {
    await this.manager.destroy();
  }
}
e("arcgis-daylight", n);
export {
  n as ArcgisDaylight
};
