// @arcgis/charts-components/dist/chunks/axis-model-wrapper.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
class e {
  /**
   * `AxisModelWrapper` constructor.
   *
   * @param model The wrapped chart model.
   * @param isXAxis Whether the wrapper is for X-axis or Y-axis.
   */
  constructor(i, s = !1) {
    this.model = i, this.isXAxis = s;
  }
  getMaxBound() {
    return (this.isXAxis ? this.model?.getMaxXBound() : this.model?.getMaxYBound()) ?? Number.NaN;
  }
  async setMaxBound(i) {
    this.isXAxis ? await this.model?.setMaxXBound(i) : await this.model?.setMaxYBound(i);
  }
  getMinBound() {
    return (this.isXAxis ? this.model?.getMinXBound() : this.model?.getMinYBound()) ?? Number.NaN;
  }
  async setMinBound(i) {
    this.isXAxis ? await this.model?.setMinXBound(i) : await this.model?.setMinYBound(i);
  }
  getValueFormat() {
    return this.isXAxis ? this.model?.getXAxisValueFormat() : this.model?.getYAxisValueFormat();
  }
  async setValueFormat(i) {
    this.isXAxis ? await this.model?.setXAxisValueFormat(i) : await this.model?.setYAxisValueFormat(i);
  }
  getLogarithmic() {
    return (this.isXAxis ? this.model?.getXLogarithmic() : this.model?.getYLogarithmic()) ?? !1;
  }
  async setLogarithmic(i) {
    this.isXAxis ? await this.model?.setXLogarithmic(i) : await this.model?.setYLogarithmic(i);
  }
}
export {
  e as A
};
