// @arcgis/charts-components/dist/chunks/common.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = {
  xAxisAsc: "xAxisAsc",
  xAxisDesc: "xAxisDesc",
  yAxisAsc: "yAxisAsc",
  yAxisDesc: "yAxisDesc",
  customSort: "customSort"
}, s = {
  xAxisAsc: "xAxisAsc",
  xAxisDesc: "xAxisDesc",
  meanAsc: "meanAsc",
  meanDesc: "meanDesc",
  medianAsc: "medianAsc",
  medianDesc: "medianDesc",
  customSort: "customSort"
}, a = {
  labelAsc: "labelAsc",
  labelDesc: "labelDesc",
  valueAsc: "valueAsc",
  valueDesc: "valueDesc",
  customSort: "customSort"
}, t = {
  Category: "category",
  Fields: "fields"
}, i = {
  Category: "category",
  Value: "value",
  Percentage: "percentage",
  Both: "both"
}, n = {
  TextElements: "textElements",
  XAxisTitle: "xAxisTitle",
  YAxisTitle: "yAxisTitle",
  AxisLabels: "axisLabels",
  GuideLabels: "guideLabels",
  Title: "title",
  LegendText: "legendText",
  LegendTitle: "legendTitle",
  Description: "description",
  DataLabels: "dataLabels",
  Background: "background",
  LeaderLines: "leaderLines",
  AxisLines: "axisLines",
  GridLines: "gridLines"
}, o = {
  AxisLines: "axisLines",
  GridLines: "gridLines",
  LeaderLines: "leaderLines"
}, r = {
  Title: "title",
  XAxisTitle: "xAxisTitle",
  YAxisTitle: "yAxisTitle",
  AxisLabels: "axisLabels",
  GuideLabels: "guideLabels",
  LegendText: "legendText",
  LegendTitle: "legendTitle",
  Description: "description",
  DataLabels: "dataLabels"
}, l = {
  Background: "background"
}, c = {
  X: "x",
  Y: "y"
}, x = {
  BarChart: "barChart",
  LineChart: "lineChart",
  ComboBarLineChart: "comboBarLineChart",
  BoxPlot: "boxPlot",
  PieChart: "pieChart",
  Scatterplot: "scatterplot",
  Histogram: "histogram",
  Gauge: "gauge",
  RadarChart: "radarChart"
};
export {
  s as B,
  l as C,
  n as F,
  c as G,
  o as L,
  x as M,
  a as P,
  e as S,
  r as T,
  t as a,
  i as b
};
