// @arcgis/charts-components/dist/chunks/index2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const t = {
  MonoSelection: "monoSelection",
  MultiSelection: "multiSelection",
  MultiSelectionWithCtrlKey: "multiSelectionWithCtrlKey",
  Zoom: "zoom",
  None: "none"
}, a = {
  MultiAxes: "multi-axes",
  XAxis: "x-axis",
  YAxis: "y-axis"
}, o = {
  DataProcessComplete: "arcgisDataProcessComplete",
  DataFetchComplete: "arcgisDataFetchComplete",
  DataProcessError: "arcgisDataProcessError",
  LegendItemVisibilityChange: "arcgisLegendItemVisibilityChange",
  UpdateComplete: "arcgisUpdateComplete",
  SeriesColorChange: "arcgisSeriesColorChange",
  AxesMinMaxChange: "arcgisAxesMinMaxChange",
  DefaultActionSelect: "arcgisDefaultActionSelect",
  SelectionComplete: "arcgisSelectionComplete",
  ActionBarToggle: "arcgisActionBarToggle",
  NoRenderPropChange: "arcgisNoRenderPropChange",
  BadDataWarningRaise: "arcgisBadDataWarningRaise",
  SeriesOrder: "arcgisSeriesOrder",
  RuntimeError: "arcgisRuntimeError",
  ConfigChange: "arcgisConfigChange"
}, i = {
  FilterBySelection: "filterBySelection",
  FilterByExtent: "filterByExtent",
  Legend: "legend",
  RotateChart: "rotateChart",
  Selection: "selection",
  ClearSelection: "clearSelection",
  SwitchSelection: "switchSelection",
  Zoom: "zoom",
  FullExtent: "fullExtent",
  EditChart: "editChart",
  ExportAsImage: "exportAsImage",
  ExportAsCSV: "exportAsCSV"
}, n = {
  Active: "active",
  Inactive: "inactive"
}, r = {
  Enabled: "enabled",
  Disabled: "disabled"
}, s = {
  Ignore: "ignore",
  Throw: "throw"
}, e = {
  Ignore: "ignore",
  Refresh: "refresh"
}, c = e, l = e, g = e, C = e, h = {
  /**
   * Enable syncing selections between chart and layer view
   */
  Enabled: "enabled",
  /**
   * Disable syncing selections between chart and layer view
   */
  Disabled: "disabled"
};
export {
  t as A,
  a as C,
  i as D,
  s as E,
  c as L,
  h as S,
  C as V,
  o as a,
  n as b,
  r as c,
  l as d,
  g as e
};
