// @arcgis/charts-components/dist/chunks/interfaces2.js
import { WebChartStatisticType as e } from "@arcgis/charts-spec";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const i = {
  actionBar: "action-bar",
  customActionTop: "custom-action-top",
  customActionBelow: "custom-action-below"
}, a = {
  Mean: "mean",
  Median: "median",
  StandardDev: "standardDev",
  NormalDist: "normalDist"
}, o = {
  Mean: "mean",
  Median: "median",
  StandardDev: "standardDev",
  Count: e.Count,
  CountExcludingNulls: "countExcludingNulls",
  Min: e.Minimum,
  Max: e.Maximum,
  Sum: e.Sum,
  Nulls: "nulls"
}, n = {
  Data: "data",
  Series: "series",
  Slices: "slices",
  Axes: "axes",
  Guides: "guides",
  Format: "format",
  General: "general"
}, s = {
  Title: "title",
  Text: "text",
  Label: "label",
  Palette: "palette"
}, r = {
  xAxisField: "xAxisField",
  aggregation: "aggregation",
  numericFields: "numericField",
  splitByField: "splitByField",
  interval: "interval",
  intervalAlignment: "intervalAlignment"
}, l = {
  category: "category",
  aggregation: "aggregation",
  numericFields: "numericField",
  splitByField: "splitByField",
  sortOrder: "sortOrder"
}, d = {
  category: "category",
  aggregation: "aggregation",
  numericFields: "numericFields",
  sortOrder: "sortOrder"
}, c = {
  numericFields: "numericField",
  transformation: "transformation",
  fillSymbol: "fillSymbol",
  lineSymbol: "lineSymbol"
}, m = {
  xAxisField: "xAxis",
  yAxisField: "yAxis",
  sizeVariable: "sizeVariable",
  symbol: "symbol",
  trendLine: "trendLine",
  none: "none"
}, g = {
  category: "category",
  numericFields: "numericField",
  splitByField: "splitByField",
  sortOrder: "sortOrder"
}, u = {
  Fixed: "fixed",
  Proportional: "proportional"
}, p = {
  Up: "up",
  Down: "down"
};
export {
  l as B,
  n as C,
  s as F,
  c as H,
  r as L,
  d as P,
  m as S,
  g as a,
  u as b,
  a as c,
  o as d,
  p as e,
  i as f
};
