// @arcgis/charts-components/dist/chunks/messages.sk.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const a = "Pri načítavaní grafu sa vyskytla chyba.", e = "Nie je možné aplikovať Log transformáciu na záporné alebo nulové hodnoty.", o = "Nie je možné aplikovať Square Root transformáciu na záporné hodnoty.", n = "Pri načítavaní vrstvy sa vyskytla chyba. URL = ${ url }. ID položky portálu = ${ portalItemId }.", t = "Polia [${ fieldNames }] definované v konfigurácii sa nenašli vo vrstve.", s = "${ dataPath } musí byť jedinečný. Séria pomenovaná ${ seriesName } má id (${ seriesID }), ktoré už je používané inou sériou.", i = "${ dataPath } nemôže vykonávať agregáciu bez počtov na nečíselnom poli.", r = "${ dataPath } chýba vlastnosť s názvom ${ missingProperty }.", l = "${ dataPath } musí byť kratší než ${ limit } znakov.", m = "${ dataPath } nesmie mať menej než ${ limit } položiek.", d = "${ dataPath } nesmie mať viac než ${ limit } položiek.", p = "${ dataPath } musí mať aspoň jeden znak bez prázdneho miesta.", c = "${ dataPath } nesmie mať ${ additionalProperty }.", u = "${ dataPath } musí byť rovné jednej z týchto povolených hodnôt: ${ allowedValues }.", v = "${ dataPath } sa musí zhodovať so schémou jednej z týchto: ${ schemaOptions }.", y = "Korelačné grafy s proporcionálnymi symbolmi nie sú podporované. Bola aplikovaná predvolená veľkosť symbolu.", h = "Zlyhanie načítania vstupných údajov.", $ = "Operácia queryBins nie je v tejto službe podporovaná.", P = "Očakávaný typ série pri indexe ${ seriesIndex } je '${ expectedType }', ale namiesto toho bolo prijaté '${ receivedType }'", b = "alebo", k = {
  defaultError: a,
  negativeValueInDataForLogTransformation: e,
  negativeValueInDataForSqrtTransformation: o,
  layerLoadFailure: n,
  layerFieldsNotFound: t,
  duplicateSeriesID: s,
  nonNumericAggregation: i,
  requiredProperty: r,
  minLength: l,
  minItems: m,
  maxItems: d,
  whiteSpacePattern: p,
  additionalProperty: c,
  enumValues: u,
  anyOfValues: v,
  bubbleChartValidateMsg: y,
  queryError: h,
  queryBinsNotSupported: $,
  invalidSeriesType: P,
  or: b
};
export {
  c as additionalProperty,
  v as anyOfValues,
  y as bubbleChartValidateMsg,
  k as default,
  a as defaultError,
  s as duplicateSeriesID,
  u as enumValues,
  P as invalidSeriesType,
  t as layerFieldsNotFound,
  n as layerLoadFailure,
  d as maxItems,
  m as minItems,
  l as minLength,
  e as negativeValueInDataForLogTransformation,
  o as negativeValueInDataForSqrtTransformation,
  i as nonNumericAggregation,
  b as or,
  $ as queryBinsNotSupported,
  h as queryError,
  r as requiredProperty,
  p as whiteSpacePattern
};
