// @arcgis/charts-components/dist/chunks/messages.hu.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "Hiba történt a diagram betöltése közben.", t = "A log transzformáció nem alkalmazható negatív vagy nulla értékre.", a = "A négyzetgyök transzformáció nem alkalmazható negatív értékre.", n = "Hiba történt a réteg betöltés közben. URL = ${ url }. Portál elemazonosítója = ${ portalItemId }.", l = "A konfigurációban definiált [${ fieldNames }] mezők nem találhatók a rétegben.", r = "Egyedinek kell lennie: ${ dataPath } A(z) ${ seriesName } nevű sorozat azonosítóját (${ seriesID }) egy másik sorozat már használja.", s = "A(z) ${ dataPath } nem hajthat végre nem megszámlálható összevonást nem numerikus mezőn.", o = "A(z) ${ dataPath } esetében hiányzik a(z) ${ missingProperty } nevű tulajdonság.", i = "A(z) ${ dataPath } hossza legalább ${ limit } karakter kell legyen.", m = "A(z) ${ dataPath } elemeinek száma legalább ${ limit } kell legyen.", k = "A(z) ${ dataPath } elemeinek száma legfeljebb ${ limit } lehet.", z = "A következőnek legalább egy nem-whitespace karaktert kell tartalmaznia: ${ dataPath }", g = "${ dataPath } nem rendelkezhet ezzel: ${ additionalProperty }.", d = "${ dataPath } értékének egyeznie kell a következő megengedett értékek egyikével: ${ allowedValues }.", c = "A(z) ${ dataPath } értékének egyeznie kell a következők egyikének sémájával: ${ schemaOptions }.", y = "A méretarányos szimbólummal rendelkező szórásdiagramok nem támogatottak. Az alapértelmezett szimbólum méret lesz alkalmazva.", h = "Nem sikerült beolvasni a bemeneti adatokat.", u = "A queryBins művelet nem támogatott ebben a szolgáltatásban.", b = "A várt sorozattípus a(z) ${ seriesIndex } indexnél '${ expectedType }', de helyette '${ receivedType }' érkezett.", $ = "vagy", v = {
  defaultError: e,
  negativeValueInDataForLogTransformation: t,
  negativeValueInDataForSqrtTransformation: a,
  layerLoadFailure: n,
  layerFieldsNotFound: l,
  duplicateSeriesID: r,
  nonNumericAggregation: s,
  requiredProperty: o,
  minLength: i,
  minItems: m,
  maxItems: k,
  whiteSpacePattern: z,
  additionalProperty: g,
  enumValues: d,
  anyOfValues: c,
  bubbleChartValidateMsg: y,
  queryError: h,
  queryBinsNotSupported: u,
  invalidSeriesType: b,
  or: $
};
export {
  g as additionalProperty,
  c as anyOfValues,
  y as bubbleChartValidateMsg,
  v as default,
  e as defaultError,
  r as duplicateSeriesID,
  d as enumValues,
  b as invalidSeriesType,
  l as layerFieldsNotFound,
  n as layerLoadFailure,
  k as maxItems,
  m as minItems,
  i as minLength,
  t as negativeValueInDataForLogTransformation,
  a as negativeValueInDataForSqrtTransformation,
  s as nonNumericAggregation,
  $ as or,
  u as queryBinsNotSupported,
  h as queryError,
  o as requiredProperty,
  z as whiteSpacePattern
};
