// @arcgis/charts-components/dist/chunks/messages.uk4.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "Відтворити", o = "Стоп", t = "Легенда", r = "Завантажується...", s = "Головна сторінка", a = "Діаграма", n = "Серія", i = "Експорт", l = "Зображення", c = "Дані", d = "Друк", m = {
  "Zoom Out": "Зменшити масштаб",
  Play: e,
  Stop: o,
  Legend: t,
  "Press ENTER to toggle": "Натисніть Enter, щоб перемкнути",
  Loading: r,
  Home: s,
  Chart: a,
  "Serial chart": "Серійна діаграма",
  "X/Y chart": "Діаграма X/Y",
  "Pie chart": "Кругова діаграма",
  "Gauge chart": "Діаграма з індикаторами",
  "Radar chart": "Діаграма радара",
  "Sankey diagram": "Діаграма Санкея",
  "Flow diagram": "Функціональна діаграма",
  "Chord diagram": "Діаграма хорди",
  "TreeMap chart": "Діаграма деревоподібної карти",
  "Sliced chart": "Секторна діаграма",
  Series: n,
  "Candlestick Series": "Свічкові ряди",
  "OHLC Series": "Ряди OHLC",
  "Column Series": "Стовпчасті ряди",
  "Line Series": "Лінійні ряди",
  "Pie Slice Series": "Секторні ряди",
  "Funnel Series": "Ворока рядів",
  "Pyramid Series": "Пірамідні ряди",
  "X/Y Series": "Ряди X/Y",
  Map: "Карта",
  "Press ENTER to zoom in": "Натисніть Enter, щоб збільшити",
  "Press ENTER to zoom out": "Натисніть Enter, щоб зменшити",
  "Use arrow keys to zoom in and out": "Для масштабування використовуйте клавіші зі стрілками",
  "Use plus and minus keys on your keyboard to zoom in and out": "Використовуйте клавіші «плюс» та «мінус» на клавіатурі для збільшення та зменшення масштабу",
  Export: i,
  Image: l,
  Data: c,
  Print: d,
  "Press ENTER to open": "Натисніть Enter, щоб відкрити",
  "Press ENTER to print.": "Натисніть Enter, щоб надрукувати.",
  "Press ENTER to export as %1.": "Натисніть Enter, щоб експортувати як %1.",
  "(Press ESC to close this message)": "(Натисніть ESC, щоб закрити це повідомлення)",
  "Image Export Complete": "Експорт зображення завершено",
  "Export operation took longer than expected. Something might have gone wrong.": "Операція експорту займає більше часу, ніж очікувалось. Щось могло піти не так.",
  "Saved from": "Збережено з",
  "Use TAB to select grip buttons or left and right arrows to change selection": "Використовуйте TAB, щоб вибрати кнопки маніпулювання, або стрілку вліво та стрілку вправо, щоб змінити вибірку",
  "Use left and right arrows to move selection": "Скористайтеся стрілкою вліво та стрілкою вправо, щоб перемістити вибірку",
  "Use left and right arrows to move left selection": "Скористайтеся стрілкою вліво та стрілкою вправо, щоб перемістити ліву вибірку",
  "Use left and right arrows to move right selection": "Скористайтеся стрілкою вліво та стрілкою вправо, щоб перемістити праву вибірку",
  "Use TAB select grip buttons or up and down arrows to change selection": "Використовуйте кнопки вибору маніпулювання TAB або стрілку вгору та стрілку вниз, щоб змінити вибірку",
  "Use up and down arrows to move selection": "Скористайтеся стрілкою вгору та стрілкою вниз, щоб перемістити вибірку",
  "Use up and down arrows to move lower selection": "Скористайтеся стрілкою вгору та стрілкою вниз, щоб перемістити нижню вибірку",
  "Use up and down arrows to move upper selection": "Скористайтеся стрілкою вгору та стрілкою вниз, щоб перемістити верхню вибірку",
  "From %1 to %2": "Від %1 до %2",
  "From %1": "З %1",
  "To %1": "До %1",
  "No parser available for file: %1": "Немає доступного аналізатора для файлу: %1",
  "Error parsing file: %1": "Помилка аналізу файлу: %1",
  "Unable to load file: %1": "Не вдалося завантажити файл: %1",
  "Invalid date": "Некоректна дата"
};
export {
  a as Chart,
  c as Data,
  i as Export,
  s as Home,
  l as Image,
  t as Legend,
  r as Loading,
  e as Play,
  d as Print,
  n as Series,
  o as Stop,
  m as default
};
