// @arcgis/charts-components/dist/chunks/messages.es.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "Se ha producido un error al cargar el gráfico.", a = "No se puede aplicar una transformación del registro a valores negativos o de cero.", o = "No se puede aplicar una transformación de raíz cuadrada a valores negativos.", n = "Se ha producido un error al cargar la capa. URL = ${ url }. Id. de elemento de portal = ${ portalItemId }.", r = "Los campos [${ fieldNames }] definidos en la configuración no se han encontrado en la capa.", t = "El elemento ${ dataPath } debe ser único. La serie denominada ${ seriesName } tiene un Id. (${ seriesID }) que ya se utiliza en otra serie.", s = "${ dataPath } no puede realizar una agregación no numérica en un campo no numérico.", i = "A ${ dataPath } le falta la propiedad denominada ${ missingProperty }.", d = "${ dataPath } no debe tener menos de ${ limit } caracteres.", c = "${ dataPath } no debe tener menos de ${ limit } elementos.", l = "${ dataPath } no debe tener más de ${ limit } elementos.", u = "${ dataPath } debe tener al menos un carácter que no sea un espacio en blanco.", m = "${ dataPath } no debe tener ${ additionalProperty }", p = "${ dataPath } debe ser igual a uno de los siguientes valores permitidos: ${ allowedValues }.", $ = "${ dataPath } debe coincidir con el esquema de uno de estos: ${ schemaOptions }.", g = "No se admiten diagramas de dispersión con símbolos proporcionales. Se ha aplicado el tamaño de símbolo predeterminado.", h = "Error al leer los datos de entrada.", b = "En este servicio no se admite la operación queryBins.", P = 'El tipo de serie esperado en el índice ${ seriesIndex } es "${ expectedType }", pero en su lugar se ha recibido "${ receivedType }"', y = "o", f = {
  defaultError: e,
  negativeValueInDataForLogTransformation: a,
  negativeValueInDataForSqrtTransformation: o,
  layerLoadFailure: n,
  layerFieldsNotFound: r,
  duplicateSeriesID: t,
  nonNumericAggregation: s,
  requiredProperty: i,
  minLength: d,
  minItems: c,
  maxItems: l,
  whiteSpacePattern: u,
  additionalProperty: m,
  enumValues: p,
  anyOfValues: $,
  bubbleChartValidateMsg: g,
  queryError: h,
  queryBinsNotSupported: b,
  invalidSeriesType: P,
  or: y
};
export {
  m as additionalProperty,
  $ as anyOfValues,
  g as bubbleChartValidateMsg,
  f as default,
  e as defaultError,
  t as duplicateSeriesID,
  p as enumValues,
  P as invalidSeriesType,
  r as layerFieldsNotFound,
  n as layerLoadFailure,
  l as maxItems,
  c as minItems,
  d as minLength,
  a as negativeValueInDataForLogTransformation,
  o as negativeValueInDataForSqrtTransformation,
  s as nonNumericAggregation,
  y as or,
  b as queryBinsNotSupported,
  h as queryError,
  i as requiredProperty,
  u as whiteSpacePattern
};
