// @arcgis/charts-components/dist/chunks/messages.vi2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const n = { chart: "Biểu đồ", barChart: "Biểu đồ cột", lineChart: "Biểu đồ dạng đường", pieChart: "Biểu đồ hình tròn", histogram: "Biểu đồ tần suất", scatterplot: "Biểu đồ phân tán", boxPlot: "Biểu đồ hộp" }, t = { mean: "Trung bình", median: "Bình quân", sum: "Tổng", discretePercentile: "Phân vị rời rạc", minimum: "Tối thiểu", maximum: "Tối đa", variance: "Phương sai", count: "Số lượng", aggregation: "${ statistics } / ${ fieldName }", noAggregation: "Không tổng hợp" }, i = "Biểu đồ", a = "Đếm Số chuỗi", s = "Trục X", e = "Không Khả dụng", u = { start: "Chuyển sang điểm dữ liệu đầu tiên", end: "Chuyển sang điểm dữ liệu cuối cùng" }, c = {
  chartType: n,
  statistics: t,
  defaultTitle: i,
  countSeries: a,
  xAxis: s,
  notAvailable: e,
  timeAggregationTypes: u
};
export {
  n as chartType,
  a as countSeries,
  c as default,
  i as defaultTitle,
  e as notAvailable,
  t as statistics,
  u as timeAggregationTypes,
  s as xAxis
};
