// @arcgis/charts-components/dist/chunks/messages.hu4.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "Lejátszás", a = "Megálló", t = "Jelmagyarázat", o = "Betöltés", l = "Kezdőlap", s = "Diagram", n = "Sorozat", r = "Exportálás", i = "Kép", m = "Adatok", g = "Nyomtatás", d = {
  "Zoom Out": "Kicsinyítés",
  Play: e,
  Stop: a,
  Legend: t,
  "Press ENTER to toggle": "Nyomja meg az ENTER billentyűt a váltáshoz",
  Loading: o,
  Home: l,
  Chart: s,
  "Serial chart": "Értéksordiagram",
  "X/Y chart": "X/Y-diagram",
  "Pie chart": "Kördiagram",
  "Gauge chart": "Mérőpanel diagram",
  "Radar chart": "Radardiagram",
  "Sankey diagram": "Sankey-diagram",
  "Flow diagram": "Áramlásdiagram",
  "Chord diagram": "Húrdiagram",
  "TreeMap chart": "TreeMap-diagram",
  "Sliced chart": "Szeletelt diagram",
  Series: n,
  "Candlestick Series": "Gyertyadiagramok",
  "OHLC Series": "OHLC-diagramok",
  "Column Series": "Oszlopdiagramok",
  "Line Series": "Vonaldiagramok",
  "Pie Slice Series": "Tortadiagramok",
  "Funnel Series": "Tölcsérdiagramok",
  "Pyramid Series": "Piramisdiagramok",
  "X/Y Series": "X/Y-diagramok",
  Map: "Térkép",
  "Press ENTER to zoom in": "Nyomja meg az ENTER billentyűt a nagyításhoz",
  "Press ENTER to zoom out": "Nyomja meg az ENTER billentyűt a kicsinyítéshez",
  "Use arrow keys to zoom in and out": "Használja a nyílbillentyűket a nagyításhoz és kicsinyítéshez",
  "Use plus and minus keys on your keyboard to zoom in and out": "Használja a billentyűzetén található plusz- és mínusz billentyűket a nagyításhoz és kicsinyítéshez",
  Export: r,
  Image: i,
  Data: m,
  Print: g,
  "Press ENTER to open": "Nyomja meg az ENTER billentyűt a megnyitáshoz",
  "Press ENTER to print.": "Nyomja meg az ENTER billentyűt a nyomításhoz",
  "Press ENTER to export as %1.": "Nyomja meg az ENTER billentyűt a(z) %1-ként való exportáláshoz.",
  "(Press ESC to close this message)": "(Nyomja meg az ESC billentyűt az üzenet bezárásához)",
  "Image Export Complete": "Kép exportálása befejeződött",
  "Export operation took longer than expected. Something might have gone wrong.": "Az exportálási művelet a vártnál hosszabb ideig tartott. Valami hiba történhetett.",
  "Saved from": "Mentve innen:",
  "Use TAB to select grip buttons or left and right arrows to change selection": "A TAB billentyűvel kiválaszthatja a fogópont-gombokat, a bal és jobb nyílbillentyűkkel pedig módosíthatja a kijelölést",
  "Use left and right arrows to move selection": "A bal és jobb nyílbillentyűkkel mozgathatja a kijelölést",
  "Use left and right arrows to move left selection": "A bal és jobb nyílbillentyűkkel mozgathatja a bal kijelölést",
  "Use left and right arrows to move right selection": "A bal és jobb nyílbillentyűkkel mozgathatja a jobb kijelölést",
  "Use TAB select grip buttons or up and down arrows to change selection": "A TAB billentyűvel kiválaszthatja a fogópont-gombokat, a fel és le nyílbillentyűkkel pedig módosíthatja a kijelölést",
  "Use up and down arrows to move selection": "A fel és le nyílbillentyűkkel mozgathatja a kijelölést",
  "Use up and down arrows to move lower selection": "A fel és le nyílbillentyűkkel mozgathatja az alsó kijelölést",
  "Use up and down arrows to move upper selection": "A fel és le nyílbillentyűkkel mozgathatja a felső kijelölést",
  "From %1 to %2": "Innen: %1, Eddig: %2",
  "From %1": "Innen: %1",
  "To %1": "Eddig: %1",
  "No parser available for file: %1": "Nem érhető el feldolgozó a fájlhoz: %1",
  "Error parsing file: %1": "Hiba a fájl feldolgozásakor: %1",
  "Unable to load file: %1": "Nem lehet betölteni a fájlt: %1",
  "Invalid date": "Érvénytelen dátum"
};
export {
  s as Chart,
  m as Data,
  r as Export,
  l as Home,
  i as Image,
  t as Legend,
  o as Loading,
  e as Play,
  g as Print,
  n as Series,
  a as Stop,
  d as default
};
