// @arcgis/charts-components/dist/chunks/messages.sv2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const a = { chart: "Diagram", barChart: "Liggande stapeldiagram", lineChart: "Linjediagram", pieChart: "Cirkeldiagram", histogram: "Histogram", scatterplot: "Sambandsdiagram", boxPlot: "Låddiagram" }, t = { mean: "Medelvärde", median: "Median", sum: "Summa", discretePercentile: "Diskret percentil", minimum: "Minimum", maximum: "Maximum", variance: "Varians", count: "Antal", aggregation: "${ statistics } av ${ fieldName }", noAggregation: "Ingen aggregering" }, e = "Diagram", i = "Antalsserie", n = "X-axel", s = "Ej tillgänglig", r = { start: "Snappa till den första datapunkten", end: "Snappa till den sista datapunkten" }, g = {
  chartType: a,
  statistics: t,
  defaultTitle: e,
  countSeries: i,
  xAxis: n,
  notAvailable: s,
  timeAggregationTypes: r
};
export {
  a as chartType,
  i as countSeries,
  g as default,
  e as defaultTitle,
  s as notAvailable,
  t as statistics,
  r as timeAggregationTypes,
  n as xAxis
};
