// @arcgis/charts-components/dist/chunks/messages.ar2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const t = { chart: "الرسم البياني", barChart: "المخطط الشريطي", lineChart: "مخطط خطي", pieChart: "مخطط دائري", histogram: "مدرج تكراري", scatterplot: "Scatterplot", boxPlot: "رسم مربع" }, e = { mean: "المتوسط", median: "متوسط", sum: "مجموع", discretePercentile: "النسبة المئوية المنفصلة", minimum: "الحد الأدنى", maximum: "الحد الأقصى", variance: "تنوع", count: "عدد", aggregation: "${ statistics } من ${ fieldName }", noAggregation: "لا يوجد تجميع" }, a = "الرسم البياني", s = "سلسلة العدد", i = "محور س", n = "غير مرئي", o = { start: "انطباق على نقطة البيانات الأولى", end: "انطباق على نقطة البيانات الأخيرة" }, c = {
  chartType: t,
  statistics: e,
  defaultTitle: a,
  countSeries: s,
  xAxis: i,
  notAvailable: n,
  timeAggregationTypes: o
};
export {
  t as chartType,
  s as countSeries,
  c as default,
  a as defaultTitle,
  n as notAvailable,
  e as statistics,
  o as timeAggregationTypes,
  i as xAxis
};
