// @arcgis/charts-components/dist/chunks/messages.ru4.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "Воспроизвести", o = "Остановить", t = "Легенда", r = "Загрузка", s = "На главную", a = "Диаграмма", n = "Серия", i = "Экспорт", l = "Изображение", c = "Данные", d = "Печать", E = {
  "Zoom Out": "Отдалить",
  Play: e,
  Stop: o,
  Legend: t,
  "Press ENTER to toggle": "Нажать ENTER, чтобы переключить",
  Loading: r,
  Home: s,
  Chart: a,
  "Serial chart": "Серийная диаграмма",
  "X/Y chart": "Диаграмма X/Y",
  "Pie chart": "Круговая диаграмма",
  "Gauge chart": "Диаграмма-датчик",
  "Radar chart": "Лепестковая диаграмма",
  "Sankey diagram": "Диаграмма Сэнки",
  "Flow diagram": "Диаграмма процесса",
  "Chord diagram": "Хордовая диаграмма",
  "TreeMap chart": "Диаграмма Treemap",
  "Sliced chart": "Ступенчатая диаграмма",
  Series: n,
  "Candlestick Series": 'Серия "японских свечей"',
  "OHLC Series": "Серия OHLC",
  "Column Series": "Серия столбцов",
  "Line Series": "Серия графиков",
  "Pie Slice Series": "Серия долей круга",
  "Funnel Series": "Серия воронок",
  "Pyramid Series": "Серия пирамид",
  "X/Y Series": "Серия X/Y",
  Map: "Карта",
  "Press ENTER to zoom in": "Нажмите ENTER, чтобы приблизиться",
  "Press ENTER to zoom out": "Нажмите ENTER, чтобы отдалиться",
  "Use arrow keys to zoom in and out": "Использовать клавиши со стрелками для масштабирования",
  "Use plus and minus keys on your keyboard to zoom in and out": "Использовать клавиши плюс и минус на клавиатуре для масштабирования",
  Export: i,
  Image: l,
  Data: c,
  Print: d,
  "Press ENTER to open": "Нажмите ENTER, чтобы открыть",
  "Press ENTER to print.": "Нажмите ENTER, чтобы напечатать.",
  "Press ENTER to export as %1.": "Нажмите ENTER, чтобы экспортировать как %1.",
  "(Press ESC to close this message)": "(Нажмите ESC, чтобы закрыть это сообщение)",
  "Image Export Complete": "Экспорт изображения завершен",
  "Export operation took longer than expected. Something might have gone wrong.": "Операция экспорта заняла больше времени, чем ожидалось. Возможно, что-то пошло не так.",
  "Saved from": "Сохранено из",
  "Use TAB to select grip buttons or left and right arrows to change selection": "Используйте TAB, чтобы выбрать кнопку захвата или стрелки вправо и влево для изменения выборки",
  "Use left and right arrows to move selection": "Используйте стрелки вправо и влево для перемещения выборки",
  "Use left and right arrows to move left selection": "Используйте стрелки вправо и влево для перемещения левой выборки",
  "Use left and right arrows to move right selection": "Используйте стрелки вправо и влево для перемещения правой выборки",
  "Use TAB select grip buttons or up and down arrows to change selection": "Используйте TAB, чтобы выбрать кнопку захвата или стрелки вверх и вниз для изменения выборки",
  "Use up and down arrows to move selection": "Используйте стрелки вверх и вниз для перемещения выборки",
  "Use up and down arrows to move lower selection": "Используйте стрелки вверх и вниз для перемещения нижней выборки",
  "Use up and down arrows to move upper selection": "Используйте стрелки вверх и вниз для перемещения верхней выборки",
  "From %1 to %2": "Из %1 в %2",
  "From %1": "Из %1",
  "To %1": "В %1",
  "No parser available for file: %1": "Нет доступного синтаксического анализатора для файла: %1",
  "Error parsing file: %1": "Ошибка разбора файла: %1",
  "Unable to load file: %1": "Невозможно загрузить файл: %1",
  "Invalid date": "Некорректная дата"
};
export {
  a as Chart,
  c as Data,
  i as Export,
  s as Home,
  l as Image,
  t as Legend,
  r as Loading,
  e as Play,
  d as Print,
  n as Series,
  o as Stop,
  E as default
};
