// @arcgis/charts-components/dist/chunks/messages.de4.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "Wiedergeben", n = "Beenden", r = "Legende", a = "Wird geladen", t = "Start", s = "Diagramm", i = "Reihe", o = "Exportieren", d = "Bild", u = "Daten", l = "Drucken", c = {
  "Zoom Out": "Verkleinern",
  Play: e,
  Stop: n,
  Legend: r,
  "Press ENTER to toggle": "Eingabetaste zum Umschalten drücken",
  Loading: a,
  Home: t,
  Chart: s,
  "Serial chart": "Reihen-Diagramm",
  "X/Y chart": "X/Y-Diagramm",
  "Pie chart": "Kreisdiagramm",
  "Gauge chart": "Skalendiagramm",
  "Radar chart": "Radar-Diagramm",
  "Sankey diagram": "Sankey-Diagramm",
  "Flow diagram": "Flussdiagramm",
  "Chord diagram": "Sehnendiagramm",
  "TreeMap chart": "Baumkartendiagramm",
  "Sliced chart": "Ausgeschnittenes Diagramm",
  Series: i,
  "Candlestick Series": "Kerzen-Reihe",
  "OHLC Series": "OHLC-Reihe",
  "Column Series": "Spalten-Reihe",
  "Line Series": "Linien-Reihe",
  "Pie Slice Series": "Segment-Reihe",
  "Funnel Series": "Trichter-Reihe",
  "Pyramid Series": "Pyramiden-Reihe",
  "X/Y Series": "X/Y-Reihe",
  Map: "Karte",
  "Press ENTER to zoom in": "Eingabetaste zum Vergrößern drücken",
  "Press ENTER to zoom out": "Eingabetaste zum Verkleinern drücken",
  "Use arrow keys to zoom in and out": "Pfeiltasten zum Vergrößern und Verkleinern drücken",
  "Use plus and minus keys on your keyboard to zoom in and out": "Plus- und Minus-Tasten auf der Tastatur zum Vergrößern und Verkleinern drücken",
  Export: o,
  Image: d,
  Data: u,
  Print: l,
  "Press ENTER to open": "Eingabetaste zum Öffnen drücken",
  "Press ENTER to print.": "Eingabetaste zum Drucken drücken",
  "Press ENTER to export as %1.": "Eingabetaste zum Exportieren als %1 drücken",
  "(Press ESC to close this message)": "Esc-Taste drücken, um diese Meldung zu schließen",
  "Image Export Complete": "Bildexport abgeschlossen",
  "Export operation took longer than expected. Something might have gone wrong.": "Export dauert länger als erwartet. Möglicherweise ist ein Fehler aufgetreten.",
  "Saved from": "Gespeichert aus",
  "Use TAB to select grip buttons or left and right arrows to change selection": "TAB-Taste zur Auswahl der Grifftasten verwenden, oder Nach-rechts- und Nach-links-Pfeil-Tasten verwenden, um die Auswahl zu ändern",
  "Use left and right arrows to move selection": "Nach-rechts- und Nach-links-Pfeil-Tasten verwenden, um die Auswahl zu verschieben",
  "Use left and right arrows to move left selection": "Nach-rechts- und Nach-links-Pfeil-Tasten verwenden, um linke Auswahl zu verschieben",
  "Use left and right arrows to move right selection": "Nach-rechts- und Nach-links-Pfeil-Tasten verwenden, um rechte Auswahl zu verschieben",
  "Use TAB select grip buttons or up and down arrows to change selection": "TAB-Taste zur Auswahl der Grifftasten verwenden, oder Nach-oben- und Nach-unten-Pfeil-Tasten verwenden, um die Auswahl zu ändern",
  "Use up and down arrows to move selection": "Nach-oben- und Nach-unten-Pfeil-Tasten verwenden, um die Auswahl zu verschieben",
  "Use up and down arrows to move lower selection": "Nach-oben- und Nach-unten-Pfeil-Tasten verwenden, um untere Auswahl zu verschieben",
  "Use up and down arrows to move upper selection": "Nach-oben- und Nach-unten-Pfeil-Tasten verwenden, um obere Auswahl zu verschieben",
  "From %1 to %2": "Von %1 bis %2",
  "From %1": "Von %1",
  "To %1": "Bis %1",
  "No parser available for file: %1": "Kein Parser für Datei verfügbar: %1",
  "Error parsing file: %1": "Fehler beim Parsen der Datei: %1",
  "Unable to load file: %1": "Datei kann nicht geladen werden: %1",
  "Invalid date": "Ungültiges Datum"
};
export {
  s as Chart,
  u as Data,
  o as Export,
  t as Home,
  d as Image,
  r as Legend,
  a as Loading,
  e as Play,
  l as Print,
  i as Series,
  n as Stop,
  c as default
};
