// @arcgis/charts-components/dist/chunks/messages.no.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "Det oppsto en feil under innlasting av diagrammet.", n = "Loggtransformasjon kan ikke brukes på negative verdier eller nullverdier.", t = "Kvadratrottransformasjon kan ikke brukes på negative verdier.", a = "Det oppsto en feil under innlasting av laget. URL = ${ url }. Portalelement-ID = ${ portalItemId }.", r = "Feltene [${ fieldNames }] som er definert i konfigurasjonen, ble ikke funnet i laget.", s = "${ dataPath } må være unik. Serien med navnet ${ seriesName } har en ID (${ seriesID }) som allerede brukes av en annen serie.", i = "${ dataPath } kan ikke utføre aggregering av annet enn antall på ikke-numeriske felt.", o = "${ dataPath } mangler en egenskap ved navn ${ missingProperty }.", l = "${ dataPath } må ikke være kortere enn ${ limit } tegn.", d = "${ dataPath } kan ikke inneholde færre enn ${ limit } elementer.", m = "${ limit } kan ikke inneholde flere enn ${ dataPath } elementer.", k = "${ dataPath } må inneholde minst ett tegn som ikke er mellomrom.", u = "${ dataPath } kan ikke ha ${ additionalProperty }.", g = "${ dataPath } må være lik én av følgende tillatte verdier: ${ allowedValues }.", p = "${ dataPath } må samsvare med skjemaet for én av følgende: ${ schemaOptions }.", c = "Punktdiagrammer med proporsjonale symboler støttes ikke. Standard symbolstørrelse er brukt.", v = "Kan ikke lese inndata.", $ = "Operasjonen queryBins støttes ikke på denne tjenesten.", f = "Forventet serietype ved indeks ${ seriesIndex } er ${ expectedType }, men ${ receivedType } ble mottatt i stedet", h = "eller", P = {
  defaultError: e,
  negativeValueInDataForLogTransformation: n,
  negativeValueInDataForSqrtTransformation: t,
  layerLoadFailure: a,
  layerFieldsNotFound: r,
  duplicateSeriesID: s,
  nonNumericAggregation: i,
  requiredProperty: o,
  minLength: l,
  minItems: d,
  maxItems: m,
  whiteSpacePattern: k,
  additionalProperty: u,
  enumValues: g,
  anyOfValues: p,
  bubbleChartValidateMsg: c,
  queryError: v,
  queryBinsNotSupported: $,
  invalidSeriesType: f,
  or: h
};
export {
  u as additionalProperty,
  p as anyOfValues,
  c as bubbleChartValidateMsg,
  P as default,
  e as defaultError,
  s as duplicateSeriesID,
  g as enumValues,
  f as invalidSeriesType,
  r as layerFieldsNotFound,
  a as layerLoadFailure,
  m as maxItems,
  d as minItems,
  l as minLength,
  n as negativeValueInDataForLogTransformation,
  t as negativeValueInDataForSqrtTransformation,
  i as nonNumericAggregation,
  h as or,
  $ as queryBinsNotSupported,
  v as queryError,
  o as requiredProperty,
  k as whiteSpacePattern
};
