// @arcgis/charts-components/dist/chunks/messages.th4.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "เล่น", o = "หยุด", t = "คำอธิบายสัญลักษณ์", r = "กำลังโหลด", s = "หน้าหลัก", a = "แผนภูมิ", n = "ชุด", i = "ส่งออก", l = "รูปภาพ", c = "ข้อมูล", d = "พิมพ์", E = {
  "Zoom Out": "ซูมออก",
  Play: e,
  Stop: o,
  Legend: t,
  "Press ENTER to toggle": "กด ENTER เพื่อสลับ",
  Loading: r,
  Home: s,
  Chart: a,
  "Serial chart": "แผนภูมิแบบอนุกรม",
  "X/Y chart": "แผนภูมิ X/Y",
  "Pie chart": "แผนภูมิทรงกลม",
  "Gauge chart": "แผนภูมิเกจ",
  "Radar chart": "แผนภูมิเรดาร์",
  "Sankey diagram": "ไดอะแกรม Sankey",
  "Flow diagram": "ไดอะแกรมโฟลว์",
  "Chord diagram": "ไดอะแกรมคอร์ด",
  "TreeMap chart": "แผนภูมิทรีแมป",
  "Sliced chart": "แผนภูมิสไลซ์",
  Series: n,
  "Candlestick Series": "ซีรีส์แท่งเทียน",
  "OHLC Series": "ซีรีส์ OHLC",
  "Column Series": "ซีรีส์คอลัมน์",
  "Line Series": "ซีรีส์เส้น",
  "Pie Slice Series": "ซีรีส์สไลซ์วงกลม",
  "Funnel Series": "ซีรีส์ฟันเนล",
  "Pyramid Series": "ซีรีส์พีรามิด",
  "X/Y Series": "ซีรีส์ X/Y",
  Map: "แผนที่",
  "Press ENTER to zoom in": "กด ENTER เพื่อซูมเข้า",
  "Press ENTER to zoom out": "กด ENTER เพื่อซูมออก",
  "Use arrow keys to zoom in and out": "ใช้ปุ่มลูกศรในการซูมเข้าและออก",
  "Use plus and minus keys on your keyboard to zoom in and out": "ใช้ปุ่มบวกและลบบนแป้นพิมพ์ของคุณเพื่อซูมเข้าและออก",
  Export: i,
  Image: l,
  Data: c,
  Print: d,
  "Press ENTER to open": "กด ENTER เพื่อเปิด",
  "Press ENTER to print.": "กด ENTER เพื่อพิมพ์",
  "Press ENTER to export as %1.": "กด ENTER เพื่อส่งออกในรูปแบบ %1",
  "(Press ESC to close this message)": "(กด ESC เพื่อปิดข้อความนี้)",
  "Image Export Complete": "ส่งออกภาพเสร็จสมบูรณ์",
  "Export operation took longer than expected. Something might have gone wrong.": "การดำเนินการส่งออกใช้เวลานานกว่าที่คาดไว้ อาจเกิดความผิดปกติบางอย่าง",
  "Saved from": "บันทึกจาก",
  "Use TAB to select grip buttons or left and right arrows to change selection": "ใช้ TAB เพื่อเลือกปุ่มกริป หรือลูกศรซ้ายและขวาเพื่อเปลี่ยนแปลงการเลือก",
  "Use left and right arrows to move selection": "ใช้ลูกศรซ้ายและขวาเพื่อย้ายการเลือก",
  "Use left and right arrows to move left selection": "ใช้ลูกศรซ้ายและขวาเพื่อย้ายการเลือกทางซ้าย",
  "Use left and right arrows to move right selection": "ใช้ลูกศรซ้ายและขวาเพื่อย้ายการเลือกทางขวา",
  "Use TAB select grip buttons or up and down arrows to change selection": "ใช้ TAB เพื่อเลือกปุ่มกริป หรือลูกศรขึ้นและลงเพื่อเปลี่ยนแปลงการเลือก",
  "Use up and down arrows to move selection": "ใช้ลูกศรขึ้นและลงเพื่อย้ายการเลือก",
  "Use up and down arrows to move lower selection": "ใช้ลูกศรขึ้นและลงเพื่อย้ายการเลือกด้านล่าง",
  "Use up and down arrows to move upper selection": "ใช้ลูกศรขึ้นและลงเพื่อย้ายการเลือกด้านบน",
  "From %1 to %2": "จาก %1 เป็น %2",
  "From %1": "จาก %1",
  "To %1": "เป็น %1",
  "No parser available for file: %1": "ไม่มีตัวแยกวิเคราะห์ที่พร้อมใช้งานสำหรับไฟล์: %1",
  "Error parsing file: %1": "เกิดข้อผิดพลาดในการแยกวิเคราะห์ไฟล์: %1",
  "Unable to load file: %1": "ไม่สามารถโหลดไฟล์: %1",
  "Invalid date": "วันที่ใช้งานไม่ได้"
};
export {
  a as Chart,
  c as Data,
  i as Export,
  s as Home,
  l as Image,
  t as Legend,
  r as Loading,
  e as Play,
  d as Print,
  n as Series,
  o as Stop,
  E as default
};
