// @arcgis/charts-components/dist/chunks/messages.lv4.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const a = "Demonstrēt", t = "Apturēt", i = "Apzīmējumi", e = "Ielādē", s = "Sākums", r = "Diagramma", o = "Sērijas", u = "Eksportēt", l = "Attēls", n = "Dati", m = "Drukāt", d = {
  "Zoom Out": "Attālināt",
  Play: a,
  Stop: t,
  Legend: i,
  "Press ENTER to toggle": "Nospiediet ENTER, lai pārslēgtu",
  Loading: e,
  Home: s,
  Chart: r,
  "Serial chart": "Sērijas diagramma",
  "X/Y chart": "X/Y diagramma",
  "Pie chart": "Sektoru diagramma",
  "Gauge chart": "Skalas diagramma",
  "Radar chart": "Radaru diagramma",
  "Sankey diagram": "Sankija diagramma",
  "Flow diagram": "Plūsmas diagramma",
  "Chord diagram": "Hordas diagramma",
  "TreeMap chart": "TreeMap diagramma",
  "Sliced chart": "Sagrieztu sektoru diagramma",
  Series: o,
  "Candlestick Series": "Sveču diagrammas sērijas",
  "OHLC Series": "OHLC sērijas",
  "Column Series": "Kolonnu sērijas",
  "Line Series": "Līniju diagrammas sērijas",
  "Pie Slice Series": "Sektoru diagrammas sērijas",
  "Funnel Series": "Tuneļu diagrammas sērijas",
  "Pyramid Series": "Piramīdas diagrammas sērijas",
  "X/Y Series": "X/Y sērijas",
  Map: "Karte",
  "Press ENTER to zoom in": "Nospiediet ENTER, lai tuvinātu",
  "Press ENTER to zoom out": "Nospiediet ENTER, lai attālinātu",
  "Use arrow keys to zoom in and out": "Izmantojiet bulttaustiņus, tuvinātu un attālinātu",
  "Use plus and minus keys on your keyboard to zoom in and out": "Izmantojot plusa vai mīnusa taustiņus uz klaviatūras, lai pakāpeniski tuvinātu un tālinātu.",
  Export: u,
  Image: l,
  Data: n,
  Print: m,
  "Press ENTER to open": "Nospiediet ENTER, lai atvērtu",
  "Press ENTER to print.": "Nospiediet ENTER, lai drukātu.",
  "Press ENTER to export as %1.": "Nospiediet ENTER, lai ēksportētu kā %1.",
  "(Press ESC to close this message)": "(Nospiediet ESC, lai aizvērtu šo ziņu)",
  "Image Export Complete": "Attēlu eksportēšana ir pabeigta",
  "Export operation took longer than expected. Something might have gone wrong.": "Darbības izpilde bija ilgāka, nekā paredzēts. Kaut kas, iespējams, neizdevās.",
  "Saved from": "Saglabāts no",
  "Use TAB to select grip buttons or left and right arrows to change selection": "Izmantojiet TAB, lai izvēlētos saķeres pogas, vai pa kreisi un pa labi vērstās bultas, lai mainītu izvēli",
  "Use left and right arrows to move selection": "Izmantojiet pa kreisi un pa labi vērstās bultas, lai pārvietotu izvēli",
  "Use left and right arrows to move left selection": "Izmantojiet pa kreisi un pa labi vērstās bultas, lai pārvietotu kreisās puses izvēli",
  "Use left and right arrows to move right selection": "Izmantojiet pa kreisi un pa labi vērstās bultas, lai pārvietotu labās puses izvēli",
  "Use TAB select grip buttons or up and down arrows to change selection": "Izmantojiet TAB, lai izvēlētos saķeres pogas, vai uz augšu un uz leju vērstās bultas, lai mainītu izvēli",
  "Use up and down arrows to move selection": "Izmantojiet uz augšu un uz leju vērstās bultas, lai pārvietotu izvēli",
  "Use up and down arrows to move lower selection": "Izmantojiet uz augšu un uz leju vērstās bultas, lai pārvietotu zemāko izvēli",
  "Use up and down arrows to move upper selection": "Izmantojiet augšup un lejup vērstās bultas, lai pārvietotu augšējo izvēli",
  "From %1 to %2": "No %1 līdz %2",
  "From %1": "No %1",
  "To %1": "Līdz %1",
  "No parser available for file: %1": "Parsētājs failam nav pieejams: %1",
  "Error parsing file: %1": "Kļūda parsējot failu: %1",
  "Unable to load file: %1": "Nevar ielādēt failu: %1",
  "Invalid date": "Nederīgs datums"
};
export {
  r as Chart,
  n as Data,
  u as Export,
  s as Home,
  l as Image,
  i as Legend,
  e as Loading,
  a as Play,
  m as Print,
  o as Series,
  t as Stop,
  d as default
};
