// @arcgis/charts-components/dist/chunks/messages.sl2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const a = { chart: "Grafikon", barChart: "Palični grafikon", lineChart: "Linijski grafikon", pieChart: "Tortni grafikon", histogram: "Histogram", scatterplot: "Raztreseni grafikon", boxPlot: "Škatlasti graf" }, i = { mean: "Srednja vrednost", median: "Mediana", sum: "Vsota", discretePercentile: "Diskretni percentil", minimum: "Minimum", maximum: "Maksimum", variance: "Varianca", count: "Število", aggregation: "${ statistics } od ${ fieldName }", noAggregation: "Brez agregacije" }, t = "Grafikon", o = "Števec nizov", n = "Os X", e = "Ni na voljo", r = { start: "Privlači na prvo podatkovno točko", end: "Privlači na zadnjo podatkovno točko" }, s = {
  chartType: a,
  statistics: i,
  defaultTitle: t,
  countSeries: o,
  xAxis: n,
  notAvailable: e,
  timeAggregationTypes: r
};
export {
  a as chartType,
  o as countSeries,
  s as default,
  t as defaultTitle,
  e as notAvailable,
  i as statistics,
  r as timeAggregationTypes,
  n as xAxis
};
