// @arcgis/charts-components/dist/chunks/messages.sr4.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "Reprodukuj", a = "Zaustavi", i = "Legenda", t = "Učitavanje", r = "Matična stranica", o = "Grafikon", s = "Serija", n = "Izvezi", l = "Snimak", m = "Podaci", d = "Štampaj", p = {
  "Zoom Out": "Umanji",
  Play: e,
  Stop: a,
  Legend: i,
  "Press ENTER to toggle": "Pritisnite taster ENTER da biste uključili/isključili",
  Loading: t,
  Home: r,
  Chart: o,
  "Serial chart": "Serijski grafikon",
  "X/Y chart": "X/Y grafikon",
  "Pie chart": "Kružni grafikon",
  "Gauge chart": "Grafikon sa meračem",
  "Radar chart": "Radarski grafikon",
  "Sankey diagram": "Sankey dijagram",
  "Flow diagram": "Dijagram toka",
  "Chord diagram": "Trakasti dijagram",
  "TreeMap chart": "TreeMap grafikon",
  "Sliced chart": "Isečeni grafikon",
  Series: s,
  "Candlestick Series": "Serija svećica",
  "OHLC Series": "OHLC serija",
  "Column Series": "Serija kolona",
  "Line Series": "Serija linija",
  "Pie Slice Series": "Serija isečaka kružnog grafikona",
  "Funnel Series": "Serija levkova",
  "Pyramid Series": "Serija piramida",
  "X/Y Series": "X/Y serija",
  Map: "Mapa",
  "Press ENTER to zoom in": "Pritisnite taster ENTER da biste uvećali zumiranjem",
  "Press ENTER to zoom out": "Pritisnite taster ENTER da biste umanjili zumiranjem",
  "Use arrow keys to zoom in and out": "Upotrebite tastere sa strelicama da biste uvećavali ili umanjivali zumiranjem",
  "Use plus and minus keys on your keyboard to zoom in and out": "Upotrebite tastere plus i minus na tastaturi da biste uvećavali ili umanjivali zumiranjem",
  Export: n,
  Image: l,
  Data: m,
  Print: d,
  "Press ENTER to open": "Pritisnite taster ENTER da biste otvorili",
  "Press ENTER to print.": "Pritisnite taster ENTER da biste odštampali.",
  "Press ENTER to export as %1.": "Pritisnite taster ENTER da biste izvezli kao %1.",
  "(Press ESC to close this message)": "(Pritisnite taster ESC da biste zatvorili ovu poruku)",
  "Image Export Complete": "Izvoz snimka je završen",
  "Export operation took longer than expected. Something might have gone wrong.": "Operacija izvoza je trajala duže od očekivanog. Moguće je da nešto nije bilo u redu.",
  "Saved from": "Sačuvano iz",
  "Use TAB to select grip buttons or left and right arrows to change selection": "Upotrebite taster TAB da biste izabrali dugmad za hvatanje ili tastere sa levom i desnom strelicom da biste izmenili izbor",
  "Use left and right arrows to move selection": "Upotrebite tastere sa levom i desnom strelicom da biste premestili izbor",
  "Use left and right arrows to move left selection": "Upotrebite tastere sa levom i desnom strelicom da biste premestili levi izbor",
  "Use left and right arrows to move right selection": "Upotrebite tastere sa levom i desnom strelicom da biste premestili desni izbor",
  "Use TAB select grip buttons or up and down arrows to change selection": "Upotrebite taster TAB da biste izabrali dugmad za hvatanje ili tastere sa strelicom prema gore i prema dole da biste izmenili izbor",
  "Use up and down arrows to move selection": "Upotrebite tastere sa strelicom prema gore i prema dole da biste premestili izbor",
  "Use up and down arrows to move lower selection": "Upotrebite tastere sa strelicom prema gore i prema dole da biste premestili donji izbor",
  "Use up and down arrows to move upper selection": "Upotrebite tastere sa strelicom prema gore i prema dole da biste premestili gornji izbor",
  "From %1 to %2": "Od %1 ddo %2",
  "From %1": "Od %1",
  "To %1": "Do %1",
  "No parser available for file: %1": "Nijedan raščlanjivač nije dostupan za datoteku: %1",
  "Error parsing file: %1": "Greška prilikom raščlanjivanja datoteke: %1",
  "Unable to load file: %1": "Nije moguće učitati datoteku: %1",
  "Invalid date": "Nevažeći datum"
};
export {
  o as Chart,
  m as Data,
  n as Export,
  r as Home,
  l as Image,
  i as Legend,
  t as Loading,
  e as Play,
  d as Print,
  s as Series,
  a as Stop,
  p as default
};
