// @arcgis/charts-components/dist/chunks/messages.sk4.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "Spustiť", o = "Stop", a = "Legenda", r = "Načítava sa", t = "Domov", s = "Graf", i = "Série", n = "Export", p = "Snímka", l = "Dáta", v = "Tlačiť", d = {
  "Zoom Out": "Oddialiť",
  Play: e,
  Stop: o,
  Legend: a,
  "Press ENTER to toggle": "Stlačte ENTER pre prepnutie",
  Loading: r,
  Home: t,
  Chart: s,
  "Serial chart": "Sériový graf",
  "X/Y chart": "Graf X/Y",
  "Pie chart": "Koláčový graf",
  "Gauge chart": "Manometrický graf",
  "Radar chart": "Radarový graf",
  "Sankey diagram": "Sankey diagram",
  "Flow diagram": "Diagram prúdenia",
  "Chord diagram": "Tetivový diagram",
  "TreeMap chart": "TreeMap graf",
  "Sliced chart": "Segmentový graf",
  Series: i,
  "Candlestick Series": "Candlestick séria",
  "OHLC Series": "OHLC séria",
  "Column Series": "Stĺpcová séria",
  "Line Series": "Líniová séria",
  "Pie Slice Series": "Koláčová séria",
  "Funnel Series": "Lieviková séria",
  "Pyramid Series": "Pyramídová séria",
  "X/Y Series": "Séria X/Y",
  Map: "Mapa",
  "Press ENTER to zoom in": "Stlačte ENTER pre zväčšenie",
  "Press ENTER to zoom out": "Stlačte ENTER pre zmenšenie",
  "Use arrow keys to zoom in and out": "Použite šípky na priblíženie a oddialenie",
  "Use plus and minus keys on your keyboard to zoom in and out": "Pomocou klávesov plus a mínus na klávesnici môžete priblížiť alebo oddialiť zobrazenie",
  Export: n,
  Image: p,
  Data: l,
  Print: v,
  "Press ENTER to open": "Stlačte ENTER pre otvorenie",
  "Press ENTER to print.": "Stlačte ENTER pre tlač.",
  "Press ENTER to export as %1.": "Stlačte ENTER pre export ako %1.",
  "(Press ESC to close this message)": "(Stlačte ESC pre zatvorenie tejto správy)",
  "Image Export Complete": "Export obrázkov dokončený",
  "Export operation took longer than expected. Something might have gone wrong.": "Operácia exportu trvala dlhšie, ako sa očakávalo. Niečo sa mohlo pokaziť.",
  "Saved from": "Uložené z",
  "Use TAB to select grip buttons or left and right arrows to change selection": "Pomocou klávesy TAB vyberte tlačidlá uchopenia alebo pomocou šípok vľavo a vpravo zmeňte výber",
  "Use left and right arrows to move selection": "Pomocou šípok vľavo a vpravo presuňte výber",
  "Use left and right arrows to move left selection": "Pomocou šípok vľavo a vpravo presuňte výber doľava",
  "Use left and right arrows to move right selection": "Pomocou šípok vľavo a vpravo presuňte výber doprava",
  "Use TAB select grip buttons or up and down arrows to change selection": "Pomocou klávesy TAB vyberte tlačidlá uchopenia alebo pomocou šípok hore a dole zmeňte výber",
  "Use up and down arrows to move selection": "Pomocou šípok hore a dole presuňte výber",
  "Use up and down arrows to move lower selection": "Pomocou šípok hore a dole presuňte výber nižšie",
  "Use up and down arrows to move upper selection": "Pomocou šípok hore a dole presuňte výber vyššie",
  "From %1 to %2": "Od %1 do %2",
  "From %1": "Od %1",
  "To %1": "Do %1",
  "No parser available for file: %1": "Nie je k dispozícii parser pre súbor: %1",
  "Error parsing file: %1": "Chyba pri parsovaní súboru: %1",
  "Unable to load file: %1": "Nie je možné načítať súbor: %1",
  "Invalid date": "Neplatný dátum"
};
export {
  s as Chart,
  l as Data,
  n as Export,
  t as Home,
  p as Image,
  a as Legend,
  r as Loading,
  e as Play,
  v as Print,
  i as Series,
  o as Stop,
  d as default
};
