// @arcgis/charts-components/dist/chunks/messages.ca.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "S'ha produït un error en carregar el gràfic.", a = "No es pot aplicar una transformació del registre a valors negatius o de zero.", t = "No es pot aplicar una transformació d'arrel quadrada a valors negatius.", r = "S'ha produït un error en carregar la capa. URL = ${ url }. ID d'element del portal = ${ portalItemId }.", n = "Els camps [${ fieldNames }] definits a la configuració no s'han trobat a la capa.", s = "L'element ${ dataPath } ha de ser únic. L'ID (${ seriesID }) de la sèrie anomenada ${ seriesName } ja l'utilitza una altra sèrie.", o = "${ dataPath } no pot dur a terme una agregació que no sigui de recompte en un camp que no sigui numèric.", i = "A ${ dataPath } li falta una propietat anomenada ${ missingProperty }.", d = "${ dataPath } no pot tenir menys de ${ limit } caràcters.", l = "${ dataPath } no pot tenir menys de ${ limit } elements.", c = "${ dataPath } no pot tenir més de ${ limit } elements.", u = "${ dataPath } ha de tenir almenys un caràcter que no sigui un espai en blanc.", p = "${ dataPath } no pot tenir ${ additionalProperty }.", m = "${ dataPath } ha de ser igual a un d'aquests valors permesos: ${ allowedValues }.", $ = "${ dataPath } ha de coincidir amb l'esquema d'un d'aquests: ${ schemaOptions }.", g = "No s'admeten gràfics de dispersió amb símbols proporcionals. S'ha aplicat la mida de símbol per defecte.", h = "Error de lectura de les dades d'entrada.", y = "L'operació queryBins no s'admet en aquest servei.", P = 'El tipus de sèrie esperat a l\'índex ${ seriesIndex } és "${ expectedType }", però s\'ha rebut el tipus "${ receivedType }"', f = "o", q = {
  defaultError: e,
  negativeValueInDataForLogTransformation: a,
  negativeValueInDataForSqrtTransformation: t,
  layerLoadFailure: r,
  layerFieldsNotFound: n,
  duplicateSeriesID: s,
  nonNumericAggregation: o,
  requiredProperty: i,
  minLength: d,
  minItems: l,
  maxItems: c,
  whiteSpacePattern: u,
  additionalProperty: p,
  enumValues: m,
  anyOfValues: $,
  bubbleChartValidateMsg: g,
  queryError: h,
  queryBinsNotSupported: y,
  invalidSeriesType: P,
  or: f
};
export {
  p as additionalProperty,
  $ as anyOfValues,
  g as bubbleChartValidateMsg,
  q as default,
  e as defaultError,
  s as duplicateSeriesID,
  m as enumValues,
  P as invalidSeriesType,
  n as layerFieldsNotFound,
  r as layerLoadFailure,
  c as maxItems,
  l as minItems,
  d as minLength,
  a as negativeValueInDataForLogTransformation,
  t as negativeValueInDataForSqrtTransformation,
  o as nonNumericAggregation,
  f as or,
  y as queryBinsNotSupported,
  h as queryError,
  i as requiredProperty,
  u as whiteSpacePattern
};
