// @arcgis/charts-components/dist/chunks/messages.es4.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "Reproducir", a = "Detener", r = "Leyenda", o = "Cargando", i = "Inicio", s = "Gráfico", l = "Serie", t = "Exportar", c = "Imagen", n = "Datos", d = "Imprimir", p = {
  "Zoom Out": "Alejar",
  Play: e,
  Stop: a,
  Legend: r,
  "Press ENTER to toggle": "Pulse Intro para activar o desactivar",
  Loading: o,
  Home: i,
  Chart: s,
  "Serial chart": "Gráfico de serie",
  "X/Y chart": "Gráfico X/Y",
  "Pie chart": "Gráfico circular",
  "Gauge chart": "Gráfico de medidor",
  "Radar chart": "Gráfico de radar",
  "Sankey diagram": "Diagrama de Sankey",
  "Flow diagram": "Diagrama de flujo",
  "Chord diagram": "Diagrama de cuerdas",
  "TreeMap chart": "Gráfico de mapa de árbol",
  "Sliced chart": "Gráfico de sectores",
  Series: l,
  "Candlestick Series": "Serie de velas",
  "OHLC Series": "Serie OHLC",
  "Column Series": "Serie de columnas",
  "Line Series": "Serie de líneas",
  "Pie Slice Series": "Serie de sectores",
  "Funnel Series": "Serie de embudos",
  "Pyramid Series": "Serie de pirámides",
  "X/Y Series": "Serie X/Y",
  Map: "Mapa",
  "Press ENTER to zoom in": "Pulsar Intro para acercar",
  "Press ENTER to zoom out": "Pulsar Intro para alejar",
  "Use arrow keys to zoom in and out": "Utilizar las teclas de flecha para acercar y alejar",
  "Use plus and minus keys on your keyboard to zoom in and out": "Utilizar las teclas de signo más y menos del teclado para acercar y alejar",
  Export: t,
  Image: c,
  Data: n,
  Print: d,
  "Press ENTER to open": "Pulsar Intro para abrir",
  "Press ENTER to print.": "Pulse Intro para imprimir.",
  "Press ENTER to export as %1.": "Pulse Intro para exportar como %1.",
  "(Press ESC to close this message)": "(Pulse Esc para cerrar este mensaje)",
  "Image Export Complete": "Exportación de imagen completa",
  "Export operation took longer than expected. Something might have gone wrong.": "La operación de exportación ha durado más de lo previsto. Es posible que haya fallado algo.",
  "Saved from": "Guardado desde",
  "Use TAB to select grip buttons or left and right arrows to change selection": "Utilizar la tecla Tab para seleccionar los botones de control o las flechas izquierda y derecha para cambiar la selección",
  "Use left and right arrows to move selection": "Utilizar las flechas izquierda y derecha para mover la selección",
  "Use left and right arrows to move left selection": "Utilizar las flechas izquierda y derecha para mover la selección izquierda",
  "Use left and right arrows to move right selection": "Utilizar las flechas izquierda y derecha para mover la selección derecha",
  "Use TAB select grip buttons or up and down arrows to change selection": "Utilizar la tecla Tab para seleccionar los botones de control o las flechas arriba y abajo para cambiar la selección",
  "Use up and down arrows to move selection": "Utilizar las flechas arriba y abajo para mover la selección",
  "Use up and down arrows to move lower selection": "Utilizar las flechas arriba y abajo para mover la selección inferior",
  "Use up and down arrows to move upper selection": "Utilizar las flechas arriba y abajo para mover la selección superior",
  "From %1 to %2": "Desde %1 hasta %2",
  "From %1": "Desde %1",
  "To %1": "Hasta %1",
  "No parser available for file: %1": "Ningún analizador disponible para el archivo: %1",
  "Error parsing file: %1": "Error al analizar el archivo: %1",
  "Unable to load file: %1": "Imposible cargar el archivo: %1",
  "Invalid date": "Fecha no válida"
};
export {
  s as Chart,
  n as Data,
  t as Export,
  i as Home,
  c as Image,
  r as Legend,
  o as Loading,
  e as Play,
  d as Print,
  l as Series,
  a as Stop,
  p as default
};
