// @arcgis/charts-components/dist/chunks/messages.el2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const t = { chart: "Γράφημα", barChart: "Γράφημα ράβδων", lineChart: "Γράφημα γραμμών", pieChart: "Γράφημα πίτας", histogram: "Ιστόγραμμα", scatterplot: "Διάγραμμα διασποράς", boxPlot: "Διάγραμμα πλαισίου" }, e = { mean: "Μέσος όρος", median: "Διάμεσο", sum: "Άθροισμα", discretePercentile: "Διακριτό ποσοστημόριο", minimum: "Ελάχιστο", maximum: "Μέγιστο", variance: "Διακύμανση", count: "Πλήθος", aggregation: "${ statistics } από ${ fieldName }", noAggregation: "Δεν υπάρχει συνάθροιση" }, a = "Γράφημα", s = "Σειρά μέτρησης", i = "Άξονας Χ", n = "Μη διαθέσιμο", o = { start: "Συνδεθείτε με το πρώτο σημείο δεδομένων", end: "Συνδεθείτε με το τελευταίο σημείο δεδομένων" }, c = {
  chartType: t,
  statistics: e,
  defaultTitle: a,
  countSeries: s,
  xAxis: i,
  notAvailable: n,
  timeAggregationTypes: o
};
export {
  t as chartType,
  s as countSeries,
  c as default,
  a as defaultTitle,
  n as notAvailable,
  e as statistics,
  o as timeAggregationTypes,
  i as xAxis
};
