// @arcgis/charts-components/dist/chunks/messages.id2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const a = { chart: "Diagram", barChart: "Diagram batang", lineChart: "Diagram garis", pieChart: "Diagram pai", histogram: "Histogram", scatterplot: "Scatter plot", boxPlot: "Plot kotak" }, t = { mean: "Rata-rata", median: "Median", sum: "Sum", discretePercentile: "Persentil diskrit", minimum: "Minimal", maximum: "Maksimal", variance: "Varian", count: "Count", aggregation: "${ statistics } dari ${ fieldName }", noAggregation: "Tidak ada agregasi" }, i = "Diagram", e = "Jumlah Rangkaian", s = "Sumbu X", r = "Tidak Tersedia", n = { start: "Posisikan ke titik data pertama", end: "Posisikan ke titik data terakhir" }, o = {
  chartType: a,
  statistics: t,
  defaultTitle: i,
  countSeries: e,
  xAxis: s,
  notAvailable: r,
  timeAggregationTypes: n
};
export {
  a as chartType,
  e as countSeries,
  o as default,
  i as defaultTitle,
  r as notAvailable,
  t as statistics,
  n as timeAggregationTypes,
  s as xAxis
};
