// @arcgis/charts-components/dist/chunks/messages.it.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "Si è verificato un errore durante il caricamento del grafico.", a = "Impossibile applicare la trasformazione di log su valori negativi o zero.", t = "Impossibile applicare la trasformazione di radice quadrata su valori negativi.", i = "Si è verificato un errore durante il caricamento del layer. URL = ${ url }. ID elemento del portale = ${ portalItemId }.", o = "I campi [${ fieldNames }] definiti nel config non sono stati trovati nel layer.", n = "${ dataPath } deve essere univoco. La serie denominata ${ seriesName } ha un ID (${ seriesID }) che è già utilizzato da un'altra serie.", r = "${ dataPath } non può effettuare l'aggregazione senza conteggio su un campo non numerico.", s = "A ${ dataPath } manca una proprietà denominata ${ missingProperty }.", d = "${ dataPath } non deve essere più corto di ${ limit } caratteri.", l = "${ dataPath } non deve contenere meno di ${ limit } elementi.", c = "${ dataPath } non deve contenere più di ${ limit } elementi.", u = "${ dataPath } deve contenere almeno un carattere.", p = "${ dataPath } non deve contenere ${ additionalProperty }.", m = "${ dataPath } deve essere uguale a uno di questi valori consentiti: ${ allowedValues }.", g = "${ dataPath } deve corrispondere allo schema di uno di questi‎: ${ schemaOptions }.", $ = "I grafici a dispersione con simboli proporzionali non sono supportati. Sono state applicate delle dimensioni simbolo predefinite.", v = "Impossibile leggere i dati di input.", f = "L'operazione queryBins non è supportata da questo servizio.", h = "Il tipo di serie atteso all'indice ${ seriesIndex } è '${ expectedType }' ma invece è stato ricevuto '${ receivedType }'", I = "oppure", y = {
  defaultError: e,
  negativeValueInDataForLogTransformation: a,
  negativeValueInDataForSqrtTransformation: t,
  layerLoadFailure: i,
  layerFieldsNotFound: o,
  duplicateSeriesID: n,
  nonNumericAggregation: r,
  requiredProperty: s,
  minLength: d,
  minItems: l,
  maxItems: c,
  whiteSpacePattern: u,
  additionalProperty: p,
  enumValues: m,
  anyOfValues: g,
  bubbleChartValidateMsg: $,
  queryError: v,
  queryBinsNotSupported: f,
  invalidSeriesType: h,
  or: I
};
export {
  p as additionalProperty,
  g as anyOfValues,
  $ as bubbleChartValidateMsg,
  y as default,
  e as defaultError,
  n as duplicateSeriesID,
  m as enumValues,
  h as invalidSeriesType,
  o as layerFieldsNotFound,
  i as layerLoadFailure,
  c as maxItems,
  l as minItems,
  d as minLength,
  a as negativeValueInDataForLogTransformation,
  t as negativeValueInDataForSqrtTransformation,
  r as nonNumericAggregation,
  I as or,
  f as queryBinsNotSupported,
  v as queryError,
  s as requiredProperty,
  u as whiteSpacePattern
};
