// @arcgis/charts-components/dist/chunks/messages.zh-TW.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const t = "載入圖表時出錯。", a = "無法將對數轉換套用於負值或零值。", e = "無法將平方根轉換套用於負值。", s = "載入圖層時出錯。 URL = ${ url }。 入口網站項目 ID = ${ portalItemId }。", o = "在圖層中找不到配置中定義的欄位 [${ fieldNames }]。", n = "${ dataPath } 必須唯一。 名為 ${ seriesName } 的數列具有其他數列已在使用的 ID (${ seriesID })。", r = "${ dataPath } 無法在非數值欄位上執行非計數匯聚。", i = "${ dataPath } 缺少名為 ${ missingProperty } 的屬性。", d = "${ dataPath } 不應短於 ${ limit } 個字元。", c = "${ dataPath } 的項目數不應少於 ${ limit } 個。", l = "${ dataPath } 的項目數不應多於 ${ limit } 個。", $ = "${ dataPath } 至少必須有一個非空白字元。", u = "${ dataPath } 不能有 ${ additionalProperty }。", m = "${ dataPath } 必須等於下列其中一個允許值: ${ allowedValues }。", h = "${ dataPath } 必須符合下列其中一項結構描述: ${ schemaOptions }。", p = "不支援包含比例符號的散點圖。 已套用預設符號大小。", P = "無法讀取輸入資料。", y = "此服務不支援 queryBins 操作。", I = "索引 ${ seriesIndex } 上預期的數列類型為 '${ expectedType }'，但改為接收 '${ receivedType }'", g = "或", f = {
  defaultError: t,
  negativeValueInDataForLogTransformation: a,
  negativeValueInDataForSqrtTransformation: e,
  layerLoadFailure: s,
  layerFieldsNotFound: o,
  duplicateSeriesID: n,
  nonNumericAggregation: r,
  requiredProperty: i,
  minLength: d,
  minItems: c,
  maxItems: l,
  whiteSpacePattern: $,
  additionalProperty: u,
  enumValues: m,
  anyOfValues: h,
  bubbleChartValidateMsg: p,
  queryError: P,
  queryBinsNotSupported: y,
  invalidSeriesType: I,
  or: g
};
export {
  u as additionalProperty,
  h as anyOfValues,
  p as bubbleChartValidateMsg,
  f as default,
  t as defaultError,
  n as duplicateSeriesID,
  m as enumValues,
  I as invalidSeriesType,
  o as layerFieldsNotFound,
  s as layerLoadFailure,
  l as maxItems,
  c as minItems,
  d as minLength,
  a as negativeValueInDataForLogTransformation,
  e as negativeValueInDataForSqrtTransformation,
  r as nonNumericAggregation,
  g as or,
  y as queryBinsNotSupported,
  P as queryError,
  i as requiredProperty,
  $ as whiteSpacePattern
};
