// @arcgis/charts-components/dist/chunks/messages.el.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const t = "Παρουσιάστηκε κάποιο σφάλμα κατά τη φόρτωση του γραφήματος.", a = "Δεν είναι δυνατή η εφαρμογή του μετασχηματισμού αρχείου καταγραφής σε αρνητικές ή μηδενικές τιμές.", e = "Δεν είναι δυνατή η εφαρμογή του μετασχηματισμού τετραγωνικής ρίζας σε αρνητικές τιμές.", o = "Παρουσιάστηκε κάποιο σφάλμα κατά τη φόρτωση του θεματικού επιπέδου. Διεύθυνση URL = ${ url }. Αναγνωριστικό αντικειμένου του portal = ${ portalItemId }.", s = "Τα πεδία [${ fieldNames }] που ορίστηκαν στη διαμόρφωση δεν βρέθηκαν στο θεματικό επίπεδο.", n = "Το ${ dataPath } πρέπει να είναι μοναδικό. Η σειρά με το όνομα ${ seriesName } έχει αναγνωριστικό (${ seriesID }) που ήδη χρησιμοποιείται από κάποια άλλη σειρά.", r = "Το ${ dataPath } δεν μπορεί να εκτελέσει συνάθροιση χωρίς μέτρηση σε μη αριθμητικό πεδίο.", i = "Από το ${ dataPath } λείπει μια ιδιότητα με το όνομα ${ missingProperty }.", d = "Το ${ dataPath } δεν πρέπει να έχει λιγότερους από ${ limit } χαρακτήρες.", c = "Το ${ dataPath } δεν πρέπει να έχει λιγότερα από ${ limit } στοιχεία.", l = "Το ${ dataPath } δεν πρέπει να έχει περισσότερα από ${ limit } στοιχεία.", $ = "Το ${ dataPath } πρέπει να έχει τουλάχιστον έναν χαρακτήρα που να μην είναι κενό διάστημα.", u = "Το ${ dataPath } δεν πρέπει να έχει ${ additionalProperty }.", m = "Το ${ dataPath } πρέπει να ισούται με μία από τις εξής επιτρεπόμενες τιμές: ${ allowedValues }.", p = "Το ${ dataPath } πρέπει να συμφωνεί με τη δομή ενός από τα εξής: ${ schemaOptions }.", P = "Δεν υποστηρίζονται γραφήματα διασποράς με αναλογικά σύμβολα. Έχει εφαρμοστεί το προεπιλεγμένο μέγεθος συμβόλου.", h = "Αποτυχία ανάγνωσης δεδομένων εισόδου", y = "Η λειτουργία queryBins δεν υποστηρίζεται σε αυτήν την υπηρεσία.", g = "Ο αναμενόμενος τύπος σειράς στο ευρετήριο ${ seriesIndex } είναι '${ expectedType }', αλλά λήφθηκε '${ receivedType }' αντί γι' αυτό", I = "ή", f = {
  defaultError: t,
  negativeValueInDataForLogTransformation: a,
  negativeValueInDataForSqrtTransformation: e,
  layerLoadFailure: o,
  layerFieldsNotFound: s,
  duplicateSeriesID: n,
  nonNumericAggregation: r,
  requiredProperty: i,
  minLength: d,
  minItems: c,
  maxItems: l,
  whiteSpacePattern: $,
  additionalProperty: u,
  enumValues: m,
  anyOfValues: p,
  bubbleChartValidateMsg: P,
  queryError: h,
  queryBinsNotSupported: y,
  invalidSeriesType: g,
  or: I
};
export {
  u as additionalProperty,
  p as anyOfValues,
  P as bubbleChartValidateMsg,
  f as default,
  t as defaultError,
  n as duplicateSeriesID,
  m as enumValues,
  g as invalidSeriesType,
  s as layerFieldsNotFound,
  o as layerLoadFailure,
  l as maxItems,
  c as minItems,
  d as minLength,
  a as negativeValueInDataForLogTransformation,
  e as negativeValueInDataForSqrtTransformation,
  r as nonNumericAggregation,
  I as or,
  y as queryBinsNotSupported,
  h as queryError,
  i as requiredProperty,
  $ as whiteSpacePattern
};
