// @arcgis/charts-components/dist/chunks/messages.pl4.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const a = "Odtwórz", e = "Zatrzymaj", o = "Legenda", r = "Wczytywanie", i = "Strona główna", n = "Diagram", s = "Serie", t = "Eksportuj", w = "Obraz", z = "Dane", p = "Drukuj", y = {
  "Zoom Out": "Pomniejsz",
  Play: a,
  Stop: e,
  Legend: o,
  "Press ENTER to toggle": "Naciśnij ENTER, aby przełączyć",
  Loading: r,
  Home: i,
  Chart: n,
  "Serial chart": "Diagram szeregowy",
  "X/Y chart": "Diagram X/Y",
  "Pie chart": "Diagram kołowy",
  "Gauge chart": "Diagram wskaźnikowy",
  "Radar chart": "Diagram radarowy",
  "Sankey diagram": "Diagram Sankey'a",
  "Flow diagram": "Diagram przepływu",
  "Chord diagram": "Diagram strunowy",
  "TreeMap chart": "Diagram mapy drzewa",
  "Sliced chart": "Diagram z przekrojami",
  Series: s,
  "Candlestick Series": "Seria świecowa",
  "OHLC Series": "Seria OHLC",
  "Column Series": "Seria kolumnowa",
  "Line Series": "Seria liniowa",
  "Pie Slice Series": "Diagram kołowy z serią przekrojową",
  "Funnel Series": "Seria lejkowa",
  "Pyramid Series": "Seria piramidowa",
  "X/Y Series": "Seria X/Y",
  Map: "Mapa",
  "Press ENTER to zoom in": "Naciśnij ENTER, aby powiększyć",
  "Press ENTER to zoom out": "Naciśnij ENTER, aby pomniejszyć",
  "Use arrow keys to zoom in and out": "Użyj klawiszy strzałek do powiększania i pomniejszania",
  "Use plus and minus keys on your keyboard to zoom in and out": "Użyj klawiszy plus i minus na klawiaturze, aby powiększać i pomniejszać",
  Export: t,
  Image: w,
  Data: z,
  Print: p,
  "Press ENTER to open": "Naciśnij ENTER, aby otworzyć",
  "Press ENTER to print.": "Naciśnij ENTER, aby wydrukować.",
  "Press ENTER to export as %1.": "Naciśnij ENTER, aby wyeksportować jako %1.",
  "(Press ESC to close this message)": "(Naciśnij ESC, aby zamknąć ten komunikat)",
  "Image Export Complete": "Eksportowanie obrazu zostało zakończone",
  "Export operation took longer than expected. Something might have gone wrong.": "Operacja eksportowania trwała dłużej niż oczekiwano. Coś mogło pójść nie tak.",
  "Saved from": "Zapisano z",
  "Use TAB to select grip buttons or left and right arrows to change selection": "Użyj klawisza TAB, aby wybrać przyciski uchwytu, lub strzałek w lewo i w prawo, aby zmienić zaznaczenie",
  "Use left and right arrows to move selection": "Użyj strzałek w lewo i w prawo, aby przesunąć zaznaczenie",
  "Use left and right arrows to move left selection": "Użyj strzałek w lewo i w prawo, aby przesunąć lewe zaznaczenie",
  "Use left and right arrows to move right selection": "Użyj strzałek w lewo i w prawo, aby przesunąć prawe zaznaczenie",
  "Use TAB select grip buttons or up and down arrows to change selection": "Użyj klawisza TAB, aby wybrać przyciski uchwytu, lub strzałek w górę i w dół, aby zmienić zaznaczenie",
  "Use up and down arrows to move selection": "Użyj strzałek w górę i w dół, aby przesunąć zaznaczenie",
  "Use up and down arrows to move lower selection": "Użyj strzałek w górę i w dół, aby przesunąć dolne zaznaczenie",
  "Use up and down arrows to move upper selection": "Użyj strzałek w górę i w dół, aby przesunąć górne zaznaczenie",
  "From %1 to %2": "Od %1 do %2",
  "From %1": "Od %1",
  "To %1": "Do %1",
  "No parser available for file: %1": "Brak dostępnego analizatora składni pliku: %1",
  "Error parsing file: %1": "Błąd analizy składni pliku: %1",
  "Unable to load file: %1": "Nie można wczytywać pliku: %1",
  "Invalid date": "Nieprawidłowa data"
};
export {
  n as Chart,
  z as Data,
  t as Export,
  i as Home,
  w as Image,
  o as Legend,
  r as Loading,
  a as Play,
  p as Print,
  s as Series,
  e as Stop,
  y as default
};
