// @arcgis/charts-components/dist/chunks/messages.ro2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const a = { chart: "Diagramă", barChart: "Diagramă cu bare", lineChart: "Diagramă liniară", pieChart: "Diagramă radială", histogram: "Histogramă", scatterplot: "Diagramă de dispersie", boxPlot: "Diagramă tip casetă" }, i = { mean: "Medie", median: "Mediană", sum: "Sumă", discretePercentile: "Percentilă discretă", minimum: "Minim", maximum: "Maxim", variance: "Variaţie", count: "Număr", aggregation: "${ statistics } din ${ fieldName }", noAggregation: "Fără însumare" }, e = "Diagramă", t = "Serie sumă", r = "Axa X", s = "Indisponibil", n = { start: "Fixare la primul punct de date", end: "Fixare la ultimul punct de date" }, m = {
  chartType: a,
  statistics: i,
  defaultTitle: e,
  countSeries: t,
  xAxis: r,
  notAvailable: s,
  timeAggregationTypes: n
};
export {
  a as chartType,
  t as countSeries,
  m as default,
  e as defaultTitle,
  s as notAvailable,
  i as statistics,
  n as timeAggregationTypes,
  r as xAxis
};
