// @arcgis/charts-components/dist/chunks/messages.sr2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const a = { chart: "Grafikon", barChart: "Trakasti grafikon", lineChart: "Linijski grafikon", pieChart: "Kružni grafikon", histogram: "Histogram", scatterplot: "Grafikon", boxPlot: "Kvadratni grafikon" }, i = { mean: "Sredina", median: "Medijana", sum: "Zbir", discretePercentile: "Diskretni percentil", minimum: "Minimum", maximum: "Maksimum", variance: "Odstupanje", count: "Broj", aggregation: "${ statistics } od ${ fieldName }", noAggregation: "Nema grupisanja" }, t = "Grafikon", n = "Broj serija", e = "X-osa", r = "Nije dostupno", s = { start: "Zakači za prvu tačku podataka", end: "Zakači za zadnju tačku podataka" }, o = {
  chartType: a,
  statistics: i,
  defaultTitle: t,
  countSeries: n,
  xAxis: e,
  notAvailable: r,
  timeAggregationTypes: s
};
export {
  a as chartType,
  n as countSeries,
  o as default,
  t as defaultTitle,
  r as notAvailable,
  i as statistics,
  s as timeAggregationTypes,
  e as xAxis
};
