// @arcgis/charts-components/dist/chunks/messages.zh-HK4.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "播放", o = "停止", t = "圖例", r = "正在載入", s = "首頁", a = "圖表", n = "數列", i = "匯出", l = "影像", c = "資料", d = "列印", m = {
  "Zoom Out": "縮小",
  Play: e,
  Stop: o,
  Legend: t,
  "Press ENTER to toggle": "按下 ENTER 以切換",
  Loading: r,
  Home: s,
  Chart: a,
  "Serial chart": "序列圖",
  "X/Y chart": "X/Y 圖表",
  "Pie chart": "圓餅圖",
  "Gauge chart": "儀表圖",
  "Radar chart": "雷達圖",
  "Sankey diagram": "桑基圖",
  "Flow diagram": "流程圖",
  "Chord diagram": "和弦圖",
  "TreeMap chart": "樹狀圖圖表",
  "Sliced chart": "切片圖",
  Series: n,
  "Candlestick Series": "燭台數列",
  "OHLC Series": "OHLC 數列",
  "Column Series": "欄數列",
  "Line Series": "線數列",
  "Pie Slice Series": "圓餅圖數列",
  "Funnel Series": "漏斗數列",
  "Pyramid Series": "金字塔數列",
  "X/Y Series": "X/Y 數列",
  Map: "地圖",
  "Press ENTER to zoom in": "按下 ENTER 以放大",
  "Press ENTER to zoom out": "按下 ENTER 以縮小",
  "Use arrow keys to zoom in and out": "使用方向鍵以放大和縮小",
  "Use plus and minus keys on your keyboard to zoom in and out": "使用鍵盤上的加號和減號鍵來放大和縮小",
  Export: i,
  Image: l,
  Data: c,
  Print: d,
  "Press ENTER to open": "按下 Enter 以開啟",
  "Press ENTER to print.": "按下 Enter 以列印。",
  "Press ENTER to export as %1.": "按下 Enter 以匯出為 %1。",
  "(Press ESC to close this message)": "(按下 ESC 關閉此訊息)",
  "Image Export Complete": "圖像匯出完成",
  "Export operation took longer than expected. Something might have gone wrong.": "匯口操作耗時比預期的要長。 可能發生一些問題。",
  "Saved from": "儲存自",
  "Use TAB to select grip buttons or left and right arrows to change selection": "使用 TAB 選擇手把按鈕，或使用左右箭頭變更選擇",
  "Use left and right arrows to move selection": "使用左右箭頭移動選擇",
  "Use left and right arrows to move left selection": "使用左右箭頭移動左側選擇",
  "Use left and right arrows to move right selection": "使用左右箭頭移動右側選擇",
  "Use TAB select grip buttons or up and down arrows to change selection": "使用 TAB 選擇夾點按鈕或上下箭頭來變更選擇",
  "Use up and down arrows to move selection": "使用上下箭頭移動選擇",
  "Use up and down arrows to move lower selection": "使用上下箭頭移動較低的選擇",
  "Use up and down arrows to move upper selection": "使用上下箭頭移動較高的選擇",
  "From %1 to %2": "從 %1 到 %2",
  "From %1": "從 %1",
  "To %1": "到 %1",
  "No parser available for file: %1": "沒有可用於檔案的剖析器: %1",
  "Error parsing file: %1": "剖析檔案時出錯: %1",
  "Unable to load file: %1": "無法載入檔案: %1",
  "Invalid date": "日期無效"
};
export {
  a as Chart,
  c as Data,
  i as Export,
  s as Home,
  l as Image,
  t as Legend,
  r as Loading,
  e as Play,
  d as Print,
  n as Series,
  o as Stop,
  m as default
};
