// @arcgis/charts-components/dist/chunks/messages.pt-BR.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "Ocorreu um erro ao carregar o gráfico.", a = "Não é possível aplicar transformação de log para valores negativos ou zero.", o = "Não é possível aplicar transformação de raiz quadrada para valores negativos.", r = "Ocorreu um erro ao carregar a camada. URL = ${ url }. D de item do portal = ${ portalItemId }.", t = "Os campos [${ fieldNames }] definidos na configuração não foram encontrados na camada.", s = "${ dataPath } deve ser único. A série denominada ${ seriesName } tem um id (${ seriesID }) que já é usado por outra série.", n = "${ dataPath } não deve realizar agregação sem contagem em campo não numérico.", d = "${ dataPath } está sem uma propriedade denominada ${ missingProperty }.", i = "${ dataPath } não deve ser menor que ${ limit } caracteres.", c = "${ dataPath } não deve ter menos que ${ limit } itens.", m = "${ dataPath } não deve ter mais de ${ limit } itens.", u = "${ dataPath } deve ter pelo menos um caractere diferente de espaço em branco.", l = "${ dataPath } não deve ter ${ additionalProperty }.", p = "${ dataPath } deve ser igual a um destes valores permitidos: ${ allowedValues }.", $ = "${ dataPath } deve corresponder ao esquema de um destes: ${ schemaOptions }.", g = "Gráficos de dispersão com símbolos proporcionais não são suportados. O tamanho padrão do símbolo foi aplicado.", v = "Falha ao ler os dados de entrada.", f = "A operação queryBins não é suportada neste serviço.", h = "O tipo de série esperado no índice${ seriesIndex } é '${ expectedType }' mas '${ receivedType }' foi recebido em seu lugar", P = "ou", y = {
  defaultError: e,
  negativeValueInDataForLogTransformation: a,
  negativeValueInDataForSqrtTransformation: o,
  layerLoadFailure: r,
  layerFieldsNotFound: t,
  duplicateSeriesID: s,
  nonNumericAggregation: n,
  requiredProperty: d,
  minLength: i,
  minItems: c,
  maxItems: m,
  whiteSpacePattern: u,
  additionalProperty: l,
  enumValues: p,
  anyOfValues: $,
  bubbleChartValidateMsg: g,
  queryError: v,
  queryBinsNotSupported: f,
  invalidSeriesType: h,
  or: P
};
export {
  l as additionalProperty,
  $ as anyOfValues,
  g as bubbleChartValidateMsg,
  y as default,
  e as defaultError,
  s as duplicateSeriesID,
  p as enumValues,
  h as invalidSeriesType,
  t as layerFieldsNotFound,
  r as layerLoadFailure,
  m as maxItems,
  c as minItems,
  i as minLength,
  a as negativeValueInDataForLogTransformation,
  o as negativeValueInDataForSqrtTransformation,
  n as nonNumericAggregation,
  P as or,
  f as queryBinsNotSupported,
  v as queryError,
  d as requiredProperty,
  u as whiteSpacePattern
};
