// @arcgis/charts-components/dist/chunks/messages.he2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const t = { chart: "תרשים", barChart: "תרשים עמודות", lineChart: "תרשים קו", pieChart: "תרשים עוגה", histogram: "היסטוגרמה", scatterplot: "תרשים פיזור", boxPlot: "תרשים תיבה" }, e = { mean: "ממוצע", median: "חציון", sum: "סכום", discretePercentile: "אחוזונים בנפרד", minimum: "מינימום", maximum: "מקסימום", variance: "שונות", count: "מונה", aggregation: "${ statistics } מתוך ${ fieldName }", noAggregation: "ללא צבירה" }, a = "תרשים", s = "ספירה בסדרה", i = "ציר X", n = "לא זמין", o = { start: "הצמד לנקודת נתונים ראשונה", end: "הצמד לנקודת נתונים אחרונה" }, c = {
  chartType: t,
  statistics: e,
  defaultTitle: a,
  countSeries: s,
  xAxis: i,
  notAvailable: n,
  timeAggregationTypes: o
};
export {
  t as chartType,
  s as countSeries,
  c as default,
  a as defaultTitle,
  n as notAvailable,
  e as statistics,
  o as timeAggregationTypes,
  i as xAxis
};
