// @arcgis/charts-components/dist/chunks/messages.nl4.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "Afspelen", o = "Stoppen", r = "Legenda", t = "Bezig met laden", a = "Start", n = "Diagram", i = "Reeks", s = "Exporteren", l = "Afbeelding", d = "Gegevens", m = "Afdrukken", p = {
  "Zoom Out": "Uitzoomen",
  Play: e,
  Stop: o,
  Legend: r,
  "Press ENTER to toggle": "Druk op ENTER om in/uit te schakelen",
  Loading: t,
  Home: a,
  Chart: n,
  "Serial chart": "Seriekaart",
  "X/Y chart": "X/Y-diagram",
  "Pie chart": "Cirkeldiagram",
  "Gauge chart": "Maattabel",
  "Radar chart": "Radardiagram",
  "Sankey diagram": "Sankey-diagram",
  "Flow diagram": "Stroomdiagram",
  "Chord diagram": "Koorddiagram",
  "TreeMap chart": "Treemap-diagram",
  "Sliced chart": "Taartdiagram",
  Series: i,
  "Candlestick Series": "Kaarsreeks",
  "OHLC Series": "OHLC-reeks",
  "Column Series": "Kolommenreeks",
  "Line Series": "Lijnenreeks",
  "Pie Slice Series": "Taartsegmentreeks",
  "Funnel Series": "Trechterreeks",
  "Pyramid Series": "Pyramidereeks",
  "X/Y Series": "X/Y-reeks",
  Map: "Kaart",
  "Press ENTER to zoom in": "Druk op ENTER om in te zoomen",
  "Press ENTER to zoom out": "Druk op ENTER om uit te zoomen",
  "Use arrow keys to zoom in and out": "Gebruik pijltoetsen om in en uit te zoomen",
  "Use plus and minus keys on your keyboard to zoom in and out": "Gebruik de plus- en min-toetsen op uw toetsenbord om in en uit te zoomen",
  Export: s,
  Image: l,
  Data: d,
  Print: m,
  "Press ENTER to open": "Druk op ENTER om te openen",
  "Press ENTER to print.": "Druk op ENTER om af te drukken",
  "Press ENTER to export as %1.": "Druk op ENTER om te exporteren als %1.",
  "(Press ESC to close this message)": "(Druk op ESC als u dit bericht wilt sluiten)",
  "Image Export Complete": "Exporteren afbeelding is voltooid",
  "Export operation took longer than expected. Something might have gone wrong.": "Exporteren duurde langer dan verwacht. Er is mogelijk iets fout gegaan.",
  "Saved from": "Opgeslagen vanuit",
  "Use TAB to select grip buttons or left and right arrows to change selection": "Gebruik TAB om greepknoppen te selecteren of gebruik de pijl-links en pijl-rechts om de selectie te wijzigen",
  "Use left and right arrows to move selection": "Gebruik de pijl-links en pijl-rechts om de selectie te verplaatsen",
  "Use left and right arrows to move left selection": "Gebruik de pijl-links en pijl-rechts om de linkerselectie te verplaatsen",
  "Use left and right arrows to move right selection": "Gebruik de pijl-links en pijl-rechts om de rechterselectie te verplaatsen",
  "Use TAB select grip buttons or up and down arrows to change selection": "Gebruik TAB om greepknoppen te selecteren of gebruik de pijl-omhoog en pijl-omlaag om de selectie te wijzigen",
  "Use up and down arrows to move selection": "Gebruik de pijl-omhoog en pijl-omlaag om de selectie te wijzigen",
  "Use up and down arrows to move lower selection": "Gebruik de pijl-omhoog en pijl-omlaag om de lagere selectie te wijzigen",
  "Use up and down arrows to move upper selection": "Gebruik de pijl-omhoog en pijl-omlaag om de bovenste selectie te wijzigen",
  "From %1 to %2": "Van %1 naar %2",
  "From %1": "Van %1",
  "To %1": "Naar %1",
  "No parser available for file: %1": "Er is geen parser beschikbaar voor bestand: %1",
  "Error parsing file: %1": "Fout bij parsen bestand: %1",
  "Unable to load file: %1": "Kan bestand niet laden: %1",
  "Invalid date": "Ongeldige datum"
};
export {
  n as Chart,
  d as Data,
  s as Export,
  a as Home,
  l as Image,
  r as Legend,
  t as Loading,
  e as Play,
  m as Print,
  i as Series,
  o as Stop,
  p as default
};
