// @arcgis/charts-components/dist/chunks/messages.lt.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const a = "Įkeliant diagramą įvyko klaida.", i = "Negalima pritaikyti log. transformacijos neigiamoms arba nulinėms reikšmėms.", t = "Negalima pritaikyti kvadratinės šaknies transformacijos neigiamoms reikšmėms.", e = "Įkeliant sluoksnį įvyko klaida. URL = ${ url }. Portalo elemento ID = ${ portalItemId }.", s = "Sluoksnyje nepavyko rasti konfigūracijoje apibrėžtų laukų [${ fieldNames }].", n = "${ dataPath } turi būti unikalus. Serijos, kurios pavadinimas ${ seriesName }, ID (${ seriesID }) jau naudojamas kitai serijai.", o = "${ dataPath } negali atikti neskaičiavimo agregavimo ne skaičių lauke.", r = "${ dataPath } trūksta savybės pavadinimu ${ missingProperty }.", l = "${ dataPath } negali būti trumpesnis negu ${ limit } simb.", u = "${ dataPath } negali turėti mažiau negu ${ limit } elementų.", m = "${ dataPath } negali turėti daugiau negu ${ limit } elementų.", d = "${ dataPath } turi sudaryti bent vienas simbolis, kuris nėra tarpas.", k = "${ dataPath } negali sudaryti ${ additionalProperty }.", c = "${ dataPath } turi būti lygus vienai iš šių leidžiamų reikšmių: ${ allowedValues }.", g = "${ dataPath } turi sutapti su viena iš šių schemų: ${ schemaOptions }.", p = "Sklaidos diagramos su proporciniais simboliais nėra palaikomos. Pritaikytas numatytasis simbolių dydis.", y = "Nepavyko nuskaityti įvesties duomenų.", $ = "Šioje paslaugoje queryBins operacija nepalaikoma.", v = "Numatytas serijos tipas indeksui ${ seriesIndex } yra ${ expectedType }, bet vietoj to gauta ${ receivedType }", P = "arba", b = {
  defaultError: a,
  negativeValueInDataForLogTransformation: i,
  negativeValueInDataForSqrtTransformation: t,
  layerLoadFailure: e,
  layerFieldsNotFound: s,
  duplicateSeriesID: n,
  nonNumericAggregation: o,
  requiredProperty: r,
  minLength: l,
  minItems: u,
  maxItems: m,
  whiteSpacePattern: d,
  additionalProperty: k,
  enumValues: c,
  anyOfValues: g,
  bubbleChartValidateMsg: p,
  queryError: y,
  queryBinsNotSupported: $,
  invalidSeriesType: v,
  or: P
};
export {
  k as additionalProperty,
  g as anyOfValues,
  p as bubbleChartValidateMsg,
  b as default,
  a as defaultError,
  n as duplicateSeriesID,
  c as enumValues,
  v as invalidSeriesType,
  s as layerFieldsNotFound,
  e as layerLoadFailure,
  m as maxItems,
  u as minItems,
  l as minLength,
  i as negativeValueInDataForLogTransformation,
  t as negativeValueInDataForSqrtTransformation,
  o as nonNumericAggregation,
  P as or,
  $ as queryBinsNotSupported,
  y as queryError,
  r as requiredProperty,
  d as whiteSpacePattern
};
