// @arcgis/charts-components/dist/chunks/messages.fi4.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const a = "Toista", e = "Lopeta", o = "Selite", i = "Ladataan", t = "Aloitussivu", l = "Kaavio", n = "Sarja", s = "Vie", r = "Kuva", m = "Aineisto", p = "Tulosta", u = {
  "Zoom Out": "Loitonna",
  Play: a,
  Stop: e,
  Legend: o,
  "Press ENTER to toggle": "Vaihda painamalla ENTER",
  Loading: i,
  Home: t,
  Chart: l,
  "Serial chart": "Sarjakaavio",
  "X/Y chart": "X/Y-kaavio",
  "Pie chart": "Ympyräkaavio",
  "Gauge chart": "Mittarikaavio",
  "Radar chart": "Tutkakaavio",
  "Sankey diagram": "Sankey-kaavio",
  "Flow diagram": "Virtakaavio",
  "Chord diagram": "Puoliympyräkaavio",
  "TreeMap chart": "Puukaavio",
  "Sliced chart": "Viipalekaavio",
  Series: n,
  "Candlestick Series": "Kynttiläsarja",
  "OHLC Series": "OHLC-sarja",
  "Column Series": "Sarakesarja",
  "Line Series": "Viivasarja",
  "Pie Slice Series": "Viipalesarja",
  "Funnel Series": "Suppilosarja",
  "Pyramid Series": "Pyramidisarja",
  "X/Y Series": "X/Y-sarja",
  Map: "Kartta",
  "Press ENTER to zoom in": "Lähennä painamalla ENTER",
  "Press ENTER to zoom out": "Loitonna painamalla ENTER",
  "Use arrow keys to zoom in and out": "Lähennä ja loitonna nuolinäppäimillä",
  "Use plus and minus keys on your keyboard to zoom in and out": "Lähennä ja loitonna näppäimistön plus- ja miinuspainikkeilla",
  Export: s,
  Image: r,
  Data: m,
  Print: p,
  "Press ENTER to open": "Avaa painamalla ENTER",
  "Press ENTER to print.": "Tulosta painamalla ENTER.",
  "Press ENTER to export as %1.": "Vie muodossa %1 painamalla ENTER.",
  "(Press ESC to close this message)": "(Sulje tämä viesti painamalla ESC)",
  "Image Export Complete": "Kuvan vienti valmis",
  "Export operation took longer than expected. Something might have gone wrong.": "Vientitoiminto kesti odotettua kauemmin. Jossain tapahtui ehkä virhe.",
  "Saved from": "Tallennettu kohteesta",
  "Use TAB to select grip buttons or left and right arrows to change selection": "Valitse kahvapainikkeet painamalla TAB tai muuta valintaa vasemmalla ja oikealla nuolella",
  "Use left and right arrows to move selection": "Siirrä valintaa vasemmalla ja oikealla nuolella",
  "Use left and right arrows to move left selection": "Siirrä valintaa vasemmalle oikealla ja vasemmalla nuolella",
  "Use left and right arrows to move right selection": "Siirrä valintaa oikealle vasemmalla ja oikealla nuolella",
  "Use TAB select grip buttons or up and down arrows to change selection": "Valitse kahvapainikkeet painamalla TAB tai muuta valintaa ylä- ja alanuolella",
  "Use up and down arrows to move selection": "Siirrä valintaa ylä- ja alanuolella",
  "Use up and down arrows to move lower selection": "Siirrä alempaa valintaa ylä- ja alanuolella",
  "Use up and down arrows to move upper selection": "Siirrä ylempää valintaa ylä- ja alanuolella",
  "From %1 to %2": "Kohteesta %1 kohteeseen %2",
  "From %1": "Kohteesta %1",
  "To %1": "Kohteeseen %1",
  "No parser available for file: %1": "Jäsennintä ei ole saatavilla tiedostolle: %1",
  "Error parsing file: %1": "Virhe jäsennettäessä tiedostoa: %1",
  "Unable to load file: %1": "Ei voi ladata tiedostoa: %1",
  "Invalid date": "Virheellinen päivämäärä"
};
export {
  l as Chart,
  m as Data,
  s as Export,
  t as Home,
  r as Image,
  o as Legend,
  i as Loading,
  a as Play,
  p as Print,
  n as Series,
  e as Stop,
  u as default
};
