// @arcgis/charts-components/dist/chunks/messages.zh-CN.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const t = "加载图表时出错。", a = "无法向负值或零值应用对数变换。", e = "无法向负值应用平方根变换。", s = "加载图层时出错。 URL = ${ url }。 门户项目 ID = ${ portalItemId }。", o = "未在图层中找到配置中定义的字段 [${ fieldNames }]。", n = "${ dataPath } 必须唯一。 名为 ${ seriesName } 的系列具有一个 ID (${ seriesID })，此 ID 已被另一个系列使用。", r = "${ dataPath } 无法在非数值字段上执行非计数聚合。", i = "${ dataPath } 缺少名为 ${ missingProperty } 的属性。", d = "${ dataPath } 不应短于 ${ limit } 个字符。", c = "${ dataPath } 包含的项目数量不应少于 ${ limit } 个。", l = "${ dataPath } 包含的项目数量不应多于 ${ limit } 个。", $ = "${ dataPath } 必须至少包含一个空白字符。", u = "${ dataPath } 不能有 ${ additionalProperty }。", m = "${ dataPath } 必须等于以下允许值之一：${ allowedValues }。", h = "${ dataPath } 必须与下列各项之一的方案匹配：${ schemaOptions }。", p = "包含比例符号的散点图不受支持。 已应用默认符号大小。", P = "无法读取输入数据。", y = "此服务不支持 queryBins 操作。", I = "索引 ${ seriesIndex } 处的预期系列类型是 '${ expectedType }'，但接受的是 '${ receivedType }'", g = "或", D = {
  defaultError: t,
  negativeValueInDataForLogTransformation: a,
  negativeValueInDataForSqrtTransformation: e,
  layerLoadFailure: s,
  layerFieldsNotFound: o,
  duplicateSeriesID: n,
  nonNumericAggregation: r,
  requiredProperty: i,
  minLength: d,
  minItems: c,
  maxItems: l,
  whiteSpacePattern: $,
  additionalProperty: u,
  enumValues: m,
  anyOfValues: h,
  bubbleChartValidateMsg: p,
  queryError: P,
  queryBinsNotSupported: y,
  invalidSeriesType: I,
  or: g
};
export {
  u as additionalProperty,
  h as anyOfValues,
  p as bubbleChartValidateMsg,
  D as default,
  t as defaultError,
  n as duplicateSeriesID,
  m as enumValues,
  I as invalidSeriesType,
  o as layerFieldsNotFound,
  s as layerLoadFailure,
  l as maxItems,
  c as minItems,
  d as minLength,
  a as negativeValueInDataForLogTransformation,
  e as negativeValueInDataForSqrtTransformation,
  r as nonNumericAggregation,
  g as or,
  y as queryBinsNotSupported,
  P as queryError,
  i as requiredProperty,
  $ as whiteSpacePattern
};
