// @arcgis/charts-components/dist/chunks/messages.uk.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const t = "Під час завантаження діаграми виникла помилка.", a = "Неможливо застосувати перетворення логарифму до від’ємних або нульових значень.", e = "Неможливо застосувати перетворення квадратного кореня до від’ємних значень.", s = "Під час завантаження шару виникла помилка. URL = ${ url }. ІД елемента порталу = ${ portalItemId }.", o = "Визначені у конфігурації поля [${ fieldNames }] не знайдено на шарі.", n = "${ dataPath } повинен бути унікальним. Серія з назвою ${ seriesName } має ідентифікатор (${ seriesID }), який вже використовується іншою серією.", r = "${ dataPath } не може виконувати агрегування без підрахунку в нечисловому полі.", i = "${ dataPath } не має властивості з назвою ${ missingProperty }.", d = "${ dataPath } не може мати менше ${ limit } символів.", c = "${ dataPath } не може мати менше ${ limit } елементів.", l = "${ dataPath } не може мати більше ${ limit } елементів.", $ = "${ dataPath } повинен мати принаймні один символ, який не є пробілом.", u = "${ dataPath } не може мати ${ additionalProperty }.", m = "${ dataPath } має дорівнювати одному з цих допустимих значень: ${ allowedValues }.", p = "${ dataPath } має збігатися зі схемою одного з цих: ${ schemaOptions }.", P = "Діаграми розсіювання з пропорційними символами не підтримуються. Застосовано розмір символу за замовчуванням.", h = "Помилка читання вхідних даних.", y = "Операція queryBins не підтримується на цьому сервісі.", g = "Очікуваним типом серії з індексом ${ seriesIndex } є '${ expectedType }', але замість нього було отримано '${ receivedType }'", I = "або", f = {
  defaultError: t,
  negativeValueInDataForLogTransformation: a,
  negativeValueInDataForSqrtTransformation: e,
  layerLoadFailure: s,
  layerFieldsNotFound: o,
  duplicateSeriesID: n,
  nonNumericAggregation: r,
  requiredProperty: i,
  minLength: d,
  minItems: c,
  maxItems: l,
  whiteSpacePattern: $,
  additionalProperty: u,
  enumValues: m,
  anyOfValues: p,
  bubbleChartValidateMsg: P,
  queryError: h,
  queryBinsNotSupported: y,
  invalidSeriesType: g,
  or: I
};
export {
  u as additionalProperty,
  p as anyOfValues,
  P as bubbleChartValidateMsg,
  f as default,
  t as defaultError,
  n as duplicateSeriesID,
  m as enumValues,
  g as invalidSeriesType,
  o as layerFieldsNotFound,
  s as layerLoadFailure,
  l as maxItems,
  c as minItems,
  d as minLength,
  a as negativeValueInDataForLogTransformation,
  e as negativeValueInDataForSqrtTransformation,
  r as nonNumericAggregation,
  I as or,
  y as queryBinsNotSupported,
  h as queryError,
  i as requiredProperty,
  $ as whiteSpacePattern
};
