// @arcgis/charts-components/dist/chunks/messages.ja.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const t = "チャートの読み込み中にエラーが発生しました。", a = "対数変換は負またはゼロの値に適用できません。", e = "平方根変換は負の値に適用できません。", s = "レイヤーの読み込み中にエラーが発生しました。 URL = ${ url } ポータル アイテム ID = ${ portalItemId }", o = "構成に定義されたフィールド [${ fieldNames }] がレイヤー内に見つかりませんでした。", n = "${ dataPath } は一意である必要があります。 ${ seriesName } という系列の ID (${ seriesID }) は、別の系列ですでに使用されています。", r = "${ dataPath } は、数値以外のフィールドで、個数以外の集約を実行することはできません。", i = "${ dataPath } に ${ missingProperty } というプロパティがありません。", d = "${ dataPath } は ${ limit } 文字以上である必要があります。", c = "${ dataPath } のアイテムは ${ limit } 以上にする必要があります。", l = "${ dataPath } のアイテムは ${ limit } 以下にする必要があります。", $ = "${ dataPath } には空白以外の文字が 1 つ以上含まれている必要があります。", u = "${ dataPath } が ${ additionalProperty } を持つことはできません。", m = "${ dataPath } は、次のいずれかの許可された値と等しい必要があります: ${ allowedValues }", p = "${ dataPath } は、次のいずれかのスキーマと一致する必要があります: ${ schemaOptions }", P = "比例シンボルを使用した散布図はサポートされていません。 デフォルトのシンボル サイズが適用されています。", h = "入力データを読み取れません。", y = "queryBins 操作は、このサービスでサポートされていません", I = "インデックス ${ seriesIndex } で必要な系列タイプは ${ expectedType } ですが、代わりに ${ receivedType } が指定されました", g = "または", f = {
  defaultError: t,
  negativeValueInDataForLogTransformation: a,
  negativeValueInDataForSqrtTransformation: e,
  layerLoadFailure: s,
  layerFieldsNotFound: o,
  duplicateSeriesID: n,
  nonNumericAggregation: r,
  requiredProperty: i,
  minLength: d,
  minItems: c,
  maxItems: l,
  whiteSpacePattern: $,
  additionalProperty: u,
  enumValues: m,
  anyOfValues: p,
  bubbleChartValidateMsg: P,
  queryError: h,
  queryBinsNotSupported: y,
  invalidSeriesType: I,
  or: g
};
export {
  u as additionalProperty,
  p as anyOfValues,
  P as bubbleChartValidateMsg,
  f as default,
  t as defaultError,
  n as duplicateSeriesID,
  m as enumValues,
  I as invalidSeriesType,
  o as layerFieldsNotFound,
  s as layerLoadFailure,
  l as maxItems,
  c as minItems,
  d as minLength,
  a as negativeValueInDataForLogTransformation,
  e as negativeValueInDataForSqrtTransformation,
  r as nonNumericAggregation,
  g as or,
  y as queryBinsNotSupported,
  h as queryError,
  i as requiredProperty,
  $ as whiteSpacePattern
};
