// @arcgis/charts-components/dist/chunks/messages.fi.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const t = "Kaavion lataamisessa tapahtui virhe.", a = "Logaritmimuunnosta ei voi käyttää negatiivisiin arvoihin tai nolla-arvoihin.", e = "Neliöjuurimuunnosta ei voi käyttää negatiivisiin arvoihin.", i = "Karttatason lataamisessa tapahtui virhe. URL = ${ url }. Portaalikohteen tunnus = ${ portalItemId }.", s = "Määrityksessä määritettyjä kenttiä [${ fieldNames }] ei löytynyt karttatasolta.", n = "Kohteen ${ dataPath } on oltava yksilöllinen. Sarjan ${ seriesName } tunnus (${ seriesID }) on jo toisen sarjan käytössä.", o = "${ dataPath } ei saa suorittaa muuta kuin määrän koostamista muussa kuin numerokentässä.", r = "Kohteesta ${ dataPath } puuttuu ominaisuus ${ missingProperty }.", u = "${ dataPath } ei saa olla alle ${ limit } merkkiä pitkä.", l = "Kohteessa ${ dataPath } ei saa olla alle ${ limit } kohdetta.", m = "Kohteessa ${ dataPath } ei saa olla yli ${ limit } kohdetta.", h = "Kohteessa ${ dataPath } on oltava vähintään yksi muu kuin välimerkki.", d = "Kohteessa ${ dataPath } ei saa olla kohdetta ${ additionalProperty }.", k = "Kohteen ${ dataPath } tulee olla yhtä suuri kuin jokin näistä sallituista arvoista: ${ allowedValues }.", y = "Kohteen ${ dataPath } on vastattava jotakin näistä skeemoista: ${ schemaOptions }.", c = "Pistekaavioita, joissa on suhteutettuja symboleja, ei tueta. Symbolin oletuskokoa on käytetty.", $ = "Lähtöaineiston lukeminen epäonnistui.", v = "Tämä palvelu ei tue queryBins-toimintoa.", p = 'Odotettu sarjan tyyppi indeksissä ${ seriesIndex } on "${ expectedType }", mutta sen sijaan vastaanotettiin "${ receivedType }"', P = "tai", g = {
  defaultError: t,
  negativeValueInDataForLogTransformation: a,
  negativeValueInDataForSqrtTransformation: e,
  layerLoadFailure: i,
  layerFieldsNotFound: s,
  duplicateSeriesID: n,
  nonNumericAggregation: o,
  requiredProperty: r,
  minLength: u,
  minItems: l,
  maxItems: m,
  whiteSpacePattern: h,
  additionalProperty: d,
  enumValues: k,
  anyOfValues: y,
  bubbleChartValidateMsg: c,
  queryError: $,
  queryBinsNotSupported: v,
  invalidSeriesType: p,
  or: P
};
export {
  d as additionalProperty,
  y as anyOfValues,
  c as bubbleChartValidateMsg,
  g as default,
  t as defaultError,
  n as duplicateSeriesID,
  k as enumValues,
  p as invalidSeriesType,
  s as layerFieldsNotFound,
  i as layerLoadFailure,
  m as maxItems,
  l as minItems,
  u as minLength,
  a as negativeValueInDataForLogTransformation,
  e as negativeValueInDataForSqrtTransformation,
  o as nonNumericAggregation,
  P as or,
  v as queryBinsNotSupported,
  $ as queryError,
  r as requiredProperty,
  h as whiteSpacePattern
};
