// @arcgis/charts-components/dist/chunks/messages.ar.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const t = "حدث خطأ أثناء تحميل المخطط.", a = "يتعذر تطبيق تحويل السِّجل إلى قيم سلبية أو إلى صفر.", e = "يتعذر تطبيق تحويل الجذر التربيعي إلى قيم سلبية.", s = "حدث خطأ أثناء تحميل الطبقة. URL = ${ url }. معرف عنصر البوابة = ${ portalItemId }.", o = "لم يُعثر في الطبقة على الحقول [${ fieldNames }] المحددة في التكوين.", n = "يجب أن يكون ${ dataPath } مميزًا. السلسلة التي اسمها ${ seriesName } تتضمن معرفًا (${ seriesID }) مستخدمًا بالفعل من قبل سلسلة أخرى.", r = "يتعذر على ${ dataPath } إجراء تجميع من دون حساب في حقل غير رقمي.", i = "يتفقد ${ dataPath } خاصية باسم ${ missingProperty }.", d = "يجب ألا يكون ${ dataPath } أقصر من ${ limit } من الحروف.", c = "يجب ألا يحتوي ${ dataPath } على أقل من ${ limit } من العناصر.", l = "يجب ألا يحتوي ${ dataPath } على أكثر من ${ limit } من العناصر.", $ = "يجب أن يتضمن ${ dataPath } حرفًا واحدًا على الأقل ليس مسافة بيضاء.", u = "يجب أن يحتوي ${ dataPath } على ${ additionalProperty }.", m = "يجب أن يكون ${ dataPath } مساويًا لإحدى القيم هذه المسموح بها: ${ allowedValues }", p = "يجب أن يتطابق ${ dataPath } مع مخطط أحد ما يلي: ${ schemaOptions }.", P = "Scatterplots ذات الرموز المتناسبة غير مدعومة. تم تطبيق حجم الرمز الافتراضي.", h = "فشلت قراءة البيانات المدخلة.", y = "لا يتم دعم عملية queryBins على هذه الخدمة.", g = "نوع السلسلة المتوقع في الفهرس ${ seriesIndex } هو '${ expectedType }' ولكن تم استلام '${ receivedType }' بدلاً من ذلك", I = "أو", f = {
  defaultError: t,
  negativeValueInDataForLogTransformation: a,
  negativeValueInDataForSqrtTransformation: e,
  layerLoadFailure: s,
  layerFieldsNotFound: o,
  duplicateSeriesID: n,
  nonNumericAggregation: r,
  requiredProperty: i,
  minLength: d,
  minItems: c,
  maxItems: l,
  whiteSpacePattern: $,
  additionalProperty: u,
  enumValues: m,
  anyOfValues: p,
  bubbleChartValidateMsg: P,
  queryError: h,
  queryBinsNotSupported: y,
  invalidSeriesType: g,
  or: I
};
export {
  u as additionalProperty,
  p as anyOfValues,
  P as bubbleChartValidateMsg,
  f as default,
  t as defaultError,
  n as duplicateSeriesID,
  m as enumValues,
  g as invalidSeriesType,
  o as layerFieldsNotFound,
  s as layerLoadFailure,
  l as maxItems,
  c as minItems,
  d as minLength,
  a as negativeValueInDataForLogTransformation,
  e as negativeValueInDataForSqrtTransformation,
  r as nonNumericAggregation,
  I as or,
  y as queryBinsNotSupported,
  h as queryError,
  i as requiredProperty,
  $ as whiteSpacePattern
};
