// @arcgis/charts-components/dist/chunks/messages.he4.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "הפעל", o = "עצור", t = "מקרא", r = "טוען", s = "דף הבית", a = "תרשים", n = "סדרה", i = "יצוא", l = "תמונה", c = "נתונים", d = "הדפס", E = {
  "Zoom Out": "זום החוצה",
  Play: e,
  Stop: o,
  Legend: t,
  "Press ENTER to toggle": "לחץ על ENTER להחלפת מצב",
  Loading: r,
  Home: s,
  Chart: a,
  "Serial chart": "תרשים סדרה",
  "X/Y chart": "תרשים X/Y",
  "Pie chart": "תרשים עוגה",
  "Gauge chart": "תרשים מד",
  "Radar chart": "תרשים רדאר",
  "Sankey diagram": "תרשים Sankey",
  "Flow diagram": "תרשים זרימה",
  "Chord diagram": "תרשים מיתרים",
  "TreeMap chart": "תרשים מפת עץ",
  "Sliced chart": "תרשים פלחים",
  Series: n,
  "Candlestick Series": "סדרת פמוטים",
  "OHLC Series": "סדרת OHLC",
  "Column Series": "סדרת עמודות",
  "Line Series": "סדרת קווים",
  "Pie Slice Series": "סדרת פרוסות עוגה",
  "Funnel Series": "סדרת משפך",
  "Pyramid Series": "סדרת פירמידה",
  "X/Y Series": "סדרת X/Y",
  Map: "מפה",
  "Press ENTER to zoom in": "לחץ על ENTER לזום כניסה",
  "Press ENTER to zoom out": "לחץ על ENTER לזום יציאה",
  "Use arrow keys to zoom in and out": "השתמש במקשי החיצים כדי לבצע זום פנימה והחוצה",
  "Use plus and minus keys on your keyboard to zoom in and out": "השתמש במקשי הפלוס והמינוס במקלדת כדי לבצע זום פנימה והחוצה",
  Export: i,
  Image: l,
  Data: c,
  Print: d,
  "Press ENTER to open": "לחץ על ENTER לפתיחה",
  "Press ENTER to print.": "לחץ על ENTER להדפסה.",
  "Press ENTER to export as %1.": "לחץ על ENTER כדי לייצא כ-%1.",
  "(Press ESC to close this message)": "(לחץ על ESC כדי לסגור הודעה זו)",
  "Image Export Complete": "יצוא התמונה הושלם",
  "Export operation took longer than expected. Something might have gone wrong.": "פעולת הייצוא ארכה זמן רב מהצפוי. ייתכן שמשמהו השתבש.",
  "Saved from": "נשמר מ:",
  "Use TAB to select grip buttons or left and right arrows to change selection": "השתמש ב-TAB כדי לבחור לחצני ידית או חצים שמאלה וימינה לשינוי בחירה",
  "Use left and right arrows to move selection": "השתמש בחצים שמאלה וימינה להזזת בחירה",
  "Use left and right arrows to move left selection": "השתמש בחצים שמאלה וימינה להזזת בחירה שמאלית",
  "Use left and right arrows to move right selection": "השתמש בחצים שמאלה וימינה להזזת בחירה ימנית",
  "Use TAB select grip buttons or up and down arrows to change selection": "השתמש ב-TAB כדי לבחור לחצני ידית או חצים למעלה ולמטה לשינוי בחירה",
  "Use up and down arrows to move selection": "השתמש בחצים למעלה ולמטה להזזת בחירה",
  "Use up and down arrows to move lower selection": "השתמש בחצים למעלה ולמטה להזזת בחירה תחתונה",
  "Use up and down arrows to move upper selection": "השתמש בחצים למעלה ולמטה להזזת בחירה עליונה",
  "From %1 to %2": "מ: %1 אל %2",
  "From %1": "מ: %1",
  "To %1": "אל %1",
  "No parser available for file: %1": "אין parser זמין עבור קובץ: %1",
  "Error parsing file: %1": "שגיאה בפענוח קובץ: %1",
  "Unable to load file: %1": "לא ניטן לטעון קובץ: %1",
  "Invalid date": "תאריך לא חוקי"
};
export {
  a as Chart,
  c as Data,
  i as Export,
  s as Home,
  l as Image,
  t as Legend,
  r as Loading,
  e as Play,
  d as Print,
  n as Series,
  o as Stop,
  E as default
};
