// @arcgis/charts-components/dist/chunks/messages.he.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const t = "אירעה שגיאה במהלך טעינת התרשים.", a = "לא ניתן לבצע טרנספורמציה לוגריתמית על ערכים שליליים או על אפס.", e = "לא ניתן לבצע טרנספורמציה של שורש ריבועי על ערכים שליליים.", o = "אירעה שגיאה במהלך טעינת השכבה. URL = ${ url }. מזהה פריט פורטל =${ portalItemId }.", s = "השדות [${ fieldNames }] שהוגדרו ב-config לא נמצאו בשכבה.", n = "${ dataPath } חייב/ת להיות ייחודי/ת. לסדרה בשם ${ seriesName } יש מזהה (${ seriesID }) שבו משתמשת כבר סדרה אחרת.", r = "ל-${ dataPath } אסור לבצע צבירה שאינה ספירה על שדה לא מספרי.", i = "ב-${ dataPath } חסר מאפיין בשם ${ missingProperty }.", d = "${ dataPath } לא יכול להיות קצר יות מ-${ limit } תווים.", c = "ל-${ dataPath } לא יכולים להיות פחות מ-${ limit } פריטים.", l = "ל-${ dataPath } לא יכולים להיות יותר מ-${ limit } פריטים.", $ = "ל-${ dataPath } חייב להיות לפחות תו אחד שאינו רווח בלתי נראה.", u = "ל-${ dataPath } אסור לכלול ${ additionalProperty }.", m = "${ dataPath } חייב להיות שווה לאחד מהערכים המותרים הבאים: ${ allowedValues }.", h = "${ dataPath } חייב להתאים לסכמה של אחד מאלה: ${ schemaOptions }.", p = "תרשימי פיזור עם סמלים פרופורציונליים לא נתמכים. הוחל גודל סמל שמוגדר כברירת מחדל.", P = "קריאת נתוני הקלט נכשלה.", y = "פעולת queryBins אינה נתמכת על ידי שירות זה.", g = "סוג הסדרה הצפוי במדד ${ seriesIndex } הוא '${ expectedType }', אבל במקום זאת התקבל '${ receivedType }'.", I = "או", f = {
  defaultError: t,
  negativeValueInDataForLogTransformation: a,
  negativeValueInDataForSqrtTransformation: e,
  layerLoadFailure: o,
  layerFieldsNotFound: s,
  duplicateSeriesID: n,
  nonNumericAggregation: r,
  requiredProperty: i,
  minLength: d,
  minItems: c,
  maxItems: l,
  whiteSpacePattern: $,
  additionalProperty: u,
  enumValues: m,
  anyOfValues: h,
  bubbleChartValidateMsg: p,
  queryError: P,
  queryBinsNotSupported: y,
  invalidSeriesType: g,
  or: I
};
export {
  u as additionalProperty,
  h as anyOfValues,
  p as bubbleChartValidateMsg,
  f as default,
  t as defaultError,
  n as duplicateSeriesID,
  m as enumValues,
  g as invalidSeriesType,
  s as layerFieldsNotFound,
  o as layerLoadFailure,
  l as maxItems,
  c as minItems,
  d as minLength,
  a as negativeValueInDataForLogTransformation,
  e as negativeValueInDataForSqrtTransformation,
  r as nonNumericAggregation,
  I as or,
  y as queryBinsNotSupported,
  P as queryError,
  i as requiredProperty,
  $ as whiteSpacePattern
};
