// @arcgis/charts-components/dist/chunks/messages.th.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const t = "เกิดข้อผิดพลาดขึ้นขณะโหลดแผนภูมิ", a = "ไม่สามารถใช้การแปลงค่า log กับค่าติดลบหรือศูนย์ได้", e = "ไม่สามารถใช้การแปลงรากที่สองกับค่าติดลบได้", o = "เกิดข้อผิดพลาดขึ้นขณะโหลดชั้นข้อมูล URL = ${ url } ID รายการพอร์ทัล = ${ portalItemId }", s = "ไม่พบฟิลด์ [${ fieldNames }] ที่กำหนดค่าไว้ในคอนฟิกในชั้นข้อมูล", n = "${ dataPath } ต้องไม่ซ้ำกัน ซีรีส์ที่ชื่อ ${ seriesName } มี ID (${ seriesID }) ที่ซีรีส์อื่นใช้งานอยู่แล้ว", r = "${ dataPath } ไม่สามารถดำเนินการรวมแบบไม่นับกับฟิลด์ที่ไม่ใช่ตัวเลขได้", i = "${ dataPath } ไม่มีชื่อคุณสมบัติ ${ missingProperty }", d = "${ dataPath } ต้องมีความยาวมากกว่า ${ limit } อักขระ", c = "${ dataPath } ต้องมีไม่น้อยกว่า ${ limit } รายการ", l = "${ dataPath } ต้องมีไม่เกิน ${ limit } รายการ", $ = "${ dataPath } ต้องมีอย่างน้อยหนึ่งอักขระที่ไม่ใช่ไวท์สเปซ", u = "${ dataPath } ต้องไม่มี ${ additionalProperty }", m = "${ dataPath } ต้องเท่ากับหนึ่งในค่าที่อนุญาตเหล่านี้: ${ allowedValues }", h = "${ dataPath } ต้องตรงกับรูปแบบของหนึ่งในรายการเหล่านี้: ${ schemaOptions }", p = "ไม่รองรับพล็อตการกระจายที่มีสัญลักษณ์ตามสัดส่วน ใช้ขนาดสัญลักษณ์เริ่มต้นแล้ว", P = "ไม่สามารถอ่านข้อมูลขาเข้า", y = "บริการนี้ไม่รองรับการดำเนินการ queryBins", g = "ประเภทซีรีส์ที่คาดหวังที่ดัชนี ${ seriesIndex } คือ '${ expectedType }' แต่ได้รับ '${ receivedType }' แทน", I = "หรือ", f = {
  defaultError: t,
  negativeValueInDataForLogTransformation: a,
  negativeValueInDataForSqrtTransformation: e,
  layerLoadFailure: o,
  layerFieldsNotFound: s,
  duplicateSeriesID: n,
  nonNumericAggregation: r,
  requiredProperty: i,
  minLength: d,
  minItems: c,
  maxItems: l,
  whiteSpacePattern: $,
  additionalProperty: u,
  enumValues: m,
  anyOfValues: h,
  bubbleChartValidateMsg: p,
  queryError: P,
  queryBinsNotSupported: y,
  invalidSeriesType: g,
  or: I
};
export {
  u as additionalProperty,
  h as anyOfValues,
  p as bubbleChartValidateMsg,
  f as default,
  t as defaultError,
  n as duplicateSeriesID,
  m as enumValues,
  g as invalidSeriesType,
  s as layerFieldsNotFound,
  o as layerLoadFailure,
  l as maxItems,
  c as minItems,
  d as minLength,
  a as negativeValueInDataForLogTransformation,
  e as negativeValueInDataForSqrtTransformation,
  r as nonNumericAggregation,
  I as or,
  y as queryBinsNotSupported,
  P as queryError,
  i as requiredProperty,
  $ as whiteSpacePattern
};
