// @arcgis/charts-components/dist/chunks/messages.et4.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const a = "Esita", e = "Peata", s = "Legend", i = "Laadimine", t = "Avakuva", o = "Diagramm", r = "Sarjad", l = "Ekspordi", m = "Pilt", n = "Andmed", u = "Prindi", d = {
  "Zoom Out": "Vähenda",
  Play: a,
  Stop: e,
  Legend: s,
  "Press ENTER to toggle": "Vahetamiseks vajutage nuppu SISESTA",
  Loading: i,
  Home: t,
  Chart: o,
  "Serial chart": "Seeriadiagramm",
  "X/Y chart": "X/Y-diagramm",
  "Pie chart": "Sektordiagramm",
  "Gauge chart": "Mõõdikdiagramm",
  "Radar chart": "Radardiagramm",
  "Sankey diagram": "Sankey diagramm",
  "Flow diagram": "Voodiagramm",
  "Chord diagram": "Kõõldiagramm",
  "TreeMap chart": "Puukaardi diagramm",
  "Sliced chart": "Viildiagramm",
  Series: r,
  "Candlestick Series": "Küünalgraafikute sari",
  "OHLC Series": "OHLC sari",
  "Column Series": "Tulpdiagrammide sari",
  "Line Series": "Joondiagrammide sari",
  "Pie Slice Series": "Sektordiagrammide sari",
  "Funnel Series": "Lehterdiagrammide sari",
  "Pyramid Series": "Püramiiddiagrammide sari",
  "X/Y Series": "X/Y-sari",
  Map: "Kaardid",
  "Press ENTER to zoom in": "Suurendamiseks vajutage sisestusklahvi",
  "Press ENTER to zoom out": "Vähendamiseks vajutage sisestusklahvi",
  "Use arrow keys to zoom in and out": "Kasutage suumimiseks noolklahve",
  "Use plus and minus keys on your keyboard to zoom in and out": "Kasutage suumimiseks klaviatuuril olevaid pluss- ja miinusklahve.",
  Export: l,
  Image: m,
  Data: n,
  Print: u,
  "Press ENTER to open": "Avamiseks vajutage sisestusklahvi",
  "Press ENTER to print.": "Printimiseks vajutage sisestusklahvi",
  "Press ENTER to export as %1.": "Eksportimiseks kujul %1 vajutage sisestusklahvi",
  "(Press ESC to close this message)": "(Selle sõnumi sulgemiseks vajutage paoklahvi)",
  "Image Export Complete": "Pildi eksport on lõpetatud",
  "Export operation took longer than expected. Something might have gone wrong.": "Eksporditoiming võttis eeldatust rohkem aega. Midagi võis valesti minna.",
  "Saved from": "Salvestatud asukohast",
  "Use TAB to select grip buttons or left and right arrows to change selection": "Haardenuppude valimiseks kasutage tabulaatoriklahvi või valiku muutmiseks vasakut ja paremat nooleklahvi",
  "Use left and right arrows to move selection": "Valiku teisaldamiseks kasutage vasakut ja paremat nooleklahvi",
  "Use left and right arrows to move left selection": "Vasakpoolse valiku teisaldamiseks kasutage vasakut ja paremat nooleklahvi",
  "Use left and right arrows to move right selection": "Parempoolse valiku teisaldamiseks kasutage vasakut ja paremat nooleklahvi",
  "Use TAB select grip buttons or up and down arrows to change selection": "Valiku muutmiseks kasutage tabulaatoriklahvi haardenuppe või üles-allanooli",
  "Use up and down arrows to move selection": "Valiku teisaldamiseks kasutage üles-allanooli",
  "Use up and down arrows to move lower selection": "Alumise valiku teisaldamiseks kasutage üles-allanooli",
  "Use up and down arrows to move upper selection": "Ülemise valiku teisaldamiseks kasutage üles-allanooli",
  "From %1 to %2": "Alates %1 kuni %2",
  "From %1": "Alates %1",
  "To %1": "Kuni %1",
  "No parser available for file: %1": "Faili jaoks pole sõela saadaval: %1",
  "Error parsing file: %1": "Faili sõelumisel ilmnes viga: %1",
  "Unable to load file: %1": "Faili ei saa laadida: %1",
  "Invalid date": "Sobimatu kuupäev"
};
export {
  o as Chart,
  n as Data,
  l as Export,
  t as Home,
  m as Image,
  s as Legend,
  i as Loading,
  a as Play,
  u as Print,
  r as Series,
  e as Stop,
  d as default
};
