// @arcgis/charts-components/dist/chunks/messages.bs4.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "Reproduciraj", a = "Mjesto zaustavljanja", i = "Legenda", o = "Učitavanje", r = "Početna", t = "Grafikon", s = "Niz", n = "Izvezi", l = "Slika", d = "Podaci", c = "Ispis", m = {
  "Zoom Out": "Smanji",
  Play: e,
  Stop: a,
  Legend: i,
  "Press ENTER to toggle": "Pritisni ENTER za prebacivanje",
  Loading: o,
  Home: r,
  Chart: t,
  "Serial chart": "Serijski grafikon",
  "X/Y chart": "Grafikon X/Y",
  "Pie chart": "Tortni grafikon",
  "Gauge chart": "Grafikon u obliku mjerača",
  "Radar chart": "Grafikon radara",
  "Sankey diagram": "Sankey diagram",
  "Flow diagram": "Dijagram toka",
  "Chord diagram": "Dijagram tetive kružnice",
  "TreeMap chart": "Karta u obliku stabla",
  "Sliced chart": "Kružni grafikon",
  Series: s,
  "Candlestick Series": "Niz u obliku svijeća",
  "OHLC Series": "OHLC niz",
  "Column Series": "Nizovi stupaca",
  "Line Series": "Nizovi linija",
  "Pie Slice Series": "Nizovi isječaka kružnog grafikona",
  "Funnel Series": "Nizovi lijevka",
  "Pyramid Series": "Nizovi piramida",
  "X/Y Series": "Nizovi X/Y",
  Map: "Karta",
  "Press ENTER to zoom in": "Pritisni ENTER za povećavanje",
  "Press ENTER to zoom out": "Pritisni ENTER za umanjivanje",
  "Use arrow keys to zoom in and out": "Koristite tipke sa strelicama za povećanje ili umanjivanje",
  "Use plus and minus keys on your keyboard to zoom in and out": "Za zumiranje koristite tipke plus i minus na tipkovnici",
  Export: n,
  Image: l,
  Data: d,
  Print: c,
  "Press ENTER to open": "Pritisni ENTER za otvaranje",
  "Press ENTER to print.": "Pritisni ENTER za ispis.",
  "Press ENTER to export as %1.": "Pritisni ENTER za izvoz kao %1.",
  "(Press ESC to close this message)": "(Pritinit ESC za zatvaranje ove poruke)",
  "Image Export Complete": "Izvos slike dovršen",
  "Export operation took longer than expected. Something might have gone wrong.": "Radnja izvoza trajala je duže od očekivanog. Možda je nije u redu.",
  "Saved from": "Spremljeno iz",
  "Use TAB to select grip buttons or left and right arrows to change selection": "Koristi TAB za odabir gumba za hvatanje ili strelice lijevo i desno za promjenu odabira",
  "Use left and right arrows to move selection": "Koristi strelice lijevo i desno za pomicanje odabira",
  "Use left and right arrows to move left selection": "Koristi strelice lijevo i desno za pomicanje odabira ulijevo",
  "Use left and right arrows to move right selection": "Koristi strelice lijevo i desno za pomicanje odabira udesno",
  "Use TAB select grip buttons or up and down arrows to change selection": "Koristi TAB za odabir gumba za hvatanje ili strelice gore i dolje za promjenu odabira",
  "Use up and down arrows to move selection": "Koristi strelice gore i dolje za pomicanje odabira",
  "Use up and down arrows to move lower selection": "Koristi strelice gore i dolje za pomicanje odabira prema dolje",
  "Use up and down arrows to move upper selection": "Koristite strelice gore i dolje za pomicanje odabira prema gore",
  "From %1 to %2": "Od %1 do %2",
  "From %1": "Od %1",
  "To %1": "Do %1",
  "No parser available for file: %1": "Nema dostupnog parsera za datoteku: %1",
  "Error parsing file: %1": "Pogreška prilikom parsiranju datoteke: %1",
  "Unable to load file: %1": "Nije moguće učitati datoteku: %1",
  "Invalid date": "Neispravni datum"
};
export {
  t as Chart,
  d as Data,
  n as Export,
  r as Home,
  l as Image,
  i as Legend,
  o as Loading,
  e as Play,
  c as Print,
  s as Series,
  a as Stop,
  m as default
};
