// @arcgis/charts-components/dist/chunks/messages.en2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const t = { chart: "Chart", barChart: "Bar chart", lineChart: "Line chart", pieChart: "Pie chart", histogram: "Histogram", scatterplot: "Scatterplot", boxPlot: "Box plot" }, a = { mean: "Mean", median: "Median", sum: "Sum", discretePercentile: "Discrete percentile", minimum: "Minimum", maximum: "Maximum", variance: "Variance", count: "Count", aggregation: "${ statistics } of ${ fieldName }", noAggregation: "No aggregation" }, e = "Chart", i = "Count Series", n = "X Axis", o = "Not Available", s = { start: "Snap to the first data point", end: "Snap to the last data point" }, r = {
  chartType: t,
  statistics: a,
  defaultTitle: e,
  countSeries: i,
  xAxis: n,
  notAvailable: o,
  timeAggregationTypes: s
};
export {
  t as chartType,
  i as countSeries,
  r as default,
  e as defaultTitle,
  o as notAvailable,
  a as statistics,
  s as timeAggregationTypes,
  n as xAxis
};
