// @arcgis/charts-components/dist/chunks/messages.lt2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const a = { chart: "Diagrama", barChart: "Stulpelinė diagrama", lineChart: "Linijinė diagrama", pieChart: "Skritulinė diagrama", histogram: "Histograma", scatterplot: "Sklaidos diagrama", boxPlot: "Stačiakampė diagrama" }, i = { mean: "Vidurkis", median: "Mediana", sum: "Suma", discretePercentile: "Diskretinis procentilis", minimum: "Minimumas", maximum: "Maksimumas", variance: "Nuokrypis", count: "Bendras skaičius", aggregation: "${ statistics } iš ${ fieldName }", noAggregation: "Be agregavimo" }, t = "Diagrama", s = "Skaičių serija", e = "X ašis", r = "Negalima", n = { start: "Pritraukti pirmą duomenų tašką", end: "Pritraukti paskutinį duomenų tašką" }, m = {
  chartType: a,
  statistics: i,
  defaultTitle: t,
  countSeries: s,
  xAxis: e,
  notAvailable: r,
  timeAggregationTypes: n
};
export {
  a as chartType,
  s as countSeries,
  m as default,
  t as defaultTitle,
  r as notAvailable,
  i as statistics,
  n as timeAggregationTypes,
  e as xAxis
};
