// @arcgis/charts-components/dist/chunks/messages.bg4.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "Възпроизвеждане", o = "Стоп", t = "Легенда", r = "Зареждане", s = "Начало", a = "Диаграма", n = "Серии", i = "Експортиране", l = "Изображение", c = "Данни", d = "Печат", E = {
  "Zoom Out": "Отдаличаване",
  Play: e,
  Stop: o,
  Legend: t,
  "Press ENTER to toggle": "Натиснете ENTER, за да превключите",
  Loading: r,
  Home: s,
  Chart: a,
  "Serial chart": "Серийна диаграма",
  "X/Y chart": "X/Y диаграма",
  "Pie chart": "Диаграма в кръгла форма",
  "Gauge chart": "Диаграма с контролна стрелка",
  "Radar chart": "Диаграма на радар",
  "Sankey diagram": "Диаграма Sankey",
  "Flow diagram": "Диаграма на потока",
  "Chord diagram": "Диаграма на хорда",
  "TreeMap chart": "Диаграма тип TreeMap",
  "Sliced chart": "Срезова диаграма",
  Series: n,
  "Candlestick Series": "Серия тип „свещ“",
  "OHLC Series": "Серия тип OHLC",
  "Column Series": "Серия тип „колона“",
  "Line Series": "Линейна серия",
  "Pie Slice Series": "Серия тип „кръгова диаграма“",
  "Funnel Series": 'Серия тип „фуния"',
  "Pyramid Series": 'Серия тип „пирамида"',
  "X/Y Series": "Серия X/Y",
  Map: "Карта",
  "Press ENTER to zoom in": "Натиснете ENTER за увеличаване",
  "Press ENTER to zoom out": "Натиснете ENTER за намаляване",
  "Use arrow keys to zoom in and out": "Използвайте клавишите със стрелки за увеличаване и намаляване",
  "Use plus and minus keys on your keyboard to zoom in and out": "Използвайте клавишите плюс и минус на клавиатурата за увеличаване и намаляване",
  Export: i,
  Image: l,
  Data: c,
  Print: d,
  "Press ENTER to open": "Натиснете ENTER за отваряне",
  "Press ENTER to print.": "Натиснете ENTER за отпечатване",
  "Press ENTER to export as %1.": "Натиснете ENTER за експортиране като %1.",
  "(Press ESC to close this message)": "(Натиснете ESC, за да затворите това съобщение)",
  "Image Export Complete": "Експортирането на изображението е завършено",
  "Export operation took longer than expected. Something might have gone wrong.": "Операцията по експортиране отне повече време от очакваното. Възможно е нещо да се е объркало.",
  "Saved from": "Запазено от",
  "Use TAB to select grip buttons or left and right arrows to change selection": "Използвайте TAB, за да изберете бутони за захващане или стрелки наляво и надясно, за да промените селекцията",
  "Use left and right arrows to move selection": "Използвайте стрелки наляво и надясно, за да преместите селекцията",
  "Use left and right arrows to move left selection": "Използвайте стрелки наляво и надясно, за да преместите селекцията наляво",
  "Use left and right arrows to move right selection": "Използвайте стрелки наляво и надясно, за да преместите селекцията надясно",
  "Use TAB select grip buttons or up and down arrows to change selection": "Използвайте бутони за захващане TAB или стрелки нагоре и надолу, за да промените селекцията",
  "Use up and down arrows to move selection": "Използвайте стрелки нагоре и надолу, за да преместите селекцията",
  "Use up and down arrows to move lower selection": "Използвайте стрелки нагоре и надолу, за да преместите долната селекция",
  "Use up and down arrows to move upper selection": "Използвайте стрелки нагоре и надолу, за да преместите горната селекция",
  "From %1 to %2": "От %1 до %2",
  "From %1": "От %1",
  "To %1": "До %1",
  "No parser available for file: %1": "Няма наличен парсер за файл: %1",
  "Error parsing file: %1": "Грешка при сдвояването на файл: %1",
  "Unable to load file: %1": "Файлът не може да се зареди: %1",
  "Invalid date": "Невалидна дата"
};
export {
  a as Chart,
  c as Data,
  i as Export,
  s as Home,
  l as Image,
  t as Legend,
  r as Loading,
  e as Play,
  d as Print,
  n as Series,
  o as Stop,
  E as default
};
