// @arcgis/charts-components/dist/chunks/messages.de.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "Beim Laden des Diagramms ist ein Fehler aufgetreten.", n = "Eine logarithmische Transformation kann nicht auf negative oder Nullwerte angewendet werden.", t = "Eine Quadratwurzeltransformation kann nicht auf negative Werte angewendet werden.", a = "Beim Laden des Layers ist ein Fehler aufgetreten. URL = ${ url } Portal-Element-ID = ${ portalItemId }.", i = "Die in der Konfiguration definierten Felder [${ fieldNames }] wurden im Layer nicht gefunden.", r = "${ dataPath } muss eindeutig sein. Die Reihe mit dem Namen ${ seriesName } hat eine ID (${ seriesID }), die schon von einer anderen Reihe verwendet wird.", s = "${ dataPath } kann keine Aggregation, die nicht anzahlbasiert ist, für nicht numerische Felder durchführen.", d = "${ dataPath } fehlt eine Eigenschaft mit dem Namen ${ missingProperty }.", o = "${ dataPath } darf nicht kürzer als ${ limit } Zeichen sein.", l = "${ dataPath } darf nicht weniger als ${ limit } Elemente enthalten.", m = "${ dataPath } darf nicht mehr als ${ limit } Elemente enthalten.", c = "${ dataPath } muss mindestens ein Zeichen enthalten, bei dem es sich nicht um ein Leerstellenzeichen handelt.", h = "${ dataPath } darf nicht ${ additionalProperty } aufweisen.", u = "${ dataPath } muss einem der folgenden zulässigen Werte entsprechen: ${ allowedValues }.", g = "${ dataPath } muss mit einem der folgenden Schemas übereinstimmen: ${ schemaOptions }.", f = "Scatterplots mit proportionalen Symbolen werden nicht unterstützt. Die standardmäßige Symbolgröße wurde angewendet.", $ = "Fehler beim Lesen der Eingabedaten.", p = "Der queryBins-Vorgang wird für diesen Service nicht unterstützt.", w = 'Der erwartete Serientyp bei Index ${ seriesIndex } lautet "${ expectedType }", stattdessen wurde jedoch "${ receivedType }" empfangen.', y = "oder", P = {
  defaultError: e,
  negativeValueInDataForLogTransformation: n,
  negativeValueInDataForSqrtTransformation: t,
  layerLoadFailure: a,
  layerFieldsNotFound: i,
  duplicateSeriesID: r,
  nonNumericAggregation: s,
  requiredProperty: d,
  minLength: o,
  minItems: l,
  maxItems: m,
  whiteSpacePattern: c,
  additionalProperty: h,
  enumValues: u,
  anyOfValues: g,
  bubbleChartValidateMsg: f,
  queryError: $,
  queryBinsNotSupported: p,
  invalidSeriesType: w,
  or: y
};
export {
  h as additionalProperty,
  g as anyOfValues,
  f as bubbleChartValidateMsg,
  P as default,
  e as defaultError,
  r as duplicateSeriesID,
  u as enumValues,
  w as invalidSeriesType,
  i as layerFieldsNotFound,
  a as layerLoadFailure,
  m as maxItems,
  l as minItems,
  o as minLength,
  n as negativeValueInDataForLogTransformation,
  t as negativeValueInDataForSqrtTransformation,
  s as nonNumericAggregation,
  y as or,
  p as queryBinsNotSupported,
  $ as queryError,
  d as requiredProperty,
  c as whiteSpacePattern
};
