// @arcgis/charts-components/dist/chunks/messages.nl2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = { chart: "Diagram", barChart: "Staafdiagram", lineChart: "Lijndiagram", pieChart: "Cirkeldiagram", histogram: "Histogram", scatterplot: "Spreidingsdiagram", boxPlot: "Boxplot" }, a = { mean: "Gemiddelde", median: "Mediaan", sum: "Som", discretePercentile: "Discreet percentiel", minimum: "Minimum", maximum: "Maximum", variance: "Variantie", count: "Aantal", aggregation: "${ statistics } van ${ fieldName }", noAggregation: "Geen aggregatie" }, t = "Diagram", i = "Tellingserie", n = "X-as", s = "Niet beschikbaar", r = { start: "Koppelen met het eerste gegevenspunt", end: "Koppelen met het laatste gegevenspunt" }, m = {
  chartType: e,
  statistics: a,
  defaultTitle: t,
  countSeries: i,
  xAxis: n,
  notAvailable: s,
  timeAggregationTypes: r
};
export {
  e as chartType,
  i as countSeries,
  m as default,
  t as defaultTitle,
  s as notAvailable,
  a as statistics,
  r as timeAggregationTypes,
  n as xAxis
};
