// @arcgis/charts-components/dist/chunks/messages.fi2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const i = { chart: "Kaavio", barChart: "Palkkikaavio", lineChart: "Viivakaavio", pieChart: "Ympyräkaavio", histogram: "Histogrammi", scatterplot: "Pistekaavio", boxPlot: "Laatikkojana" }, a = { mean: "Keskiarvo", median: "Mediaani", sum: "Summa", discretePercentile: "Irrallinen prosenttipiste", minimum: "Vähimmäisarvo", maximum: "Enimmäisarvo", variance: "Varianssi", count: "Määrä", aggregation: "${ statistics }/${ fieldName }", noAggregation: "Ei koostetta" }, t = "Kaavio", e = "Laske sarjat", s = "X-akseli", n = "Ei käytettävissä", o = { start: "Kiinnitä ensimmäiseen aineistopisteeseen", end: "Kiinnitä viimeiseen aineistopisteeseen" }, r = {
  chartType: i,
  statistics: a,
  defaultTitle: t,
  countSeries: e,
  xAxis: s,
  notAvailable: n,
  timeAggregationTypes: o
};
export {
  i as chartType,
  e as countSeries,
  r as default,
  t as defaultTitle,
  n as notAvailable,
  a as statistics,
  o as timeAggregationTypes,
  s as xAxis
};
