// @arcgis/charts-components/dist/chunks/messages.fr.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "Une erreur est survenue lors du chargement du diagramme.", a = "Impossible d'appliquer une transformation logarithmique à des valeurs négatives ou égales à zéro.", t = "Impossible d'appliquer une transformation de racine carrée à des valeurs négatives.", s = "Une erreur est survenue lors du chargement de la couche. URL = ${ url }. ID de l'élément du portail = ${ portalItemId }.", r = "Les champs [${ fieldNames }] définis dans la configuration sont introuvables dans la couche.", n = "${ dataPath } doit être unique. La série nommée ${ seriesName } comporte un ID (${ seriesID }) qui est déjà utilisé par une autre série.", o = "${ dataPath } ne peut pas réaliser d'agrégation de type autre que Total sur un champ non numérique.", i = "Propriété ${ missingProperty } manquante dans ${ dataPath }.", u = "${ dataPath } ne doit pas comporter moins de ${ limit } caractères.", d = "${ dataPath } ne doit pas comporter moins de ${ limit } éléments.", l = "${ dataPath } ne doit pas comporter plus de ${ limit } éléments.", c = "${ dataPath } doit contenir au moins un caractère autre qu'un espace.", p = "${ dataPath } ne doit pas contenir ${ additionalProperty }.", m = "${ dataPath } doit être égal à l'une de ces valeurs autorisées : ${ allowedValues }.", $ = "${ dataPath } doit correspondre à l'une des structures suivantes : ${ schemaOptions }.", g = "Les nuages de points avec des symboles proportionnels ne sont pas pris en charge. La taille de symbole par défaut a été appliquée.", h = "Échec de la lecture des données en entrée.", y = "L’opération queryBins n’est pas prise en charge sur ce service.", P = "Le type de série attendu à l’index ${ seriesIndex } est '${ expectedType }', mais '${ receivedType }' a été reçu à la place", q = "ou", v = {
  defaultError: e,
  negativeValueInDataForLogTransformation: a,
  negativeValueInDataForSqrtTransformation: t,
  layerLoadFailure: s,
  layerFieldsNotFound: r,
  duplicateSeriesID: n,
  nonNumericAggregation: o,
  requiredProperty: i,
  minLength: u,
  minItems: d,
  maxItems: l,
  whiteSpacePattern: c,
  additionalProperty: p,
  enumValues: m,
  anyOfValues: $,
  bubbleChartValidateMsg: g,
  queryError: h,
  queryBinsNotSupported: y,
  invalidSeriesType: P,
  or: q
};
export {
  p as additionalProperty,
  $ as anyOfValues,
  g as bubbleChartValidateMsg,
  v as default,
  e as defaultError,
  n as duplicateSeriesID,
  m as enumValues,
  P as invalidSeriesType,
  r as layerFieldsNotFound,
  s as layerLoadFailure,
  l as maxItems,
  d as minItems,
  u as minLength,
  a as negativeValueInDataForLogTransformation,
  t as negativeValueInDataForSqrtTransformation,
  o as nonNumericAggregation,
  q as or,
  y as queryBinsNotSupported,
  h as queryError,
  i as requiredProperty,
  c as whiteSpacePattern
};
