// @arcgis/charts-components/dist/chunks/messages.cs2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const t = { chart: "Graf", barChart: "Pruhový graf", lineChart: "Spojnicový graf", pieChart: "Výsečový graf", histogram: "Histogram", scatterplot: "Bodový graf", boxPlot: "Krabicový graf" }, a = { mean: "Průměr", median: "Medián", sum: "Suma", discretePercentile: "Diskrétní percentil", minimum: "Minimum", maximum: "Maximum", variance: "Rozptyl", count: "Počet", aggregation: "${ statistics } z ${ fieldName }", noAggregation: "Žádná agregace" }, i = "Graf", e = "Počet sérií", o = "Osa X", r = "Není k dispozici", s = { start: "Přichytit k prvnímu datovému bodu", end: "Přichytit k poslednímu datovému bodu" }, n = {
  chartType: t,
  statistics: a,
  defaultTitle: i,
  countSeries: e,
  xAxis: o,
  notAvailable: r,
  timeAggregationTypes: s
};
export {
  t as chartType,
  e as countSeries,
  n as default,
  i as defaultTitle,
  r as notAvailable,
  a as statistics,
  s as timeAggregationTypes,
  o as xAxis
};
