// @arcgis/charts-components/dist/chunks/messages.pt-PT.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "Ocorreu um erro ao carregar o gráfico.", o = "Não é possível aplicar uma transformação de registo a valores negativos ou de zero.", a = "Não é possível aplicar uma transformação de raiz quadrada a valores negativos.", r = "Ocorreu um erro ao carregar a camada. URL = ${ url }. ID do item do portal = ${ portalItemId }.", t = "Os campos [${ fieldNames }] definidos na configuração não foram encontrados na camada.", s = "${ dataPath } tem de ser único. A série com o nome ${ seriesName } tem um ID (${ seriesID }) que já está a ser usado por outra série.", n = "${ dataPath } não pode realizar agregação sem contagem num campo não numérico.", i = "${ dataPath } tem uma propriedade com o nome ${ missingProperty } em falta.", d = "${ dataPath } não pode ser inferior a ${ limit } caracteres.", m = "${ dataPath } não pode conter menos do que ${ limit } itens.", c = "${ dataPath } não pode conter mais do que ${ limit } itens.", u = "${ dataPath } tem de ter, pelo menos, um caráter que não seja um espaço em branco.", p = "${ dataPath } não pode conter ${ additionalProperty }.", l = "${ dataPath } tem de ser igual a um destes valores permitidos: ${ allowedValues }.", $ = "${ dataPath } tem de corresponder ao esquema de um dos seguintes: ${ schemaOptions }.", g = "Os gráficos de dispersão com símbolos proporcionais não são suportados. Foi aplicado o tamanho de símbolo predefinido.", f = "Falha ao ler os dados de entrada.", h = "A operação queryBins não é suportada por este serviço.", P = "O tipo de série esperado no índice ${ seriesIndex } é '${ expectedType }', mas '${ receivedType }' foi recebido em vez disso", v = "ou", y = {
  defaultError: e,
  negativeValueInDataForLogTransformation: o,
  negativeValueInDataForSqrtTransformation: a,
  layerLoadFailure: r,
  layerFieldsNotFound: t,
  duplicateSeriesID: s,
  nonNumericAggregation: n,
  requiredProperty: i,
  minLength: d,
  minItems: m,
  maxItems: c,
  whiteSpacePattern: u,
  additionalProperty: p,
  enumValues: l,
  anyOfValues: $,
  bubbleChartValidateMsg: g,
  queryError: f,
  queryBinsNotSupported: h,
  invalidSeriesType: P,
  or: v
};
export {
  p as additionalProperty,
  $ as anyOfValues,
  g as bubbleChartValidateMsg,
  y as default,
  e as defaultError,
  s as duplicateSeriesID,
  l as enumValues,
  P as invalidSeriesType,
  t as layerFieldsNotFound,
  r as layerLoadFailure,
  c as maxItems,
  m as minItems,
  d as minLength,
  o as negativeValueInDataForLogTransformation,
  a as negativeValueInDataForSqrtTransformation,
  n as nonNumericAggregation,
  v as or,
  h as queryBinsNotSupported,
  f as queryError,
  i as requiredProperty,
  u as whiteSpacePattern
};
