// @arcgis/charts-components/dist/chunks/messages.hr.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const a = "Došlo je do pogreške tijekom učitavanja grafikona.", e = "Ne može se primijeniti transformacija dnevnika na negativne ili nulte vrijednosti.", i = "Ne može se primijeniti transformacija kvadratnog korijena na negativne vrijednosti.", o = "Došlo je do pogreške tijekom učitavanja sloja. URL = ${ url }. Id stavke portala = ${ portalItemId }.", n = "Polja [${ fieldNames }] koja su definirana u konfiguraciji nisu pronađena u sloju.", t = "${ dataPath } mora biti jedinstveno. Serija naziva ${ seriesName } ima ID (${ seriesID }) koji već koristi druga serija.", r = "${ dataPath } ne može izvoditi nebrojeno agregiranje na ne-numeričkom polju.", s = "${ dataPath } nedostaje svojstvo naziva ${ missingProperty }.", d = "${ dataPath } mora biti kraće od ${ limit } znakova.", j = "${ dataPath } ne smije imati manje od ${ limit } stavki.", m = "${ dataPath } ne smije imati više od ${ limit } stavki.", l = "${ dataPath } mora imati barem jedan znak koji nije razmak.", u = "${ dataPath } ne smije imati ${ additionalProperty }.", p = "${ dataPath } mora biti jednak jednoj od ovih dopuštenih vrijednosti: ${ allowedValues }.", v = "${ dataPath } mora odgovarati shemi jednog od ovih: ${ schemaOptions }.", c = "Raspršeni grafikoni s proporcionalnim simbolima nisu podržani. Primijenjena je zadana veličina simbola.", k = "Nije uspjelo čitanje ulaznih podataka.", g = "Radnja queryBins nije podržana u sklopu ove usluge.", $ = "Očekivana vrsta serije po indeksu je ${ seriesIndex } je '${ expectedType }' ali primljeno je '${ receivedType }' umjesto toga", h = "ili", P = {
  defaultError: a,
  negativeValueInDataForLogTransformation: e,
  negativeValueInDataForSqrtTransformation: i,
  layerLoadFailure: o,
  layerFieldsNotFound: n,
  duplicateSeriesID: t,
  nonNumericAggregation: r,
  requiredProperty: s,
  minLength: d,
  minItems: j,
  maxItems: m,
  whiteSpacePattern: l,
  additionalProperty: u,
  enumValues: p,
  anyOfValues: v,
  bubbleChartValidateMsg: c,
  queryError: k,
  queryBinsNotSupported: g,
  invalidSeriesType: $,
  or: h
};
export {
  u as additionalProperty,
  v as anyOfValues,
  c as bubbleChartValidateMsg,
  P as default,
  a as defaultError,
  t as duplicateSeriesID,
  p as enumValues,
  $ as invalidSeriesType,
  n as layerFieldsNotFound,
  o as layerLoadFailure,
  m as maxItems,
  j as minItems,
  d as minLength,
  e as negativeValueInDataForLogTransformation,
  i as negativeValueInDataForSqrtTransformation,
  r as nonNumericAggregation,
  h as or,
  g as queryBinsNotSupported,
  k as queryError,
  s as requiredProperty,
  l as whiteSpacePattern
};
