// @arcgis/charts-components/dist/chunks/messages.ar4.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "تشغيل", o = "إيقاف", t = "وسيلة الإيضاح", r = "تحميل", s = "الصفحة الرئيسية", a = "الرسم البياني", n = "سلاسل", i = "تصدير", l = "صورة", c = "البيانات", d = "طباعة", E = {
  "Zoom Out": "تصغير",
  Play: e,
  Stop: o,
  Legend: t,
  "Press ENTER to toggle": "اضغط على ENTER للتبديل",
  Loading: r,
  Home: s,
  Chart: a,
  "Serial chart": "المخطط التسلسلي",
  "X/Y chart": "مخطط X/Y",
  "Pie chart": "مخطط دائري",
  "Gauge chart": "مخطط القياس",
  "Radar chart": "مخطط الرادار",
  "Sankey diagram": "مخطط سانكي",
  "Flow diagram": "مخطط التدفق",
  "Chord diagram": "مخطط وتري",
  "TreeMap chart": "مخطط هيكلي",
  "Sliced chart": "مخطط شرائح",
  Series: n,
  "Candlestick Series": "سلسلة الشمعدان",
  "OHLC Series": "سلسلة OHLC",
  "Column Series": "سلسلة العمود",
  "Line Series": "سلسلة الخط",
  "Pie Slice Series": "سلسلة الشريحة الدائرية",
  "Funnel Series": "سلسلة القُمع",
  "Pyramid Series": "سلسلة الهرم",
  "X/Y Series": "سلسلة X/Y",
  Map: "الخريطة",
  "Press ENTER to zoom in": "اضغط على ENTER للتكبير",
  "Press ENTER to zoom out": "اضغط على ENTER للتصغير",
  "Use arrow keys to zoom in and out": "استخدم مفاتيح الأسهم للتكبير والتصغير",
  "Use plus and minus keys on your keyboard to zoom in and out": "استخدم مفاتيح الجمع والطرح من على لوحة المفاتيح للتكبير والتصغير",
  Export: i,
  Image: l,
  Data: c,
  Print: d,
  "Press ENTER to open": "اضغط على ENTER للفتح",
  "Press ENTER to print.": "اضغط على ENTER للطباعة.",
  "Press ENTER to export as %1.": "اضغط على ENTER للتصدير كـ %1.",
  "(Press ESC to close this message)": "(اضغط على ESC لإغلاق هذه الرسالة)",
  "Image Export Complete": "اكتمل تصدير الصورة",
  "Export operation took longer than expected. Something might have gone wrong.": "استغرقت عملية التصدير وقتًا أطول من المتوقع. ربما حدث خطأ ما.",
  "Saved from": "تم الحفظ من",
  "Use TAB to select grip buttons or left and right arrows to change selection": "استخدم زر TAB للتنقل بين أزرار الإمساك، أو استخدم الأسهم اليمنى واليسرى لتغيير التحديد",
  "Use left and right arrows to move selection": "استخدم الأسهم اليسرى واليمنى لتحريك التحديد",
  "Use left and right arrows to move left selection": "استخدم الأسهم اليسرى واليمنى لتحريك التحديد الأيسر",
  "Use left and right arrows to move right selection": "استخدم الأسهم اليمنى واليسرى لتحريك التحديد إلى اليمين",
  "Use TAB select grip buttons or up and down arrows to change selection": "استخدم أزرار التحكم في التحديد TAB أو الأسهم لأعلى ولأسفل لتغيير التحديد",
  "Use up and down arrows to move selection": "استخدم الأسهم لأعلى ولأسفل لتحريك التحديد",
  "Use up and down arrows to move lower selection": "استخدم الأسهم لأعلى ولأسفل لتحريك التحديد إلى الأسفل",
  "Use up and down arrows to move upper selection": "استخدم الأسهم لأعلى ولأسفل لتحريك التحديد العلوي",
  "From %1 to %2": "من %1 إلى %2",
  "From %1": "من %1",
  "To %1": "إلى %1",
  "No parser available for file: %1": "لا يوجد محلل متاح للملف: %1",
  "Error parsing file: %1": "خطأ في تحليل الملف: %1",
  "Unable to load file: %1": "يتعذر تحميل الملف: %1",
  "Invalid date": "التاريخ غير صحيح"
};
export {
  a as Chart,
  c as Data,
  i as Export,
  s as Home,
  l as Image,
  t as Legend,
  r as Loading,
  e as Play,
  d as Print,
  n as Series,
  o as Stop,
  E as default
};
