// @arcgis/charts-components/dist/chunks/messages.sv4.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const r = "Spela", t = "Stoppa", e = "Teckenförklaring", a = "Läser in", o = "Hem", n = "Diagram", i = "Serier", s = "Exportera", l = "Bild", d = "Data", p = "Skriv ut", g = {
  "Zoom Out": "Zooma ut",
  Play: r,
  Stop: t,
  Legend: e,
  "Press ENTER to toggle": "Tryck på ENTER för att växla",
  Loading: a,
  Home: o,
  Chart: n,
  "Serial chart": "Seriediagram",
  "X/Y chart": "X/Y-diagram",
  "Pie chart": "Cirkeldiagram",
  "Gauge chart": "Mätardiagram",
  "Radar chart": "Polärdiagram",
  "Sankey diagram": "Sankey-diagram",
  "Flow diagram": "Flödesdiagram",
  "Chord diagram": "Kordadiagram",
  "TreeMap chart": "TreeMap-diagram",
  "Sliced chart": "Sektordiagram",
  Series: i,
  "Candlestick Series": "Ljusstakeserie",
  "OHLC Series": "OHLC-serie",
  "Column Series": "Kolumnserie",
  "Line Series": "Linjeserie",
  "Pie Slice Series": "Cirkelsektioner – serie",
  "Funnel Series": "Trattserie",
  "Pyramid Series": "Pyramidserie",
  "X/Y Series": "X/Y-serie",
  Map: "Karta",
  "Press ENTER to zoom in": "Tryck på ENTER för att zooma in",
  "Press ENTER to zoom out": "Tryck på ENTER för att zooma ut",
  "Use arrow keys to zoom in and out": "Använd piltangenterna för att zooma in och ut",
  "Use plus and minus keys on your keyboard to zoom in and out": "Använd plus- och minustecknen på ditt tangentbord för att zooma in och ut",
  Export: s,
  Image: l,
  Data: d,
  Print: p,
  "Press ENTER to open": "Tryck på ENTER för att öppna",
  "Press ENTER to print.": "Tryck på ENTER för att skriva ut.",
  "Press ENTER to export as %1.": "Tryck på ENTER för att exportera som %1.",
  "(Press ESC to close this message)": "(Tryck på ESC för att stänga det här meddelandet)",
  "Image Export Complete": "Bildexporten är klar",
  "Export operation took longer than expected. Something might have gone wrong.": "Exportåtgärden tog längre tid än förväntat. Något kan ha gått fel.",
  "Saved from": "Sparat från",
  "Use TAB to select grip buttons or left and right arrows to change selection": "Använd TABB för att välja greppknappar eller vänster- och högerpilarna för att ändra urval",
  "Use left and right arrows to move selection": "Använd vänster- och högerpilarna för att flytta urval",
  "Use left and right arrows to move left selection": "Använd vänster- och högerpilarna för att flytta vänster urval",
  "Use left and right arrows to move right selection": "Använd vänster- och högerpilarna för att flytta höger urval",
  "Use TAB select grip buttons or up and down arrows to change selection": "Använd TABB för att välja greppknappar eller uppåt- och nedåtpilarna för att ändra urval",
  "Use up and down arrows to move selection": "Använd uppåt- och nedåtpilarna för att flytta urval",
  "Use up and down arrows to move lower selection": "Använd uppåt- och nedåtpilarna för att flytta nedre urval",
  "Use up and down arrows to move upper selection": "Använd uppåt- och nedåtpilarna för att flytta övre urval",
  "From %1 to %2": "Från %1 till %2",
  "From %1": "Från %1",
  "To %1": "Till %1",
  "No parser available for file: %1": "Ingen parser tillgänglig för fil: %1",
  "Error parsing file: %1": "Fel vid tolkning av fil: %1",
  "Unable to load file: %1": "Det går inte att läsa in fil: %1",
  "Invalid date": "Ogiltigt datum"
};
export {
  n as Chart,
  d as Data,
  s as Export,
  o as Home,
  l as Image,
  e as Legend,
  a as Loading,
  r as Play,
  p as Print,
  i as Series,
  t as Stop,
  g as default
};
