// @arcgis/charts-components/dist/chunks/messages.bg2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const t = { chart: "Диаграма", barChart: "Диаграма във форма на лента", lineChart: "Диаграма във форма на линия", pieChart: "Диаграма в кръгла форма", histogram: "Хистограма", scatterplot: "Toчкова диаграма", boxPlot: "Каре диаграма" }, e = { mean: "Осреднена стойност", median: "Медиана", sum: "Сума", discretePercentile: "Дискретен перцентил", minimum: "Минимални", maximum: "Максимални", variance: "Отклонение", count: "Брой", aggregation: "${ statistics } от ${ fieldName }", noAggregation: "Няма агрегиране" }, a = "Диаграма", s = "Серия за броене", i = "Ос Х", n = "Няма налични", o = { start: "Притеглете до първата точка от данни", end: "Притеглете до последната точка от данни" }, c = {
  chartType: t,
  statistics: e,
  defaultTitle: a,
  countSeries: s,
  xAxis: i,
  notAvailable: n,
  timeAggregationTypes: o
};
export {
  t as chartType,
  s as countSeries,
  c as default,
  a as defaultTitle,
  n as notAvailable,
  e as statistics,
  o as timeAggregationTypes,
  i as xAxis
};
