// @arcgis/charts-components/dist/chunks/messages.cs.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "Při načítání grafu došlo k chybě.", o = "Nelze uplatnit logaritmickou transformaci na záporné nebo nulové hodnoty.", a = "Nelze uplatnit transformaci odmocniny na záporné hodnoty.", t = "Při načítání vrstvy došlo k chybě. URL = ${ url }. ID položky portálu = ${ portalItemId }.", n = "Pole [${ fieldNames }] definovaná v konfiguraci nebyla nalezena ve vrstvě.", s = "${ dataPath } musí být jedinečné. Řada s názvem ${ seriesName } má ID (${ seriesID }), které již používá jiná řada.", r = "${ dataPath } nemůže provádět jiné seskupení než podle počtu na jiném než numerickém poli.", i = "V ${ dataPath } chybí vlastnost s názvem ${ missingProperty }.", l = "${ dataPath } nesmí být kratší než ${ limit } znaků.", d = "${ dataPath } nesmí obsahovat méně než ${ limit } položek.", m = "${ dataPath } nesmí obsahovat více než ${ limit } položek.", c = "${ dataPath } musí obsahovat alespoň jeden znak jiný než mezera.", p = "${ dataPath } nesmí obsahovat ${ additionalProperty }.", u = "${ dataPath } se musí rovnat jedné z těchto povolených hodnot: ${ allowedValues }.", v = "${ dataPath } se musí shodovat s jedním z těchto schémat: ${ schemaOptions }.", h = "Rozptýlené bodové grafy s proporcionálními symboly nejsou podporovány. Byla použita výchozí velikost symbolu.", y = "Vstupní data se nepodařilo přečíst.", $ = "Operace queryBins není v této službě podporována.", b = "Očekávaný typ řady s indexem ${ seriesIndex } je '${ expectedType }', ale obdrželi jsme '${ receivedType }'", P = "nebo", k = {
  defaultError: e,
  negativeValueInDataForLogTransformation: o,
  negativeValueInDataForSqrtTransformation: a,
  layerLoadFailure: t,
  layerFieldsNotFound: n,
  duplicateSeriesID: s,
  nonNumericAggregation: r,
  requiredProperty: i,
  minLength: l,
  minItems: d,
  maxItems: m,
  whiteSpacePattern: c,
  additionalProperty: p,
  enumValues: u,
  anyOfValues: v,
  bubbleChartValidateMsg: h,
  queryError: y,
  queryBinsNotSupported: $,
  invalidSeriesType: b,
  or: P
};
export {
  p as additionalProperty,
  v as anyOfValues,
  h as bubbleChartValidateMsg,
  k as default,
  e as defaultError,
  s as duplicateSeriesID,
  u as enumValues,
  b as invalidSeriesType,
  n as layerFieldsNotFound,
  t as layerLoadFailure,
  m as maxItems,
  d as minItems,
  l as minLength,
  o as negativeValueInDataForLogTransformation,
  a as negativeValueInDataForSqrtTransformation,
  r as nonNumericAggregation,
  P as or,
  $ as queryBinsNotSupported,
  y as queryError,
  i as requiredProperty,
  c as whiteSpacePattern
};
