// @arcgis/charts-components/dist/chunks/messages.et.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "Diagrammi laadimisel ilmnes viga.", a = "Logi teisendamisel ei saa kasutada negatiivseid väärtusi ega nullväärtusi.", t = "Ruutjuure teisendamist ei saa rakendada negatiivsetele väärtustele.", i = "Kihi laadimisel tekkis viga. URL = ${ url }. Portaaliüksuse ID = ${ portalItemId }.", s = "Konfiguratsioonis määratletud välju [${ fieldNames }] kihis ei leitud.", o = "${ dataPath } peab olema kordumatu. Seerial nimega ${ seriesName } on ID (${ seriesID }), mida kasutatakse juba muudes seeriates.", u = "${ dataPath } ei tohi mittenumbrilisel väljal teha loendamisega mitteseotud agregeerimist.", n = "Omadusel ${ dataPath } puudub omadus nimega ${ missingProperty }.", l = "${ dataPath } ei tohi olla lühem kui ${ limit } märki.", r = "Muutujal ${ dataPath } ei tohi olla vähem kui ${ limit } üksust.", d = "Muutujal ${ dataPath } ei tohi olla rohkem kui ${ limit } üksust.", m = "${ dataPath } peab sisaldama vähemalt ühte märki, mis pole tühimärk.", h = "Muutujal ${ dataPath } ei tohi olla: ${ additionalProperty }.", g = "${ dataPath } peab võrduma ühega allpool nimetatud väärtustest: ${ allowedValues }.", c = "${ dataPath } peab vastama ühele nendest skeemidest: ${ schemaOptions }.", p = "Proportsionaalsete sümbolitega hajuvusdiagramme ei toetata. Rakendatud on sümboli vaikesuurus.", $ = "Sisendandmete lugemine nurjus.", k = "See teenus ei toeta toimingut queryBins.", v = "Eeldatud seeriatüüp indeksis ${ seriesIndex } on '${ expectedType }', kuid saadi hoopis '${ receivedType }'", P = "või", y = {
  defaultError: e,
  negativeValueInDataForLogTransformation: a,
  negativeValueInDataForSqrtTransformation: t,
  layerLoadFailure: i,
  layerFieldsNotFound: s,
  duplicateSeriesID: o,
  nonNumericAggregation: u,
  requiredProperty: n,
  minLength: l,
  minItems: r,
  maxItems: d,
  whiteSpacePattern: m,
  additionalProperty: h,
  enumValues: g,
  anyOfValues: c,
  bubbleChartValidateMsg: p,
  queryError: $,
  queryBinsNotSupported: k,
  invalidSeriesType: v,
  or: P
};
export {
  h as additionalProperty,
  c as anyOfValues,
  p as bubbleChartValidateMsg,
  y as default,
  e as defaultError,
  o as duplicateSeriesID,
  g as enumValues,
  v as invalidSeriesType,
  s as layerFieldsNotFound,
  i as layerLoadFailure,
  d as maxItems,
  r as minItems,
  l as minLength,
  a as negativeValueInDataForLogTransformation,
  t as negativeValueInDataForSqrtTransformation,
  u as nonNumericAggregation,
  P as or,
  k as queryBinsNotSupported,
  $ as queryError,
  n as requiredProperty,
  m as whiteSpacePattern
};
