// @arcgis/charts-components/dist/chunks/messages.da2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const a = { chart: "Diagram", barChart: "Søjlediagram", lineChart: "Linjediagram", pieChart: "Cirkeldiagram", histogram: "Histogram", scatterplot: "Punktdiagram", boxPlot: "Kassediagram" }, t = { mean: "Middel", median: "Median", sum: "Sum", discretePercentile: "Særskilt percentil", minimum: "Minimum", maximum: "Maksimum", variance: "Varians", count: "Tælling", aggregation: "${ statistics } for ${ fieldName }", noAggregation: "Ingen aggregering" }, e = "Diagram", i = "Tællerække", s = "X-akse", r = "Ikke tilgængelig", n = { start: "Fastgør til det første datapunkt", end: "Fastgør til det sidste datapunkt" }, g = {
  chartType: a,
  statistics: t,
  defaultTitle: e,
  countSeries: i,
  xAxis: s,
  notAvailable: r,
  timeAggregationTypes: n
};
export {
  a as chartType,
  i as countSeries,
  g as default,
  e as defaultTitle,
  r as notAvailable,
  t as statistics,
  n as timeAggregationTypes,
  s as xAxis
};
