// @arcgis/charts-components/dist/chunks/messages.no2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const t = { chart: "Diagram", barChart: "Stolpediagram", lineChart: "Linjediagram", pieChart: "Sektordiagram", histogram: "Histogram", scatterplot: "Punktdiagram", boxPlot: "Boksdiagram" }, a = { mean: "Gjennomsnitt", median: "Median", sum: "Sum", discretePercentile: "Diskret persentil", minimum: "Minimum", maximum: "Maksimum", variance: "Varians", count: "Antall", aggregation: "${ statistics } av ${ fieldName }", noAggregation: "Ingen sammenslåing" }, e = "Diagram", i = "Tallserie", s = "X-akse", n = "Ikke tilgjengelig", r = { start: "Fest til det første datapunktet", end: "Fest til det siste datapunktet" }, m = {
  chartType: t,
  statistics: a,
  defaultTitle: e,
  countSeries: i,
  xAxis: s,
  notAvailable: n,
  timeAggregationTypes: r
};
export {
  t as chartType,
  i as countSeries,
  m as default,
  e as defaultTitle,
  n as notAvailable,
  a as statistics,
  r as timeAggregationTypes,
  s as xAxis
};
