// @arcgis/charts-components/dist/chunks/messages.en.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "An error occurred while loading the chart.", t = "Cannot apply a log transformation to negative or zero values.", a = "Cannot apply a square root transformation to negative values.", o = "An error occurred while loading the layer. URL = ${ url }. Portal item ID = ${ portalItemId }.", n = "The fields [${ fieldNames }] defined in the config were not found in the layer.", s = "${ dataPath } must be unique. The series named ${ seriesName } has an ID (${ seriesID }) that is already used by another series.", r = "${ dataPath } cannot perform non-count aggregation on non-numeric field.", i = "${ dataPath } is missing a property named ${ missingProperty }.", d = "${ dataPath } must not be shorter than ${ limit } characters.", l = "${ dataPath } must not have fewer than ${ limit } items.", c = "${ dataPath } must not have more than ${ limit } items.", h = "${ dataPath } must have at least one non-whitespace character.", u = "${ dataPath } must not have ${ additionalProperty }.", m = "${ dataPath } must be equal to one of these allowed values: ${ allowedValues }.", p = "${ dataPath } must match the schema of one of these: ${ schemaOptions }.", $ = "Scatterplots with proportional symbols are not supported. Default symbol size has been applied.", y = "Failure to read input data.", g = "The queryBins operation is not supported on this service.", f = "The expected series type at index ${ seriesIndex } is '${ expectedType }' but '${ receivedType }' was received instead", P = "or", v = {
  defaultError: e,
  negativeValueInDataForLogTransformation: t,
  negativeValueInDataForSqrtTransformation: a,
  layerLoadFailure: o,
  layerFieldsNotFound: n,
  duplicateSeriesID: s,
  nonNumericAggregation: r,
  requiredProperty: i,
  minLength: d,
  minItems: l,
  maxItems: c,
  whiteSpacePattern: h,
  additionalProperty: u,
  enumValues: m,
  anyOfValues: p,
  bubbleChartValidateMsg: $,
  queryError: y,
  queryBinsNotSupported: g,
  invalidSeriesType: f,
  or: P
};
export {
  u as additionalProperty,
  p as anyOfValues,
  $ as bubbleChartValidateMsg,
  v as default,
  e as defaultError,
  s as duplicateSeriesID,
  m as enumValues,
  f as invalidSeriesType,
  n as layerFieldsNotFound,
  o as layerLoadFailure,
  c as maxItems,
  l as minItems,
  d as minLength,
  t as negativeValueInDataForLogTransformation,
  a as negativeValueInDataForSqrtTransformation,
  r as nonNumericAggregation,
  P as or,
  g as queryBinsNotSupported,
  y as queryError,
  i as requiredProperty,
  h as whiteSpacePattern
};
