// @arcgis/charts-components/dist/chunks/messages.bg.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const t = "Възникна грешка при зареждането на диаграмата.", a = "Не може да се приложи логаритмична трансформация към отрицателни или нулеви стойности.", e = "Не може да се приложи трансформация с квадратен корен към отрицателни стойности.", s = "Възникна грешка при зареждането на слоя. URL = ${ url }. ИД на елемент на портал = ${ portalItemId }.", o = "Полетата [${ fieldNames }], които са дефинирани в конфигурацията, не са намерени в слоя.", n = "${ dataPath } трябва да бъде унукално. Серията с името ${ seriesName } има ИД (${ seriesID }), който вече се използва от друга серия.", r = "${ dataPath } не може да се извършва обобщение, различно от обобщение на брой, за нецифрово поле.", i = "${ dataPath } няма име на свойство ${ missingProperty }.", d = "${ dataPath } не трябва да има по-малко от ${ limit } символи.", c = "${ dataPath } не трябва да има по-малко от ${ limit } елементи.", l = "${ dataPath } не трябва да има повече от ${ limit } елементи.", $ = "${ dataPath } трябва да има поне един символ, който не е интервал.", u = "${ dataPath } не трябва да бъде ${ additionalProperty }.", m = "${ dataPath } трябва да бъде равна на една от тези позволени стойности: ${ allowedValues }.", p = "${ dataPath }трябва да съответства на схемата на една от тези: ${ schemaOptions }.", P = "Не се поддържат диаграми на разсейване с пропорционални символи. Използван е размер на символа по подразбиране.", h = "Входните данни не могат да се прочетат.", y = "Операцията queryBins не се поддържа от тази услуга.", g = "Очакваният тип серия в индекс${ seriesIndex } е „${ expectedType }“, но „${ receivedType }“ е получен вместо това", I = "или", f = {
  defaultError: t,
  negativeValueInDataForLogTransformation: a,
  negativeValueInDataForSqrtTransformation: e,
  layerLoadFailure: s,
  layerFieldsNotFound: o,
  duplicateSeriesID: n,
  nonNumericAggregation: r,
  requiredProperty: i,
  minLength: d,
  minItems: c,
  maxItems: l,
  whiteSpacePattern: $,
  additionalProperty: u,
  enumValues: m,
  anyOfValues: p,
  bubbleChartValidateMsg: P,
  queryError: h,
  queryBinsNotSupported: y,
  invalidSeriesType: g,
  or: I
};
export {
  u as additionalProperty,
  p as anyOfValues,
  P as bubbleChartValidateMsg,
  f as default,
  t as defaultError,
  n as duplicateSeriesID,
  m as enumValues,
  g as invalidSeriesType,
  o as layerFieldsNotFound,
  s as layerLoadFailure,
  l as maxItems,
  c as minItems,
  d as minLength,
  a as negativeValueInDataForLogTransformation,
  e as negativeValueInDataForSqrtTransformation,
  r as nonNumericAggregation,
  I as or,
  y as queryBinsNotSupported,
  h as queryError,
  i as requiredProperty,
  $ as whiteSpacePattern
};
