// @arcgis/charts-components/dist/chunks/messages.en-US4.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "Play", o = "Stop", r = "Legend", t = "Loading", s = "Home", a = "Chart", n = "Series", i = "Export", l = "Image", d = "Data", c = "Print", m = {
  "Zoom Out": "Zoom Out",
  Play: e,
  Stop: o,
  Legend: r,
  "Press ENTER to toggle": "Press ENTER to toggle",
  Loading: t,
  Home: s,
  Chart: a,
  "Serial chart": "Serial chart",
  "X/Y chart": "X/Y chart",
  "Pie chart": "Pie chart",
  "Gauge chart": "Gauge chart",
  "Radar chart": "Radar chart",
  "Sankey diagram": "Sankey diagram",
  "Flow diagram": "Flow diagram",
  "Chord diagram": "Chord diagram",
  "TreeMap chart": "TreeMap chart",
  "Sliced chart": "Sliced chart",
  Series: n,
  "Candlestick Series": "Candlestick Series",
  "OHLC Series": "OHLC Series",
  "Column Series": "Column Series",
  "Line Series": "Line Series",
  "Pie Slice Series": "Pie Slice Series",
  "Funnel Series": "Funnel Series",
  "Pyramid Series": "Pyramid Series",
  "X/Y Series": "X/Y Series",
  Map: "Map",
  "Press ENTER to zoom in": "Press ENTER to zoom in",
  "Press ENTER to zoom out": "Press ENTER to zoom out",
  "Use arrow keys to zoom in and out": "Use arrow keys to zoom in and out",
  "Use plus and minus keys on your keyboard to zoom in and out": "Use plus and minus keys on your keyboard to zoom in and out",
  Export: i,
  Image: l,
  Data: d,
  Print: c,
  "Press ENTER to open": "Press ENTER to open",
  "Press ENTER to print.": "Press ENTER to print.",
  "Press ENTER to export as %1.": "Press ENTER to export as %1.",
  "(Press ESC to close this message)": "(Press ESC to close this message)",
  "Image Export Complete": "Image Export Complete",
  "Export operation took longer than expected. Something might have gone wrong.": "Export operation took longer than expected. Something might have gone wrong.",
  "Saved from": "Saved from",
  "Use TAB to select grip buttons or left and right arrows to change selection": "Use TAB to select grip buttons or left and right arrows to change selection",
  "Use left and right arrows to move selection": "Use left and right arrows to move selection",
  "Use left and right arrows to move left selection": "Use left and right arrows to move left selection",
  "Use left and right arrows to move right selection": "Use left and right arrows to move right selection",
  "Use TAB select grip buttons or up and down arrows to change selection": "Use TAB select grip buttons or up and down arrows to change selection",
  "Use up and down arrows to move selection": "Use up and down arrows to move selection",
  "Use up and down arrows to move lower selection": "Use up and down arrows to move lower selection",
  "Use up and down arrows to move upper selection": "Use up and down arrows to move upper selection",
  "From %1 to %2": "From %1 to %2",
  "From %1": "From %1",
  "To %1": "To %1",
  "No parser available for file: %1": "No parser available for file: %1",
  "Error parsing file: %1": "Error parsing file: %1",
  "Unable to load file: %1": "Unable to load file: %1",
  "Invalid date": "Invalid date"
};
export {
  a as Chart,
  d as Data,
  i as Export,
  s as Home,
  l as Image,
  r as Legend,
  t as Loading,
  e as Play,
  c as Print,
  n as Series,
  o as Stop,
  m as default
};
