// @arcgis/charts-components/dist/chunks/messages.id.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const a = "Terjadi kesalahan saat memuat grafik.", i = "Tidak dapat menerapkan transformasi log ke nilai negatif atau nol.", e = "Tidak dapat menerapkan transformasi akar pangkat dua ke nilai negatif.", t = "Terjadi kesalahan saat memuat layer. URL = ${ url }. ID item portal = ${ portalItemId }.", n = "Kolom [${ fieldNames }] yang ditentukan dalam konfigurasi tidak ditemukan di layer.", r = "${ dataPath } harus unik. Rangkaian yang bernama ${ seriesName } memiliki ID (${ seriesID }) yang sudah digunakan oleh rangkaian lain.", s = "${ dataPath } tidak dapat melakukan agregasi non-hitungan di kolom non-numerik.", d = "${ dataPath } tidak memiliki properti bernama ${ missingProperty }.", o = "${ dataPath } tidak boleh lebih pendek dari ${ limit } karakter.", l = "${ dataPath } tidak boleh memiliki kurang dari ${ limit } item.", m = "${ dataPath } tidak boleh memiliki lebih dari ${ limit } item.", k = "${ dataPath } harus memiliki setidaknya satu karakter non-spasi.", u = "${ dataPath } tidak boleh memiliki ${ additionalProperty }.", g = "${ dataPath } harus sama dengan salah satu nilai yang diizinkan ini: ${ allowedValues }.", p = "${ dataPath } harus cocok dengan skema salah satu dari ini: ${ schemaOptions }.", h = "Scatter plot dengan simbol proporsional tidak didukung. Ukuran simbol default sudah diterapkan.", c = "Gagal membaca input data.", $ = "Operasi queryBins tidak didukung pada layanan ini.", y = "Jenis seri yang diharapkan pada indeks ${ seriesIndex } adalah '${ expectedType }' tetapi '${ receivedType }' diterima sebagai gantinya", b = "atau", P = {
  defaultError: a,
  negativeValueInDataForLogTransformation: i,
  negativeValueInDataForSqrtTransformation: e,
  layerLoadFailure: t,
  layerFieldsNotFound: n,
  duplicateSeriesID: r,
  nonNumericAggregation: s,
  requiredProperty: d,
  minLength: o,
  minItems: l,
  maxItems: m,
  whiteSpacePattern: k,
  additionalProperty: u,
  enumValues: g,
  anyOfValues: p,
  bubbleChartValidateMsg: h,
  queryError: c,
  queryBinsNotSupported: $,
  invalidSeriesType: y,
  or: b
};
export {
  u as additionalProperty,
  p as anyOfValues,
  h as bubbleChartValidateMsg,
  P as default,
  a as defaultError,
  r as duplicateSeriesID,
  g as enumValues,
  y as invalidSeriesType,
  n as layerFieldsNotFound,
  t as layerLoadFailure,
  m as maxItems,
  l as minItems,
  o as minLength,
  i as negativeValueInDataForLogTransformation,
  e as negativeValueInDataForSqrtTransformation,
  s as nonNumericAggregation,
  b as or,
  $ as queryBinsNotSupported,
  c as queryError,
  d as requiredProperty,
  k as whiteSpacePattern
};
