// @arcgis/charts-components/dist/chunks/messages.sl4.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "Predvajaj", o = "Ustavi", a = "Legenda", i = "Nalaganje", r = "Domov", n = "Grafikon", t = "Serije", s = "Izvozi", p = "Slika", l = "Podatki", d = "Natisni", m = {
  "Zoom Out": "Pomanjšaj",
  Play: e,
  Stop: o,
  Legend: a,
  "Press ENTER to toggle": "Pritisnite ENTER za preklop",
  Loading: i,
  Home: r,
  Chart: n,
  "Serial chart": "Serijski grafikon",
  "X/Y chart": "Grafikon X/Y",
  "Pie chart": "Tortni grafikon",
  "Gauge chart": "Grafikon števca",
  "Radar chart": "Grafikon radarja",
  "Sankey diagram": "Diagram Sankey",
  "Flow diagram": "Diagram poteka",
  "Chord diagram": "Kabelski grafikon",
  "TreeMap chart": "Drevesni grafikon",
  "Sliced chart": "Razdeljen grafikon",
  Series: t,
  "Candlestick Series": "Serija svečnikov",
  "OHLC Series": "Serija OHLC",
  "Column Series": "Serija stolpcev",
  "Line Series": "Serija linij",
  "Pie Slice Series": "Serija tortnih rezin",
  "Funnel Series": "Serija lijakov",
  "Pyramid Series": "Serija piramid",
  "X/Y Series": "Niz X/Y",
  Map: "Karta",
  "Press ENTER to zoom in": "Pritisnite ENTER za povečavo",
  "Press ENTER to zoom out": "Pritisnite ENTER za pomanjšavo",
  "Use arrow keys to zoom in and out": "Za povečavo in pomanjšavo uporabite smerne tipke",
  "Use plus and minus keys on your keyboard to zoom in and out": "Za povečavo in pomanjšavo uporabite tipki plus in minus",
  Export: s,
  Image: p,
  Data: l,
  Print: d,
  "Press ENTER to open": "Pritisnite ENTER za odpiranje",
  "Press ENTER to print.": "Pritisnite ENTER za tiskanje.",
  "Press ENTER to export as %1.": "Pritisnite ENTER za izvoz kot %1.",
  "(Press ESC to close this message)": "(Pritisnite ESC za zapiranje tega sporočila)",
  "Image Export Complete": "Izvoz slike je dokončan",
  "Export operation took longer than expected. Something might have gone wrong.": "Operacija izvoza traja dlje od pričakovanj. Nekaj je morda šlo narobe.",
  "Saved from": "Shranjeno iz",
  "Use TAB to select grip buttons or left and right arrows to change selection": "Za izbiro gumbov za prijem uporabite TAB, za spremembo izbire pa puščico levo in desno",
  "Use left and right arrows to move selection": "Levo in desno puščico uporabite za pomik izbire",
  "Use left and right arrows to move left selection": "Levo in desno puščico uporabite za pomik leve izbire",
  "Use left and right arrows to move right selection": "Levo in desno puščico uporabite za pomik desne izbire",
  "Use TAB select grip buttons or up and down arrows to change selection": "Za izbiro gumbov za prijem uporabite TAB, za spremembo izbire pa puščico navzgor ali navzdol",
  "Use up and down arrows to move selection": "Puščico navzgor in navzdol uporabite za pomik izbire",
  "Use up and down arrows to move lower selection": "Puščico navzgor in navzdol uporabite za pomik spodnje izbire",
  "Use up and down arrows to move upper selection": "Puščico navzgor in navzdol uporabite za pomik zgornje izbire",
  "From %1 to %2": "Od %1 do %2",
  "From %1": "Od %1",
  "To %1": "Do %1",
  "No parser available for file: %1": "Noben razčlenjevalnik ni na voljo ta datoteko: %1",
  "Error parsing file: %1": "Napaka pri razčlenjevanju datoteke: %1",
  "Unable to load file: %1": "Datoteke ni mogoče naložiti: %1",
  "Invalid date": "Neveljaven datum"
};
export {
  n as Chart,
  l as Data,
  s as Export,
  r as Home,
  p as Image,
  a as Legend,
  i as Loading,
  e as Play,
  d as Print,
  t as Series,
  o as Stop,
  m as default
};
