// @arcgis/charts-components/dist/chunks/messages.es2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = { chart: "Gráfico", barChart: "Gráfico de barras", lineChart: "Gráfico de líneas", pieChart: "Gráfico circular", histogram: "Histograma", scatterplot: "Gráfico de dispersión", boxPlot: "Diagrama de caja" }, a = { mean: "Valor medio", median: "Mediana", sum: "Suma", discretePercentile: "Percentil discreto", minimum: "Mínimo", maximum: "Máximo", variance: "Varianza", count: "Recuento", aggregation: "${ statistics } de ${ fieldName }", noAggregation: "Sin agregación" }, i = "Gráfico", t = "Recuento de series", o = "Eje X", r = "No disponible", n = { start: "Alinear con el primer punto de datos", end: "Alinear con el último punto de datos" }, s = {
  chartType: e,
  statistics: a,
  defaultTitle: i,
  countSeries: t,
  xAxis: o,
  notAvailable: r,
  timeAggregationTypes: n
};
export {
  e as chartType,
  t as countSeries,
  s as default,
  i as defaultTitle,
  r as notAvailable,
  a as statistics,
  n as timeAggregationTypes,
  o as xAxis
};
