// @arcgis/charts-components/dist/chunks/messages.it2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const a = { chart: "Grafico", barChart: "Grafico a barre", lineChart: "Grafico a linee", pieChart: "Grafico a torta", histogram: "Istogramma", scatterplot: "Grafico a dispersione", boxPlot: "Box plot" }, i = { mean: "Media", median: "Mediana", sum: "Somma", discretePercentile: "Percentile discreto", minimum: "Minimo", maximum: "Massimo", variance: "Varianza", count: "Conteggio", aggregation: "${ statistics } di ${ fieldName }", noAggregation: "Nessuna aggregazione" }, t = "Grafico", e = "Serie Conteggio", o = "Asse X", n = "Non disponibile", s = { start: "Snap al primo punto dati", end: "Snap al punto dati finale" }, r = {
  chartType: a,
  statistics: i,
  defaultTitle: t,
  countSeries: e,
  xAxis: o,
  notAvailable: n,
  timeAggregationTypes: s
};
export {
  a as chartType,
  e as countSeries,
  r as default,
  t as defaultTitle,
  n as notAvailable,
  i as statistics,
  s as timeAggregationTypes,
  o as xAxis
};
