// @arcgis/charts-components/dist/chunks/messages.vi4.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const n = "Phát", t = "Dừng", e = "Chú giải", o = "Đang tải", i = "Trang chủ", h = "Biểu đồ", r = "Chuỗi", a = "Xuất", s = "Hình ảnh", c = "Dữ liệu", g = "In", u = {
  "Zoom Out": "Thu nhỏ",
  Play: n,
  Stop: t,
  Legend: e,
  "Press ENTER to toggle": "Nhấn ENTER để chuyển đổi",
  Loading: o,
  Home: i,
  Chart: h,
  "Serial chart": "Biểu đồ chuỗi",
  "X/Y chart": "Biểu đồ X/Y",
  "Pie chart": "Biểu đồ hình tròn",
  "Gauge chart": "Biểu đồ đo lường",
  "Radar chart": "Biểu đồ radar",
  "Sankey diagram": "Biểu đồ sankey",
  "Flow diagram": "Lưu đồ",
  "Chord diagram": "Biểu đồ hợp âm",
  "TreeMap chart": "Biểu đồ bản đồ cây",
  "Sliced chart": "Biểu đồ lát cắt",
  Series: r,
  "Candlestick Series": "Chuỗi Hình nến",
  "OHLC Series": "Chuỗi OHLC",
  "Column Series": "Chuỗi Cột",
  "Line Series": "Chuỗi Đường",
  "Pie Slice Series": "Chuỗi Lát cắt Tròn",
  "Funnel Series": "Chuỗi Phễu",
  "Pyramid Series": "Chuỗi Hình chóp",
  "X/Y Series": "Chuỗi X/Y",
  Map: "Bản đồ",
  "Press ENTER to zoom in": "Nhấn ENTER để phóng to",
  "Press ENTER to zoom out": "Nhấn ENTER để thu nhỏ",
  "Use arrow keys to zoom in and out": "Sử dụng phím mũi tên để thu phóng",
  "Use plus and minus keys on your keyboard to zoom in and out": "Sử dụng phím cộng và trừ trên bàn phím để thu phóng",
  Export: a,
  Image: s,
  Data: c,
  Print: g,
  "Press ENTER to open": "Nhấn ENTER để mở",
  "Press ENTER to print.": "Nhấn ENTER để in.",
  "Press ENTER to export as %1.": "Nhấn ENTER để xuất dưới dạng %1.",
  "(Press ESC to close this message)": "(Nhấn ESC để đóng thông báo này)",
  "Image Export Complete": "Hoàn tất Xuất Hình ảnh",
  "Export operation took longer than expected. Something might have gone wrong.": "Hoạt động xuất mất nhiều thời gian hơn dự kiến. Có thể đã xảy ra sự cố.",
  "Saved from": "Đã lưu từ",
  "Use TAB to select grip buttons or left and right arrows to change selection": "Sử dụng phím TAB để chọn nút cầm hoặc mũi tên trái và phải để thay đổi lựa chọn",
  "Use left and right arrows to move selection": "Sử dụng mũi tên trái và phải để di chuyển lựa chọn",
  "Use left and right arrows to move left selection": "Sử dụng mũi tên trái và phải để di chuyển lựa chọn sang trái",
  "Use left and right arrows to move right selection": "Sử dụng mũi tên trái và phải để di chuyển lựa chọn sang phải",
  "Use TAB select grip buttons or up and down arrows to change selection": "Sử dụng phím TAB để chọn nút cầm hoặc mũi tên lên và xuống để thay đổi lựa chọn",
  "Use up and down arrows to move selection": "Sử dụng mũi tên lên và xuống để di chuyển lựa chọn",
  "Use up and down arrows to move lower selection": "Sử dụng mũi tên lên và xuống để di chuyển lựa chọn xuống",
  "Use up and down arrows to move upper selection": "Sử dụng mũi tên lên và xuống để di chuyển lựa chọn lên",
  "From %1 to %2": "Từ %1 đến %2",
  "From %1": "Từ %1",
  "To %1": "Đến %1",
  "No parser available for file: %1": "Không có sẵn trình phân tích cú pháp cho tệp: %1",
  "Error parsing file: %1": "Lỗi khi phân tích cú pháp tệp: %1",
  "Unable to load file: %1": "Không thể tải tệp: %1",
  "Invalid date": "Ngày không hợp lệ"
};
export {
  h as Chart,
  c as Data,
  a as Export,
  i as Home,
  s as Image,
  e as Legend,
  o as Loading,
  n as Play,
  g as Print,
  r as Series,
  t as Stop,
  u as default
};
