// @arcgis/charts-components/dist/chunks/messages.pt-PT2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const a = { chart: "Gráfico", barChart: "Gráfico de barras", lineChart: "Gráfico de linhas", pieChart: "Gráfico circular", histogram: "Histograma", scatterplot: "Gráfico de dispersão", boxPlot: "Gráfico de caixas" }, o = { mean: "Média", median: "Mediana", sum: "Soma", discretePercentile: "Percentil discreto", minimum: "Mínimo", maximum: "Máximo", variance: "Variação", count: "Contagem", aggregation: "${ statistics } de ${ fieldName }", noAggregation: "Sem agregação" }, t = "Gráfico", e = "Séries de Contagem", i = "Eixo X", r = "Não disponível", s = { start: "Ajustar ao primeiro ponto de dados", end: "Ajustar ao último ponto de dados" }, n = {
  chartType: a,
  statistics: o,
  defaultTitle: t,
  countSeries: e,
  xAxis: i,
  notAvailable: r,
  timeAggregationTypes: s
};
export {
  a as chartType,
  e as countSeries,
  n as default,
  t as defaultTitle,
  r as notAvailable,
  o as statistics,
  s as timeAggregationTypes,
  i as xAxis
};
