// @arcgis/charts-components/dist/chunks/messages.pl.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "Wystąpił błąd podczas wczytywania diagramu.", a = "Nie można zastosować transformacji Logarytm do wartości ujemnych lub równych zero.", n = "Nie można zastosować transformacji Pierwiastek kwadratowy do wartości ujemnych.", t = "Wystąpił błąd podczas wczytywania warstwy. URL = ${ url }. Identyfikator elementu portalu = ${ portalItemId }.", i = "Zdefiniowane w konfiguracji pola [${ fieldNames }] nie zostały znalezione w warstwie.", o = "Nazwa ${ dataPath } musi być unikalna. Seria o nazwie ${ seriesName } ma identyfikator (${ seriesID }), który jest już używany przez inną serię.", s = "Element ${ dataPath } nie może wykonywać agregacji innej niż zliczenie na polu innym niż liczbowe.", r = "W elemencie ${ dataPath } brak właściwości o nazwie ${ missingProperty }.", m = "Element ${ dataPath } nie może być krótszy niż następująca liczba znaków: ${ limit }.", c = "Element ${ dataPath } nie może mieć mniej elementów niż ${ limit }.", y = "Element ${ dataPath } nie może mieć więcej elementów niż ${ limit }.", l = "Element ${ dataPath } musi zawierać co najmniej jeden znak inny niż biały znak.", d = "Element ${ dataPath } nie może zawierać ${ additionalProperty }.", w = "Element ${ dataPath } musi mieć wartość równą jednej z tych dozwolonych wartości: ${ allowedValues }.", u = "Element ${ dataPath } musi być zgodny z jednym z tych schematów: ${ schemaOptions }.", z = "Diagramy punktowe z symbolami proporcjonalnymi nie są obsługiwane. Zastosowano domyślną wielkość symbolu.", p = "Nie powiódł się odczyt danych wejściowych.", $ = "Operacja queryBins nie jest obsługiwana w tej usłudze.", h = "Oczekiwany typ serii z indeksem ${ seriesIndex } to '${ expectedType }', ale zamiast niego otrzymano '${ receivedType }'", j = "lub", g = {
  defaultError: e,
  negativeValueInDataForLogTransformation: a,
  negativeValueInDataForSqrtTransformation: n,
  layerLoadFailure: t,
  layerFieldsNotFound: i,
  duplicateSeriesID: o,
  nonNumericAggregation: s,
  requiredProperty: r,
  minLength: m,
  minItems: c,
  maxItems: y,
  whiteSpacePattern: l,
  additionalProperty: d,
  enumValues: w,
  anyOfValues: u,
  bubbleChartValidateMsg: z,
  queryError: p,
  queryBinsNotSupported: $,
  invalidSeriesType: h,
  or: j
};
export {
  d as additionalProperty,
  u as anyOfValues,
  z as bubbleChartValidateMsg,
  g as default,
  e as defaultError,
  o as duplicateSeriesID,
  w as enumValues,
  h as invalidSeriesType,
  i as layerFieldsNotFound,
  t as layerLoadFailure,
  y as maxItems,
  c as minItems,
  m as minLength,
  a as negativeValueInDataForLogTransformation,
  n as negativeValueInDataForSqrtTransformation,
  s as nonNumericAggregation,
  j as or,
  $ as queryBinsNotSupported,
  p as queryError,
  r as requiredProperty,
  l as whiteSpacePattern
};
