// @arcgis/charts-components/dist/chunks/messages.sv.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const t = "Ett fel uppstod när diagrammet lästes in.", e = "Det går inte att tillämpa loggomvandling på negativa värden eller nollvärden.", a = "Det går inte att tillämpa kvadratrotomvandling på negativa värden.", n = "Ett fel uppstod när lagret lästes in. URL = ${ url }. Portalobjekt-ID = ${ portalItemId }.", r = "Fälten [${ fieldNames }] som är definierade i konfigurationen hittades inte i lagret.", s = "${ dataPath } måste vara unik. Serien som heter ${ seriesName } har ett ID (${ seriesID }) som redan används av en annan serie.", i = "${ dataPath } kan inte utföra icke-tidsmässig aggregering på ett icke-numeriskt fält.", o = "${ dataPath } saknar egenskapen ${ missingProperty }.", l = "${ dataPath } får inte vara kortare än ${ limit } tecken.", d = "${ dataPath } får inte ha färre än ${ limit } objekt.", m = "${ dataPath } får inte ha fler än ${ limit } objekt.", g = "${ dataPath } måste ha minst ett tecken som inte är ett blanksteg.", c = "${ dataPath } får inte innehålla ${ additionalProperty }.", p = "${ dataPath } måste vara lika med ett av dessa tillåtna värden: ${ allowedValues }.", u = "${ dataPath } måste matcha schemat för någon av dessa: ${ schemaOptions }.", h = "Punktdiagram med proportionerliga symboler stöds inte. Standardstorleken för symboler har tillämpats.", $ = "Det gick inte att läsa indata.", v = "queryBins-åtgärden stöds inte för den här tjänsten.", f = "Förväntad serietyp vid index ${ seriesIndex } är '${ expectedType }' men '${ receivedType }' togs emot istället", k = "eller", P = {
  defaultError: t,
  negativeValueInDataForLogTransformation: e,
  negativeValueInDataForSqrtTransformation: a,
  layerLoadFailure: n,
  layerFieldsNotFound: r,
  duplicateSeriesID: s,
  nonNumericAggregation: i,
  requiredProperty: o,
  minLength: l,
  minItems: d,
  maxItems: m,
  whiteSpacePattern: g,
  additionalProperty: c,
  enumValues: p,
  anyOfValues: u,
  bubbleChartValidateMsg: h,
  queryError: $,
  queryBinsNotSupported: v,
  invalidSeriesType: f,
  or: k
};
export {
  c as additionalProperty,
  u as anyOfValues,
  h as bubbleChartValidateMsg,
  P as default,
  t as defaultError,
  s as duplicateSeriesID,
  p as enumValues,
  f as invalidSeriesType,
  r as layerFieldsNotFound,
  n as layerLoadFailure,
  m as maxItems,
  d as minItems,
  l as minLength,
  e as negativeValueInDataForLogTransformation,
  a as negativeValueInDataForSqrtTransformation,
  i as nonNumericAggregation,
  k as or,
  v as queryBinsNotSupported,
  $ as queryError,
  o as requiredProperty,
  g as whiteSpacePattern
};
