// @arcgis/charts-components/dist/chunks/messages.th2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const t = { chart: "แผนภูมิ", barChart: "แผนภูมิแท่ง", lineChart: "แผนภูมิเส้น", pieChart: "แผนภูมิทรงกลม", histogram: "ฮีสโตแกรม", scatterplot: "พล็อตการกระจาย", boxPlot: "พล็อตแบบกล่อง" }, e = { mean: "ค่าเฉลี่ย", median: "ค่ามัธยฐาน", sum: "รวม", discretePercentile: "เปอร์เซนไทล์แบบไม่ต่อเนื่อง", minimum: "น้อยสุด", maximum: "สูงสุด", variance: "ความแปรปรวน", count: "นับ", aggregation: "${ statistics } จาก ${ fieldName }", noAggregation: "ไม่มีการรวม" }, a = "แผนภูมิ", s = "นับซีรีส์", i = "(แกน x)", n = "ไม่สามารถใช้ได้", o = { start: "สแน็ปเข้ากับจุดข้อมูลแรก", end: "สแน็ปเข้ากับจุดข้อมูลสุดท้าย" }, c = {
  chartType: t,
  statistics: e,
  defaultTitle: a,
  countSeries: s,
  xAxis: i,
  notAvailable: n,
  timeAggregationTypes: o
};
export {
  t as chartType,
  s as countSeries,
  c as default,
  a as defaultTitle,
  n as notAvailable,
  e as statistics,
  o as timeAggregationTypes,
  i as xAxis
};
