// @arcgis/charts-components/dist/chunks/messages.tr.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const a = "Çizge yüklenirken bir hata oluştu.", e = "Negatif veya sıfır değerlere işlem kaydı dönüştürme uygulanamaz.", t = "Negatif değerlere kare kök dönüştürme uygulanamaz.", r = "Katman yüklenirken bir hata oluştu. URL = ${ url }. Portal ögesi kimliği = ${ portalItemId }.", n = "Yapılandırmada tanımlanan [${ fieldNames }] alanları katmanda bulunamadı.", i = "${ dataPath } benzersiz olmalıdır. ${ seriesName } adlı serinin başka bir seri tarafından zaten kullanılan bir kimliği (${ seriesID }) var.", l = "${ dataPath } sayısal olmayan alanda sayım dışı kümeleme gerçekleştiremez.", s = "${ dataPath }'de ${ missingProperty } adlı özellik eksik.", d = "${ dataPath }, ${ limit } karakterden daha kısa olmamalıdır.", o = "${ dataPath }, ${ limit }'den az ögeye sahip olmamalıdır.", m = "${ dataPath }, ${ limit }'den fazla ögeye sahip olmamalıdır.", u = "${ dataPath }, en az bir adet boşluk olmayan karaktere sahip olmalıdır.", y = "${ dataPath }'de ${ additionalProperty } olmamalıdır.", k = "${ dataPath }, izin verilen şu değerlerden birine eşit olmalıdır: ${ allowedValues }.", c = "${ dataPath }, şunlardan birinin şemasıyla eşleşmelidir: ${ schemaOptions }.", $ = "Orantısal sembollere sahip dağılım grafikleri desteklenmez. Varsayılan sembol boyutu uygulandı.", g = "Girdi verileri okunamadı.", h = "queryBins işlemi bu serviste desteklenmiyor.", p = "${ seriesIndex } dizininde beklenen seri tipi '${ expectedType }' ancak '${ receivedType }' alındı", b = "veya", P = {
  defaultError: a,
  negativeValueInDataForLogTransformation: e,
  negativeValueInDataForSqrtTransformation: t,
  layerLoadFailure: r,
  layerFieldsNotFound: n,
  duplicateSeriesID: i,
  nonNumericAggregation: l,
  requiredProperty: s,
  minLength: d,
  minItems: o,
  maxItems: m,
  whiteSpacePattern: u,
  additionalProperty: y,
  enumValues: k,
  anyOfValues: c,
  bubbleChartValidateMsg: $,
  queryError: g,
  queryBinsNotSupported: h,
  invalidSeriesType: p,
  or: b
};
export {
  y as additionalProperty,
  c as anyOfValues,
  $ as bubbleChartValidateMsg,
  P as default,
  a as defaultError,
  i as duplicateSeriesID,
  k as enumValues,
  p as invalidSeriesType,
  n as layerFieldsNotFound,
  r as layerLoadFailure,
  m as maxItems,
  o as minItems,
  d as minLength,
  e as negativeValueInDataForLogTransformation,
  t as negativeValueInDataForSqrtTransformation,
  l as nonNumericAggregation,
  b as or,
  h as queryBinsNotSupported,
  g as queryError,
  s as requiredProperty,
  u as whiteSpacePattern
};
