// @arcgis/charts-components/dist/chunks/messages.sk2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const t = { chart: "Graf", barChart: "Pruhový graf", lineChart: "Čiarový graf", pieChart: "Koláčový graf", histogram: "Histogram", scatterplot: "Korelačný graf", boxPlot: "Krabicový graf" }, e = { mean: "Priemer", median: "Medián", sum: "Súčet", discretePercentile: "Diskrétny percentil", minimum: "Minimum", maximum: "Maximum", variance: "Rozptyl", count: "Počet", aggregation: "${ statistics } z ${ fieldName }", noAggregation: "Bez agregácie" }, i = "Graf", a = "Počet sérií", r = "Os X", o = "Nie je k dispozícii", s = { start: "Prichytiť k prvému dátovému bodu", end: "Prichytiť k poslednému dátovému bodu" }, n = {
  chartType: t,
  statistics: e,
  defaultTitle: i,
  countSeries: a,
  xAxis: r,
  notAvailable: o,
  timeAggregationTypes: s
};
export {
  t as chartType,
  a as countSeries,
  n as default,
  i as defaultTitle,
  o as notAvailable,
  e as statistics,
  s as timeAggregationTypes,
  r as xAxis
};
