// @arcgis/charts-components/dist/chunks/messages.id4.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const a = "Putar", n = "Stop", e = "Legenda", i = "Memuat", t = "Rumah", r = "Diagram", o = "Rangkaian", s = "Ekspor", k = "Citra", m = "Data", u = "Cetak", l = {
  "Zoom Out": "Perkecil",
  Play: a,
  Stop: n,
  Legend: e,
  "Press ENTER to toggle": "Tekan ENTER untuk beralih",
  Loading: i,
  Home: t,
  Chart: r,
  "Serial chart": "Diagram rangkaian",
  "X/Y chart": "Diagram X/Y",
  "Pie chart": "Diagram pai",
  "Gauge chart": "Diagram pengukur",
  "Radar chart": "Diagram radar",
  "Sankey diagram": "Diagram sankey",
  "Flow diagram": "Diagram alur",
  "Chord diagram": "Diagram chord",
  "TreeMap chart": "Diagram TreeMap",
  "Sliced chart": "Diagram irisan",
  Series: o,
  "Candlestick Series": "Rangkaian Candlestick",
  "OHLC Series": "Rangkaian OHLC",
  "Column Series": "Rangkaian Kolom",
  "Line Series": "Rangkaian Garis",
  "Pie Slice Series": "Rangkaian Potongan Pai",
  "Funnel Series": "Rangkaian Corong",
  "Pyramid Series": "Rangkaian Piramida",
  "X/Y Series": "Rangkaian X/Y",
  Map: "Peta",
  "Press ENTER to zoom in": "Tekan ENTER untuk memperbesar",
  "Press ENTER to zoom out": "Tekan ENTER untuk memperkecil",
  "Use arrow keys to zoom in and out": "Gunakan tombol panah untuk memperbesar dan memperkecil",
  "Use plus and minus keys on your keyboard to zoom in and out": "Gunakan tombol plus dan minus pada keyboard untuk memperbesar dan memperkecil",
  Export: s,
  Image: k,
  Data: m,
  Print: u,
  "Press ENTER to open": "Tekan ENTER untuk membuka",
  "Press ENTER to print.": "Tekan ENTER untuk mencetak",
  "Press ENTER to export as %1.": "Tekan ENTER untuk mengekspor sebagai %1.",
  "(Press ESC to close this message)": "(Tekan ESC untuk menutup pesan ini)",
  "Image Export Complete": "Ekspor Gambar Selesai",
  "Export operation took longer than expected. Something might have gone wrong.": "Operasi ekspor butuh waktu lebih lama dari yang diperkirakan Mungkin ada yang tidak beres.",
  "Saved from": "Disimpan dari",
  "Use TAB to select grip buttons or left and right arrows to change selection": "Gunakan tombol TAB untuk memilih tombol pegangan atau panah kiri dan kanan untuk mengubah pilihan",
  "Use left and right arrows to move selection": "Gunakan panah kiri dan kanan untuk memindahkan pilihan",
  "Use left and right arrows to move left selection": "Gunakan panah kiri dan kanan untuk memindahkan pilihan ke kiri",
  "Use left and right arrows to move right selection": "Gunakan panah kiri dan kanan untuk memindahkan pilihan ke kanan",
  "Use TAB select grip buttons or up and down arrows to change selection": "Gunakan tombol TAB untuk memilih tombol pegangan atau panah atas dan bawah untuk mengubah pilihan",
  "Use up and down arrows to move selection": "Gunakan panah atas dan bawah untuk memindahkan pilihan",
  "Use up and down arrows to move lower selection": "Gunakan panah atas dan bawah untuk memindahkan pilihan ke bawah",
  "Use up and down arrows to move upper selection": "Gunakan panah atas dan bawah untuk memindahkan pilihan ke atas",
  "From %1 to %2": "Dari %1 hingga %2",
  "From %1": "Dari %1",
  "To %1": "Hingga %1",
  "No parser available for file: %1": "Tidak tersedia pengurai untuk file: %1",
  "Error parsing file: %1": "Kesalahan saat mengurai file: %1",
  "Unable to load file: %1": "Tidak dapat memuat file: %1",
  "Invalid date": "Tanggal tidak valid"
};
export {
  r as Chart,
  m as Data,
  s as Export,
  t as Home,
  k as Image,
  e as Legend,
  i as Loading,
  a as Play,
  u as Print,
  o as Series,
  n as Stop,
  l as default
};
