// @arcgis/charts-components/dist/chunks/messages.fr2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = { chart: "Diagramme", barChart: "Diagramme en barres", lineChart: "Diagramme linéaire", pieChart: "Diagramme à secteurs", histogram: "Histogramme", scatterplot: "Nuage de points", boxPlot: "Boîte à moustaches" }, a = { mean: "Moyenne", median: "Médiane", sum: "Somme", discretePercentile: "Centile discret", minimum: "Minimum", maximum: "Maximum", variance: "Variance", count: "Total", aggregation: "${ statistics } de ${ fieldName }", noAggregation: "Pas d’agrégation" }, t = "Diagramme", i = "Série de nombres", r = "Axe X", n = "Non disponible", s = { start: "Capturer sur le premier point de données", end: "Capturer sur le dernier point de données" }, o = {
  chartType: e,
  statistics: a,
  defaultTitle: t,
  countSeries: i,
  xAxis: r,
  notAvailable: n,
  timeAggregationTypes: s
};
export {
  e as chartType,
  i as countSeries,
  o as default,
  t as defaultTitle,
  n as notAvailable,
  a as statistics,
  s as timeAggregationTypes,
  r as xAxis
};
