// @arcgis/charts-components/dist/chunks/messages.ru.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const t = "При загрузке диаграммы произошла ошибка.", a = "Невозможно применить Логарифмическое преобразование к отрицательным или нулевым значениям.", e = "Невозможно применить преобразование по методу квадратного корня к отрицательным значениям.", s = "При загрузке слоя произошла ошибка. URL = ${ url }. ID элемента портала = ${ portalItemId }.", o = "Поля [${ fieldNames }], заданные в конфигурации, не найдены в слое.", n = "${ dataPath } должен быть уникальным. У серии с именем ${ seriesName } есть ID (${ seriesID }), который уже используется другой серией.", r = "${ dataPath } не должен выполнять не количественное агрегирование для не числового поля.", i = "В ${ dataPath } отсутствует свойство с именем ${ missingProperty }.", d = "${ dataPath } не должен быть короче, чем ${ limit } символов.", c = "В ${ dataPath } должно быть не менее ${ limit } элементов.", l = "В ${ dataPath } не должно быть более ${ limit } элементов.", $ = "${ dataPath } должен содержать, как минимум, один символ, не являющейся пробелом.", u = "${ dataPath } не должны иметь ${ additionalProperty }", m = "${ dataPath } должен быть равен одному из этих допустимых значений: ${ allowedValues }.", p = "${ dataPath } должен совпадать со схемой одного из следующих элементов: ${ schemaOptions }.", P = "Диаграммы рассеяния с пропорциональными символами не поддерживаются. Будет использован размер символа по умолчанию.", h = "Ошибка чтения входных данных.", y = "Операция queryBins не поддерживается этим сервисом.", I = "Ожидаемый тип рядов в индексе ${ seriesIndex } - '${ expectedType }' но вместо этого было получено '${ receivedType }'", g = "или", f = {
  defaultError: t,
  negativeValueInDataForLogTransformation: a,
  negativeValueInDataForSqrtTransformation: e,
  layerLoadFailure: s,
  layerFieldsNotFound: o,
  duplicateSeriesID: n,
  nonNumericAggregation: r,
  requiredProperty: i,
  minLength: d,
  minItems: c,
  maxItems: l,
  whiteSpacePattern: $,
  additionalProperty: u,
  enumValues: m,
  anyOfValues: p,
  bubbleChartValidateMsg: P,
  queryError: h,
  queryBinsNotSupported: y,
  invalidSeriesType: I,
  or: g
};
export {
  u as additionalProperty,
  p as anyOfValues,
  P as bubbleChartValidateMsg,
  f as default,
  t as defaultError,
  n as duplicateSeriesID,
  m as enumValues,
  I as invalidSeriesType,
  o as layerFieldsNotFound,
  s as layerLoadFailure,
  l as maxItems,
  c as minItems,
  d as minLength,
  a as negativeValueInDataForLogTransformation,
  e as negativeValueInDataForSqrtTransformation,
  r as nonNumericAggregation,
  g as or,
  y as queryBinsNotSupported,
  h as queryError,
  i as requiredProperty,
  $ as whiteSpacePattern
};
