// @arcgis/charts-components/dist/chunks/messages.lv2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const a = { chart: "Diagramma", barChart: "Joslu diagramma", lineChart: "Līniju diagramma", pieChart: "Sektoru diagramma", histogram: "Histogramma", scatterplot: "Izkliedes diagramma", boxPlot: "Kastveida diagramma" }, i = { mean: "Vidējais", median: "Mediāna", sum: "Summa", discretePercentile: "Diskrētā procentīle", minimum: "Minimums", maximum: "Maksimums", variance: "Novirze", count: "Skaits", aggregation: "${ statistics } no ${ fieldName }", noAggregation: "Bez apkopošanas" }, t = "Diagramma", e = "Sēriju uzskaite", s = "X ass", m = "Nav pieejams", n = { start: "Pielipināt pie pirmā datu punkta", end: "Pielipināt pie pēdējā datu punkta" }, r = {
  chartType: a,
  statistics: i,
  defaultTitle: t,
  countSeries: e,
  xAxis: s,
  notAvailable: m,
  timeAggregationTypes: n
};
export {
  a as chartType,
  e as countSeries,
  r as default,
  t as defaultTitle,
  m as notAvailable,
  i as statistics,
  n as timeAggregationTypes,
  s as xAxis
};
