// @arcgis/charts-components/dist/chunks/messages.ru2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const t = { chart: "Диаграмма", barChart: "Линейчатая диаграмма", lineChart: "Линейная диаграмма", pieChart: "Круговая диаграмма", histogram: "Гистограмма", scatterplot: "Диаграмма рассеяния", boxPlot: "Ящичковая диаграмма" }, e = { mean: "Среднее", median: "Медиана", sum: "Сумма", discretePercentile: "Дискретный процентиль", minimum: "Минимум", maximum: "Максимум", variance: "Дисперсия", count: "Количество", aggregation: "${ statistics } из ${ fieldName }", noAggregation: "Без агрегирования" }, a = "Диаграмма", s = "Серии чисел", i = "Ось X", n = "Недоступно", o = { start: "Замкнуть на первую точку данных", end: "Замкнуть на последнюю точку данных" }, c = {
  chartType: t,
  statistics: e,
  defaultTitle: a,
  countSeries: s,
  xAxis: i,
  notAvailable: n,
  timeAggregationTypes: o
};
export {
  t as chartType,
  s as countSeries,
  c as default,
  a as defaultTitle,
  n as notAvailable,
  e as statistics,
  o as timeAggregationTypes,
  i as xAxis
};
