// @arcgis/charts-components/dist/chunks/messages.bs2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const a = { chart: "Grafikon", barChart: "Trakasti grafikon", lineChart: "Linijski grafikon", pieChart: "Tortni grafikon", histogram: "Histogram", scatterplot: "Pravokutnik raspršivanja", boxPlot: "Dijagram pravokutnika" }, i = { mean: "Prosjek", median: "Srednja vrijednost", sum: "Zbroj", discretePercentile: "Diskretni postotak", minimum: "Minimum", maximum: "Maksimum", variance: "Varijanca", count: "Zbroj", aggregation: "${ statistics } od ${ fieldName }", noAggregation: "Nema agregacije" }, t = "Grafikon", o = "Broj serija", n = "Os x", r = "Nije dostupno", e = { start: "Prilijepi na prvu podatkovnu točku", end: "Prilijepi na posljednju podatkovnu točku" }, s = {
  chartType: a,
  statistics: i,
  defaultTitle: t,
  countSeries: o,
  xAxis: n,
  notAvailable: r,
  timeAggregationTypes: e
};
export {
  a as chartType,
  o as countSeries,
  s as default,
  t as defaultTitle,
  r as notAvailable,
  i as statistics,
  e as timeAggregationTypes,
  n as xAxis
};
