// @arcgis/charts-components/dist/chunks/messages.nl.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "Er is een fout opgetreden bij het laden van de grafiek.", n = "Kan logtransformatie niet toepassen op negatieve of nulwaarden.", t = "Kan transformatie vierkantswortel niet toepassen op negatieve waarden.", a = "Er is een fout opgetreden bij het laden van de laag. URL = ${ url }. Portal item ID = ${ portalItemId }.", i = "De velden [${ fieldNames }] die in de config zijn gedefinieerd, zijn niet aangetroffen in de laag.", o = "${ dataPath } moet uniek zijn. De serie genaamd ${ seriesName } heeft een ID (${ seriesID }) die al wordt gebruikt door een andere serie.", r = "${ dataPath } kan geen niet-telbare aggregatie uitvoeren op een niet-numeriek veld.", s = "${ dataPath } mist een eigenschap met de naam ${ missingProperty }.", d = "${ dataPath } mag niet korter zijn dan ${ limit } tekens.", m = "${ dataPath } mag niet minder dan ${ limit } items hebben.", l = "${ dataPath } mag niet meer dan ${ limit } items hebben.", g = "${ dataPath } moet minstens één non-whitespace teken hebben.", c = "${ dataPath } mag geen ${ additionalProperty } hebben.", p = "${ dataPath } moet gelijk zijn aan een van deze toegelaten waarden: ${ allowedValues }.", u = "${ dataPath } moet overeenkomen met het schema van een van deze: ${ schemaOptions }.", h = "Spreidingsdiagrammen met proportionele symbolen worden niet ondersteund. Standaardsymboolgrootte is toegepast.", $ = "Kan invoergegevens niet lezen.", v = "De bewerking queryBins wordt niet ondersteund in deze service.", b = "Het verwachte serietype bij index ${ seriesIndex } is '${ expectedType }' maar '${ receivedType }' is in plaats daarvan ontvangen", f = "of", y = {
  defaultError: e,
  negativeValueInDataForLogTransformation: n,
  negativeValueInDataForSqrtTransformation: t,
  layerLoadFailure: a,
  layerFieldsNotFound: i,
  duplicateSeriesID: o,
  nonNumericAggregation: r,
  requiredProperty: s,
  minLength: d,
  minItems: m,
  maxItems: l,
  whiteSpacePattern: g,
  additionalProperty: c,
  enumValues: p,
  anyOfValues: u,
  bubbleChartValidateMsg: h,
  queryError: $,
  queryBinsNotSupported: v,
  invalidSeriesType: b,
  or: f
};
export {
  c as additionalProperty,
  u as anyOfValues,
  h as bubbleChartValidateMsg,
  y as default,
  e as defaultError,
  o as duplicateSeriesID,
  p as enumValues,
  b as invalidSeriesType,
  i as layerFieldsNotFound,
  a as layerLoadFailure,
  l as maxItems,
  m as minItems,
  d as minLength,
  n as negativeValueInDataForLogTransformation,
  t as negativeValueInDataForSqrtTransformation,
  r as nonNumericAggregation,
  f as or,
  v as queryBinsNotSupported,
  $ as queryError,
  s as requiredProperty,
  g as whiteSpacePattern
};
