// @arcgis/charts-components/dist/chunks/messages.ko.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const t = "차트를 불러오는 동안 오류가 발생했습니다.", a = "로그 변환을 음수 값 또는 0 값에 적용할 수 없습니다.", e = "제곱근 변환을 음수 값에 적용할 수 없습니다.", o = "레이어를 불러오는 동안 오류가 발생했습니다. URL = ${ url }. 포털 항목 ID = ${ portalItemId }.", s = "구성에 정의된 필드 [${ fieldNames }]를 레이어에서 찾을 수 없습니다.", n = "${ dataPath }은(는) 고유해야 합니다. ${ seriesName }(이)라는 계열에 다른 계열에서 이미 사용된 ID(${ seriesID })가 있습니다.", r = "${ dataPath }은(는) 숫자가 아닌 필드에서 불가산 집계를 수행할 수 없습니다.", i = "${ dataPath }에 ${ missingProperty } 등록정보가 누락되었습니다.", d = "${ dataPath }은(는) ${ limit }자보다 짧으면 안 됩니다.", c = "${ dataPath }에는 ${ limit }개 이상의 항목이 있어야 합니다.", l = "${ dataPath }에는 ${ limit }개 이하의 항목이 있어야 합니다.", $ = "${ dataPath }은(는) 공백 이외의 문자를 1개 이상 포함해야 합니다.", u = "${ dataPath }에는 ${ additionalProperty }이(가) 없어야 합니다.", m = "${ dataPath }은(는) 허용된 값(${ allowedValues }) 중 하나와 같아야 합니다.", p = "${ dataPath }은(는) ${ schemaOptions } 중 하나의 스키마와 일치해야 합니다.", P = "비례 심볼이 있는 산점도는 지원되지 않습니다. 기본 심볼 크기가 적용되었습니다.", h = "입력 데이터를 읽지 못했습니다.", y = "queryBins 작업은 이 서비스에서 지원하지 않습니다.", I = "인덱스 ${ seriesIndex }에 예상되는 시리즈 유형은 '${ expectedType }'인데 '${ receivedType }'이(가) 대신 수신되었습니다.", g = "또는", f = {
  defaultError: t,
  negativeValueInDataForLogTransformation: a,
  negativeValueInDataForSqrtTransformation: e,
  layerLoadFailure: o,
  layerFieldsNotFound: s,
  duplicateSeriesID: n,
  nonNumericAggregation: r,
  requiredProperty: i,
  minLength: d,
  minItems: c,
  maxItems: l,
  whiteSpacePattern: $,
  additionalProperty: u,
  enumValues: m,
  anyOfValues: p,
  bubbleChartValidateMsg: P,
  queryError: h,
  queryBinsNotSupported: y,
  invalidSeriesType: I,
  or: g
};
export {
  u as additionalProperty,
  p as anyOfValues,
  P as bubbleChartValidateMsg,
  f as default,
  t as defaultError,
  n as duplicateSeriesID,
  m as enumValues,
  I as invalidSeriesType,
  s as layerFieldsNotFound,
  o as layerLoadFailure,
  l as maxItems,
  c as minItems,
  d as minLength,
  a as negativeValueInDataForLogTransformation,
  e as negativeValueInDataForSqrtTransformation,
  r as nonNumericAggregation,
  g as or,
  y as queryBinsNotSupported,
  h as queryError,
  i as requiredProperty,
  $ as whiteSpacePattern
};
