// @arcgis/charts-components/dist/chunks/messages.pt-BR2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const a = { chart: "Gráfico", barChart: "Gráfico de barra", lineChart: "Gráfico de linhas", pieChart: "Gráfico de pizza", histogram: "Histograma", scatterplot: "Gráfico de dispersão", boxPlot: "Diagrama de caixa" }, e = { mean: "Médio", median: "Mediana", sum: "Soma", discretePercentile: "Percentil discreto", minimum: "Mínimo", maximum: "Máximo", variance: "Variância", count: "Contagem", aggregation: "${ statistics } de ${ fieldName }", noAggregation: "Sem agregação" }, t = "Gráfico", i = "Séries de contagem", o = "Eixo X", r = "Não Disponível", s = { start: "Ajustar ao primeiro ponto de dados", end: "Ajustar ao útimo ponto de dados" }, n = {
  chartType: a,
  statistics: e,
  defaultTitle: t,
  countSeries: i,
  xAxis: o,
  notAvailable: r,
  timeAggregationTypes: s
};
export {
  a as chartType,
  i as countSeries,
  n as default,
  t as defaultTitle,
  r as notAvailable,
  e as statistics,
  s as timeAggregationTypes,
  o as xAxis
};
