// @arcgis/charts-components/dist/chunks/messages.tr2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const a = { chart: "Çizge", barChart: "Çubuk grafik", lineChart: "Çizgi grafik", pieChart: "Pasta grafik", histogram: "Histogram", scatterplot: "Dağılım grafiği", boxPlot: "Kutu düzeni" }, i = { mean: "Ortalama", median: "Medyan", sum: "Toplam", discretePercentile: "Ayrık yüzdelik", minimum: "Minimum", maximum: "Maksimum", variance: "Varyans", count: "Sayım", aggregation: "${ statistics } / ${ fieldName }", noAggregation: "Kümeleme yok" }, t = "Grafik", e = "Serileri Say", r = "X Ekseni", n = "Kullanılabilir Değil", s = { start: "İlk veri noktasına yerleştir", end: "Son veri noktasına yerleştir" }, o = {
  chartType: a,
  statistics: i,
  defaultTitle: t,
  countSeries: e,
  xAxis: r,
  notAvailable: n,
  timeAggregationTypes: s
};
export {
  a as chartType,
  e as countSeries,
  o as default,
  t as defaultTitle,
  n as notAvailable,
  i as statistics,
  s as timeAggregationTypes,
  r as xAxis
};
