// @arcgis/charts-components/dist/chunks/messages.sr.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const a = "Došlo je do greške prilikom učitavanja grafika.", e = "Nije moguće primeniti transformaciju zapisa na negativne ili nulte vrednosti.", o = "Nije moguće primeniti transformaiju kvadratnog korena na negativne vrednosti.", i = "Došlo je do greške prilikom učitavanja sloja. URL = ${ url }. ID stavke portala = ${ portalItemId }.", n = "Polja [${ fieldNames }] definisana u konfiguraciji nisu pronađena u sloju.", t = "${ dataPath } mora da bude jedinstven. Serija pod nazivom ${ seriesName } ima ID (${ seriesID }) koji se već koristi od strane drugih serija.", r = "${ dataPath } ne može da obavi agregaciju bez brojanja na numeričkom polju.", s = "${ dataPath } nedostaje svojstvo pod nazivom ${ missingProperty }.", d = "${ dataPath } ne sme da bude kraće od ${ limit } karaktera.", m = "${ dataPath } ne sme da ima manje od ${ limit } stavki.", l = "${ dataPath } ne sme da ima više od ${ limit } stavki.", u = "${ dataPath } mora da ima bar jedan karakter koji nije prazan.", j = "${ dataPath } ne sme da ima ${ additionalProperty }.", p = "${ dataPath } mora da bude jednako jednoj od ovih dozvoljenih vrednosti: ${ allowedValues }.", c = "${ dataPath } mora da odgovara šemi jednog od ovih: ${ schemaOptions }.", v = "Nacrti šablona sa proporcionalnim simbolima nisu podržani. Podrazumevana veličina šablona je primenjena.", g = "Neuspešno čitanje ulaznih podataka.", $ = "Operacija queryBins nije podržana u ovom servisu.", k = "Očekivani tip serije za indeks ${ seriesIndex } je '${ expectedType }' ali je '${ receivedType }' dobijeno umesto toga", h = "ili", P = {
  defaultError: a,
  negativeValueInDataForLogTransformation: e,
  negativeValueInDataForSqrtTransformation: o,
  layerLoadFailure: i,
  layerFieldsNotFound: n,
  duplicateSeriesID: t,
  nonNumericAggregation: r,
  requiredProperty: s,
  minLength: d,
  minItems: m,
  maxItems: l,
  whiteSpacePattern: u,
  additionalProperty: j,
  enumValues: p,
  anyOfValues: c,
  bubbleChartValidateMsg: v,
  queryError: g,
  queryBinsNotSupported: $,
  invalidSeriesType: k,
  or: h
};
export {
  j as additionalProperty,
  c as anyOfValues,
  v as bubbleChartValidateMsg,
  P as default,
  a as defaultError,
  t as duplicateSeriesID,
  p as enumValues,
  k as invalidSeriesType,
  n as layerFieldsNotFound,
  i as layerLoadFailure,
  l as maxItems,
  m as minItems,
  d as minLength,
  e as negativeValueInDataForLogTransformation,
  o as negativeValueInDataForSqrtTransformation,
  r as nonNumericAggregation,
  h as or,
  $ as queryBinsNotSupported,
  g as queryError,
  s as requiredProperty,
  u as whiteSpacePattern
};
