// @arcgis/charts-components/dist/chunks/messages.da.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "Der opstod en fejl under indlæsning af diagrammet.", t = "Kan ikke anvende log-transformation på negative værdier eller nulværdier.", a = "Kan ikke anvende kvadratrod-transformation på negative værdier.", n = "Der opstod en fejl under indlæsning af laget. URL = ${ url }. Portalelement-id = ${ portalItemId }.", r = "De felter [${ fieldNames }], der er defineret i konfigurationen, blev ikke fundet i laget.", i = "${ dataPath } skal være entydig. Serien med navnet ${ seriesName } har et id (${ seriesID }), der allerede bruges i en anden serie.", s = "${ dataPath } kan ikke udføre ikke-antal-aggregering på ikke-numeriske felter.", d = "${ dataPath } mangler en egenskab med navnet ${ missingProperty }.", o = "${ dataPath } skal ikke være kortere end ${ limit } tegn.", l = "${ dataPath } må ikke have færre end ${ limit } elementer.", m = "${ dataPath } må ikke have flere end ${ limit } elementer.", k = "${ dataPath } skal have mindst et tegn, der ikke er et mellemrumstegn.", g = "${ dataPath } må ikke have ${ additionalProperty }.", u = "${ dataPath } skal være lig med en af disse tilladte værdier: ${ allowedValues }.", c = "${ dataPath } skal matche skemaet fra en af disse: ${ schemaOptions }.", v = "Punktdiagrammer med proportionale symboler understøttes ikke. Der er blevet anvendt standardstørrelse på symboler.", f = "Kunne ikke læse inputdata.", p = "queryBins-handlingen understøttes ikke i denne tjeneste.", $ = "Den forventede serietype ved indekset ${ seriesIndex } er '${ expectedType }', men '${ receivedType }' blev modtaget i stedet for", h = "eller", y = {
  defaultError: e,
  negativeValueInDataForLogTransformation: t,
  negativeValueInDataForSqrtTransformation: a,
  layerLoadFailure: n,
  layerFieldsNotFound: r,
  duplicateSeriesID: i,
  nonNumericAggregation: s,
  requiredProperty: d,
  minLength: o,
  minItems: l,
  maxItems: m,
  whiteSpacePattern: k,
  additionalProperty: g,
  enumValues: u,
  anyOfValues: c,
  bubbleChartValidateMsg: v,
  queryError: f,
  queryBinsNotSupported: p,
  invalidSeriesType: $,
  or: h
};
export {
  g as additionalProperty,
  c as anyOfValues,
  v as bubbleChartValidateMsg,
  y as default,
  e as defaultError,
  i as duplicateSeriesID,
  u as enumValues,
  $ as invalidSeriesType,
  r as layerFieldsNotFound,
  n as layerLoadFailure,
  m as maxItems,
  l as minItems,
  o as minLength,
  t as negativeValueInDataForLogTransformation,
  a as negativeValueInDataForSqrtTransformation,
  s as nonNumericAggregation,
  h as or,
  p as queryBinsNotSupported,
  f as queryError,
  d as requiredProperty,
  k as whiteSpacePattern
};
