// @arcgis/charts-components/dist/chunks/messages.pl2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const a = { chart: "Diagram", barChart: "Diagram słupkowy grupowy", lineChart: "Diagram liniowy", pieChart: "Diagram kołowy", histogram: "Histogram", scatterplot: "Diagram punktowy", boxPlot: "Diagram pudełkowy" }, i = { mean: "Średnia", median: "Mediana", sum: "Suma", discretePercentile: "Percentyl dyskretny", minimum: "Minimum", maximum: "Maksimum", variance: "Odchylenie", count: "Liczba", aggregation: "${ statistics } z ${ fieldName }", noAggregation: "Bez agregacji" }, t = "Wykres", e = "Zlicz serie", n = "Oś X", o = "Niedostępne", s = { start: "Dociągnij do pierwszego punktu danych", end: "Dociągnij do ostatniego punktu danych" }, r = {
  chartType: a,
  statistics: i,
  defaultTitle: t,
  countSeries: e,
  xAxis: n,
  notAvailable: o,
  timeAggregationTypes: s
};
export {
  a as chartType,
  e as countSeries,
  r as default,
  t as defaultTitle,
  o as notAvailable,
  i as statistics,
  s as timeAggregationTypes,
  n as xAxis
};
