// @arcgis/charts-components/dist/chunks/messages.uk2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const t = { chart: "Діаграма", barChart: "Гістограма", lineChart: "Лінійна діаграма", pieChart: "Кругова діаграма", histogram: "Гістограма", scatterplot: "Діаграма розсіяння", boxPlot: "Стовпчикова діаграма" }, e = { mean: "Середнє", median: "Медіанне", sum: "Сума", discretePercentile: "Дискретний процентиль", minimum: "Мінімум", maximum: "Максимум", variance: "Дисперсія", count: "Лічильник", aggregation: "${ statistics } з ${ fieldName }", noAggregation: "Немає агрегування" }, a = "Діаграма", s = "Кількість серій", i = "Вісь X", n = "Недоступно", o = { start: "Прив'язка до першої точки даних", end: "Прив'язка до останньої точки даних" }, c = {
  chartType: t,
  statistics: e,
  defaultTitle: a,
  countSeries: s,
  xAxis: i,
  notAvailable: n,
  timeAggregationTypes: o
};
export {
  t as chartType,
  s as countSeries,
  c as default,
  a as defaultTitle,
  n as notAvailable,
  e as statistics,
  o as timeAggregationTypes,
  i as xAxis
};
