// @arcgis/charts-components/dist/chunks/messages.lv.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const a = "Ielādējot diagrammu, radās kļūda.", t = "Žurnālfaila transformāciju nevar piemērot negatīvām vai nulles vērtībām.", e = "Kvadrātsaknes transformāciju nevar lietot negatīvām vērtībām.", s = "Ielādējot slāni, radās kļūda. Vietrādis (URL) = ${ url }. Portāla vienības ID = ${ portalItemId }.", i = "Konfigurācijā definētie lauki [${ fieldNames }] slānī nav atrasti.", r = "${ dataPath } jābūt unikālam. Sērijai ar nosaukumu ${ seriesName } ir ID (${ seriesID }), ko jau izmanto cita sērija.", n = "${ dataPath } nevar veikt apkopošanu laukā, kurā nav skaitlisku vērtību.", o = "${ dataPath } trūkst parametra ar nosaukumu ${ missingProperty }.", m = "${ dataPath } nedrīkst būt īsāks par ${ limit } rakstzīmēm.", d = "${ dataPath } nedrīkst būt mazāk par ${ limit } vienībām.", l = "${ dataPath } nedrīkst būt vairāk par ${ limit } vienībām.", u = "${ dataPath } jābūt vismaz vienai rakstzīmei, kas nav atstarpe.", k = "${ dataPath } nedrīkst būt ${ additionalProperty }.", c = "${ dataPath } jābūt vienādam ar kādu no šīm atļautajām vērtībām: ${ allowedValues }.", v = "${ dataPath } jāatbilst kādai no šādām shēmām: ${ schemaOptions }.", p = "Izkliedes diagrammas ar proporcionāliem simboliem netiek atbalstītas. Ticis piemērots noklusējuma simbola izmērs.", b = "Neizdevās nolasīt ievades datus.", $ = "QueryBins darbība šajā pakalpojumā netiek atbalstīta.", j = "Indeksa ${ seriesIndex } prognozētais sērijas veids ir “${ expectedType }”, bet tā vietā tika saņemts “${ receivedType }”", g = "vai", P = {
  defaultError: a,
  negativeValueInDataForLogTransformation: t,
  negativeValueInDataForSqrtTransformation: e,
  layerLoadFailure: s,
  layerFieldsNotFound: i,
  duplicateSeriesID: r,
  nonNumericAggregation: n,
  requiredProperty: o,
  minLength: m,
  minItems: d,
  maxItems: l,
  whiteSpacePattern: u,
  additionalProperty: k,
  enumValues: c,
  anyOfValues: v,
  bubbleChartValidateMsg: p,
  queryError: b,
  queryBinsNotSupported: $,
  invalidSeriesType: j,
  or: g
};
export {
  k as additionalProperty,
  v as anyOfValues,
  p as bubbleChartValidateMsg,
  P as default,
  a as defaultError,
  r as duplicateSeriesID,
  c as enumValues,
  j as invalidSeriesType,
  i as layerFieldsNotFound,
  s as layerLoadFailure,
  l as maxItems,
  d as minItems,
  m as minLength,
  t as negativeValueInDataForLogTransformation,
  e as negativeValueInDataForSqrtTransformation,
  n as nonNumericAggregation,
  g as or,
  $ as queryBinsNotSupported,
  b as queryError,
  o as requiredProperty,
  u as whiteSpacePattern
};
