// @arcgis/charts-components/dist/chunks/messages.et2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const a = { chart: "Diagramm", barChart: "Lintdiagramm", lineChart: "Joondiagramm", pieChart: "Sektordiagramm", histogram: "Histogramm", scatterplot: "Punktdiagramm", boxPlot: "Karpdiagramm" }, e = { mean: "Keskmine", median: "Keskmine", sum: "Summa", discretePercentile: "Diskreetne protsentiil", minimum: "Miinimum", maximum: "Maksimum", variance: "Dispersioon", count: "Kogus", aggregation: "${ statistics } /${ fieldName }", noAggregation: "Agregeerimist pole" }, t = "Diagramm", i = "Sarjade arv", s = "X-telg", m = "Pole kättesaadav", n = { start: "Haagi esimesse andmepunkti", end: "Haagi viimasesse andmepunkti" }, r = {
  chartType: a,
  statistics: e,
  defaultTitle: t,
  countSeries: i,
  xAxis: s,
  notAvailable: m,
  timeAggregationTypes: n
};
export {
  a as chartType,
  i as countSeries,
  r as default,
  t as defaultTitle,
  m as notAvailable,
  e as statistics,
  n as timeAggregationTypes,
  s as xAxis
};
