// @arcgis/charts-components/dist/chunks/messages.sl.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "Med nalaganjem grafikona je prišlo do napake.", a = "Pretvorbe dnevnika ni mogoče uporabiti za negativne ali ničelne vrednosti.", t = "Pretvorbe kvadratnega korena ni mogoče uporabiti za negativne vrednosti.", i = "Med nalaganjem sloja je prišlo do napake. URL = ${ url }. ID elementa portala = ${ portalItemId }.", n = "Polja [${ fieldNames }] opredeljena v konfiguraciji niso bila najdena v sloju.", o = "${ dataPath } mora biti edinstveno. Serija z imenom ${ seriesName } ima ID (${ seriesID }), ki ga že uporablja druga serija.", r = "${ dataPath } ne more izvajati združevanja brez števila na neštevilskem polju.", s = "${ dataPath } manjka lastnost imenovana ${ missingProperty }.", l = "${ dataPath } ne sme biti krajše od ${ limit } znakov.", d = "${ dataPath } ne sme imeti manj kot ${ limit } elementov.", m = "${ dataPath } ne sme imeti več kot ${ limit } elementov.", p = "${ dataPath } mora vsebovati vsaj en znak, ki ni presledek.", v = "${ dataPath } ne sme imeti ${ additionalProperty }.", c = "${ dataPath } mora biti enak eni od teh dovoljenih vrednosti: ${ allowedValues }.", j = "${ dataPath } se mora ujemati s shemo ene od teh: ${ schemaOptions }.", u = "Raztreseni grafikoni s proporcionalnimi simboli niso podprti. Uporabljena je bila privzeta velikost simbolov.", $ = "Branje vhodnih podatkov ni uspelo.", k = "Operacija queryBins ni podprta v tej storitvi.", g = "Pričakovani tip niza z indeksom ${ seriesIndex } je '${ expectedType }', ampak '${ receivedType }' je bil prejet namesto", h = "ali", b = {
  defaultError: e,
  negativeValueInDataForLogTransformation: a,
  negativeValueInDataForSqrtTransformation: t,
  layerLoadFailure: i,
  layerFieldsNotFound: n,
  duplicateSeriesID: o,
  nonNumericAggregation: r,
  requiredProperty: s,
  minLength: l,
  minItems: d,
  maxItems: m,
  whiteSpacePattern: p,
  additionalProperty: v,
  enumValues: c,
  anyOfValues: j,
  bubbleChartValidateMsg: u,
  queryError: $,
  queryBinsNotSupported: k,
  invalidSeriesType: g,
  or: h
};
export {
  v as additionalProperty,
  j as anyOfValues,
  u as bubbleChartValidateMsg,
  b as default,
  e as defaultError,
  o as duplicateSeriesID,
  c as enumValues,
  g as invalidSeriesType,
  n as layerFieldsNotFound,
  i as layerLoadFailure,
  m as maxItems,
  d as minItems,
  l as minLength,
  a as negativeValueInDataForLogTransformation,
  t as negativeValueInDataForSqrtTransformation,
  r as nonNumericAggregation,
  h as or,
  k as queryBinsNotSupported,
  $ as queryError,
  s as requiredProperty,
  p as whiteSpacePattern
};
