// @arcgis/charts-components/dist/chunks/messages.ro.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const e = "A survenit o eroare în timpul încărcării diagramei.", a = "Nu se poate aplica o transformare de jurnal în valori negative sau zero.", t = "Nu se poate aplica o transformare a rădăcinii pătrate în valori negative.", i = "A survenit o eroare în timpul încărcării stratului tematic. URL = ${ url }. ID-ul elementului de portal = ${ portalItemId }.", r = "Câmpurile (${ fieldNames }) definite în configurație nu au fost găsite în stratul tematic.", n = "${ dataPath } trebuie să fie unic. Seria denumită ${ seriesName } are un ID (${ seriesID }) utilizat deja de către o altă serie.", s = "${ dataPath } nu poate efectua agregarea non-numărare pe un câmp non-numeric.", o = "${ dataPath } are lipsă o proprietate denumită ${ missingProperty }.", u = "${ dataPath } nu trebuie să fie mai scurt de ${ limit } caractere.", c = "${ dataPath } nu trebuie să aibă mai puțin de ${ limit } articole.", l = "${ dataPath } nu trebuie să aibă mai mult de ${ limit } articole.", d = "${ dataPath } trebuie să aibă cel puțin un caracter non-spațiu gol.", p = "${ dataPath } nu trebuie să aibă ${ additionalProperty }.", m = "${ dataPath } trebuie să fie egal cu una dintre aceste valori permise: ${ allowedValues }.", $ = "${ dataPath } trebuie să corespundă schemei uneia dintre acestea: ${ schemaOptions }.", g = "Proprietățile diagramei de dispersie cu simbolurile proporționale nu sunt acceptate. A fost aplicată dimensiunea implicită a simbolului.", b = "A eșuat citirea datelor de intrare.", f = "Operația queryBins nu este acceptată în acest serviciu.", P = "Tipul așteptat al serie la indicele ${ seriesIndex } este „${ expectedType }”, dar a fost primit „${ receivedType }”", h = "sau", y = {
  defaultError: e,
  negativeValueInDataForLogTransformation: a,
  negativeValueInDataForSqrtTransformation: t,
  layerLoadFailure: i,
  layerFieldsNotFound: r,
  duplicateSeriesID: n,
  nonNumericAggregation: s,
  requiredProperty: o,
  minLength: u,
  minItems: c,
  maxItems: l,
  whiteSpacePattern: d,
  additionalProperty: p,
  enumValues: m,
  anyOfValues: $,
  bubbleChartValidateMsg: g,
  queryError: b,
  queryBinsNotSupported: f,
  invalidSeriesType: P,
  or: h
};
export {
  p as additionalProperty,
  $ as anyOfValues,
  g as bubbleChartValidateMsg,
  y as default,
  e as defaultError,
  n as duplicateSeriesID,
  m as enumValues,
  P as invalidSeriesType,
  r as layerFieldsNotFound,
  i as layerLoadFailure,
  l as maxItems,
  c as minItems,
  u as minLength,
  a as negativeValueInDataForLogTransformation,
  t as negativeValueInDataForSqrtTransformation,
  s as nonNumericAggregation,
  h as or,
  f as queryBinsNotSupported,
  b as queryError,
  o as requiredProperty,
  d as whiteSpacePattern
};
