// @arcgis/charts-components/dist/chunks/messages.vi.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const t = "Đã xảy ra lỗi khi tải biểu đồ.", n = "Không thể áp dụng phép biến đổi log lên giá trị âm hoặc bằng không.", h = "Không thể áp dụng phép biến đổi căn bậc hai lên giá trị âm.", a = "Đã xảy ra lỗi khi tải lớp. URL = ${ url }. ID mục cổng thông tin = ${ portalItemId }.", i = "Không tìm thấy trong lớp này các trường [${ fieldNames }] được xác định trong cấu hình.", c = "${ dataPath } phải là duy nhất. Chuỗi có tên ${ seriesName } có một ID (${ seriesID }) đã được chuỗi khác sử dụng.", e = "${ dataPath } không thể thực hiện tổng hợp không đếm trên trường không phải là số.", o = "${ dataPath } đang thiếu một tên phù hợp ${ missingProperty }.", r = "${ dataPath } không được ngắn hơn ${ limit } ký tự.", g = "${ dataPath } không được chứa dưới ${ limit } mục.", s = "${ dataPath } không được chứa nhiều hơn ${ limit } mục.", l = "${ dataPath } phải chứa ít nhất một ký tự không phải khoảng trắng.", d = "${ dataPath } không được chứa ${ additionalProperty }.", p = "${ dataPath } phải bằng một trong những giá trị được phép này: ${ allowedValues }.", m = "${ dataPath } phải khớp với lược đồ của một trong những thứ sau: ${ schemaOptions }.", u = "Biểu đồ phân tán có ký hiệu tỷ lệ không được hỗ trợ. Đã áp dụng kích cỡ ký hiệu theo mặc định.", $ = "Không thể đọc dữ liệu đầu vào.", k = "Thao tác queryBins không được hỗ trợ trên dịch vụ này.", y = "Loại chuỗi mong đợi tại chỉ mục ${ seriesIndex } là '${ expectedType }' nhưng thay vào đó lại nhận được '${ receivedType }'", P = "hoặc", I = {
  defaultError: t,
  negativeValueInDataForLogTransformation: n,
  negativeValueInDataForSqrtTransformation: h,
  layerLoadFailure: a,
  layerFieldsNotFound: i,
  duplicateSeriesID: c,
  nonNumericAggregation: e,
  requiredProperty: o,
  minLength: r,
  minItems: g,
  maxItems: s,
  whiteSpacePattern: l,
  additionalProperty: d,
  enumValues: p,
  anyOfValues: m,
  bubbleChartValidateMsg: u,
  queryError: $,
  queryBinsNotSupported: k,
  invalidSeriesType: y,
  or: P
};
export {
  d as additionalProperty,
  m as anyOfValues,
  u as bubbleChartValidateMsg,
  I as default,
  t as defaultError,
  c as duplicateSeriesID,
  p as enumValues,
  y as invalidSeriesType,
  i as layerFieldsNotFound,
  a as layerLoadFailure,
  s as maxItems,
  g as minItems,
  r as minLength,
  n as negativeValueInDataForLogTransformation,
  h as negativeValueInDataForSqrtTransformation,
  e as nonNumericAggregation,
  P as or,
  k as queryBinsNotSupported,
  $ as queryError,
  o as requiredProperty,
  l as whiteSpacePattern
};
