// @arcgis/charts-components/dist/chunks/messages.hu2.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const a = { chart: "Diagram", barChart: "Oszlop-/sávdiagram", lineChart: "Vonaldiagram", pieChart: "Kördiagram", histogram: "Hisztogram", scatterplot: "Szórásdiagram", boxPlot: "Boxdiagram" }, t = { mean: "Átlag", median: "Medián", sum: "Összeg", discretePercentile: "Diszkrét percentilis", minimum: "Minimum", maximum: "Maximum", variance: "Variancia", count: "Darabszám", aggregation: "${ statistics }/${ fieldName }", noAggregation: "Nincs összevonás" }, s = "Diagram", i = "Számsorozat", e = "X tengely", o = "Nem érhető el", r = { start: "Csatolás az első adatponthoz", end: "Csatolás az utolsó adatponthoz" }, n = {
  chartType: a,
  statistics: t,
  defaultTitle: s,
  countSeries: i,
  xAxis: e,
  notAvailable: o,
  timeAggregationTypes: r
};
export {
  a as chartType,
  i as countSeries,
  n as default,
  s as defaultTitle,
  o as notAvailable,
  t as statistics,
  r as timeAggregationTypes,
  e as xAxis
};
