// @arcgis/charts-components-react/dist/components.js
/// <reference types="@arcgis/charts-components" preserve="true" />
import React from "react";
import { createComponent } from "@lit/react";
import { makeReactWrapperFactory, getReactWrapperOptions } from "@arcgis/lumina";
const createWrapper = /*@__PURE__*/ makeReactWrapperFactory(React, createComponent);
export const ArcgisChart = /*@__PURE__*/ createWrapper(getReactWrapperOptions("arcgis-chart", {
    onArcgisAxesMinMaxChange: "arcgisAxesMinMaxChange",
    onArcgisBadDataWarningRaise: "arcgisBadDataWarningRaise",
    onArcgisConfigChange: "arcgisConfigChange",
    onArcgisDataFetchComplete: "arcgisDataFetchComplete",
    onArcgisDataProcessComplete: "arcgisDataProcessComplete",
    onArcgisDataProcessError: "arcgisDataProcessError",
    onArcgisLegendItemVisibilityChange: "arcgisLegendItemVisibilityChange",
    onArcgisNoRenderPropChange: "arcgisNoRenderPropChange",
    onArcgisRuntimeError: "arcgisRuntimeError",
    onArcgisSelectionComplete: "arcgisSelectionComplete",
    onArcgisSeriesColorChange: "arcgisSeriesColorChange",
    onArcgisSeriesOrder: "arcgisSeriesOrder",
    onArcgisUpdateComplete: "arcgisUpdateComplete",
}));
export const ArcgisChartsActionBar = /*@__PURE__*/ createWrapper(getReactWrapperOptions("arcgis-charts-action-bar", {
    onArcgisActionBarToggle: "arcgisActionBarToggle",
    onArcgisDefaultActionSelect: "arcgisDefaultActionSelect",
}));
