// @arcgis/charts-components/dist/components/arcgis-charts-config-invalid-input-message/customElement.js
import { c as s } from "../../chunks/runtime.js";
import { html as i } from "lit-html";
import { c as t } from "../../chunks/chart-ui-utils2.js";
import { LitElement as a, safeClassMap as e } from "@arcgis/lumina";
import { css as l } from "@lit/reactive-element/css-tag.js";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const n = {
  invalidMessageColor: "invalid-message-color"
}, o = l`:host{display:block}.invalid-message-color{color:red}`;
class r extends a {
  constructor() {
    super(...arguments), this.isInvalid = !1;
  }
  static {
    this.properties = { invalidInputMessage: 3, isInvalid: 5 };
  }
  static {
    this.styles = o;
  }
  render() {
    return this.isInvalid ? i`<calcite-input-message .status=${t.INVALID} icon=exclamationMarkTriangle .hidden=${!this.isInvalid} class=${e(n.invalidMessageColor)}>${this.invalidInputMessage}</calcite-input-message>` : null;
  }
}
s("arcgis-charts-config-invalid-input-message", r);
export {
  r as ArcgisChartsConfigInvalidInputMessage
};
