// @arcgis/map-components/dist/components/arcgis-bookmarks/customElement.js
import { c as i } from "../../chunks/runtime.js";
import { usePropertyChange as e } from "@arcgis/lumina/controllers";
import { reEmitEvent as t } from "@arcgis/lumina/controllers/accessor";
import s from "@arcgis/core/widgets/Bookmarks.js";
import { LitElement as o, createEvent as l, noShadowRoot as a } from "@arcgis/lumina";
import { m as h } from "../../chunks/useWidget.js";
import { css as r } from "@lit/reactive-element/css-tag.js";
import { c as d } from "../../chunks/createCloseEvent.js";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const n = r`@layer{arcgis-bookmarks{display:block}}`, m = h(s);
class g extends o {
  constructor() {
    super(...arguments), this.widget = m(this), this.viewModel = this.widget.viewModel, this.arcgisClose = d(this), this.autoDestroyDisabled = !1, this.bookmarks = this.widget.bookmarks, this.closed = !1, this.defaultCreateOptions = this.widget.defaultCreateOptions, this.defaultEditOptions = this.widget.defaultEditOptions, this.disabled = this.widget.disabled, this.dragEnabled = this.widget.dragEnabled, this.filterPlaceholder = this.widget.filterPlaceholder, this.filterText = this.widget.filterText, this.goToOverride = this.widget.goToOverride, this.hideFlow = this.widget.visibleElements.flow, this.hideThumbnail = this.widget.visibleElements.thumbnail, this.hideTime = this.widget.visibleElements.time, this.icon = this.widget.icon, this.label = this.widget.label, this.position = "bottom-left", this.showAddBookmarkButton = this.widget.visibleElements.addBookmarkButton, this.showCloseButton = this.widget.visibleElements.closeButton, this.showCollapseButton = this.widget.visibleElements.collapseButton, this.showEditBookmarkButton = this.widget.visibleElements.editBookmarkButton, this.showFilter = this.widget.visibleElements.filter, this.showHeading = this.widget.visibleElements.heading, this.state = this.viewModel.state, this.arcgisEdit = t(() => this.widget, "bookmark-edit"), this.arcgisPropertyChange = e()("state"), this.arcgisReady = l(), this.arcgisSelect = t(() => this.widget, "bookmark-select");
  }
  static {
    this.properties = { autoDestroyDisabled: 5, bookmarks: 0, closed: 5, defaultCreateOptions: 0, defaultEditOptions: 0, disabled: 7, dragEnabled: 5, filterPlaceholder: 1, filterText: 1, goToOverride: 0, hideFlow: 5, hideThumbnail: 5, hideTime: 5, icon: 1, label: 1, position: 1, referenceElement: 1, showAddBookmarkButton: 5, showCloseButton: 5, showCollapseButton: 5, showEditBookmarkButton: 5, showFilter: 5, showHeading: 5, state: 3 };
  }
  static {
    this.shadowRootOptions = a;
  }
  static {
    this.styles = n;
  }
  async destroy() {
    await this.manager.destroy();
  }
}
i("arcgis-bookmarks", g);
export {
  g as ArcgisBookmarks
};
