// @arcgis/map-components/dist/components/arcgis-print-map-only-advanced-options/customElement.js
import { c as i } from "../../chunks/runtime.js";
import { html as e } from "lit-html";
import { LitElement as s, safeClassMap as n } from "@arcgis/lumina";
import { css as a } from "@lit/reactive-element/css-tag.js";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const r = a`.arcgis-print__advanced-options-container{padding:var(--calcite-spacing-xxs);display:flex;flex-direction:column;gap:var(--calcite-spacing-xs)}`, c = "arcgis-print", o = {
  advancedOptionsContainer: `${c}__advanced-options-container`
};
class l extends s {
  static {
    this.properties = { viewModel: 0, messages: 0 };
  }
  static {
    this.styles = r;
  }
  loaded() {
    this.manager.onLifecycle(() => []);
  }
  render() {
    return e`<div aria-labelledby=${`${this.el.id}__advancedOptionsForLayout`} class=${n(o.advancedOptionsContainer)}>${this._renderScaleSection()}${this._renderDpiSection()}${this._renderOutSpatialReferenceSection()}${this._renderAttributionSection()}${this._renderCustomTextElementSection()}</div>`;
  }
  _renderScaleSection() {
    return e`<arcgis-print-scale-section .messages=${this.messages} .viewModel=${this.viewModel}></arcgis-print-scale-section>`;
  }
  _renderDpiSection() {
    return e`<arcgis-print-dpi-input .messages=${this.messages} .viewModel=${this.viewModel}></arcgis-print-dpi-input>`;
  }
  _renderOutSpatialReferenceSection() {
    return e`<arcgis-print-out-spatial-reference-input .messages=${this.messages} .viewModel=${this.viewModel}></arcgis-print-out-spatial-reference-input>`;
  }
  _renderAttributionSection() {
    return e`<div><calcite-label layout=inline><calcite-checkbox .checked=${!!this.viewModel.templateOptions.attributionEnabled} @calciteCheckboxChange=${(t) => this.viewModel.templateOptions.attributionEnabled = t.target.checked}></calcite-checkbox>${this.messages.attribution}</calcite-label></div>`;
  }
  _renderCustomTextElementSection() {
    return e`<arcgis-print-custom-text-elements .viewModel=${this.viewModel}></arcgis-print-custom-text-elements>`;
  }
}
i("arcgis-print-map-only-advanced-options", l);
export {
  l as PrintMapOnlyAdvancedOptions
};
