// @arcgis/charts-components/dist/components/arcgis-charts-config-bar-chart-popover/customElement.js
import { c as L } from "../../chunks/runtime.js";
import { repeat as b } from "lit-html/directives/repeat.js";
import { keyed as g } from "lit-html/directives/keyed.js";
import { WebChartStatisticType as $, RESTFieldType as m } from "@arcgis/charts-spec";
import { LitElement as F, createEvent as C, safeClassMap as c } from "@arcgis/lumina";
import { html as r } from "lit";
import { createRef as y, ref as p } from "lit-html/directives/ref.js";
import { isEqual as k } from "lodash-es";
import { u as I } from "../../chunks/useT9n.js";
import { B as o } from "../../chunks/interfaces2.js";
import { S as d } from "../../chunks/common.js";
import "d3-array";
import { aL as P, bS as x, bR as D, c9 as T, b3 as v, b1 as w } from "../../chunks/interfaces.js";
import "@arcgis/core/geometry/support/jsonUtils.js";
import "@arcgis/core/rest/support/AttributeBinsQuery.js";
import "@arcgis/core/rest/support/Query.js";
import "@arcgis/core/rest/support/StatisticDefinition.js";
import "@arcgis/core/time/TimeExtent.js";
import "@arcgis/core/core/promiseUtils.js";
import "@arcgis/core/request.js";
import { e as S } from "../../chunks/popover-ui-utils.js";
import { j as f, U as u, i as h, d as A, e as E, h as N } from "../../chunks/chart-ui-utils2.js";
import { css as B } from "@lit/reactive-element/css-tag.js";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.33/esri/copyright.txt for details.
v4.33.19 */
const K = B`.hide{display:none}.am5-modal{width:100%;height:100%;position:absolute;z-index:100000;top:0;left:0}.am5-modal-curtain{top:0;left:0;width:100%;height:100%;position:absolute;background:#fff!important;z-index:100}.am5-modal-wrapper{top:0;left:0;width:100%;height:100%;position:absolute;display:flex;align-items:center;justify-content:center;white-space:nowrap;background:#ffffff80;z-index:101}.am5-modal-content{display:inline-block;padding:1.2em;vertical-align:middle;text-align:start;white-space:normal;background:#fff;border-radius:4px;box-shadow:#00000073 0 0 36px;color:#000}.am5-layer-1000{z-index:1000!important}.arcgis-charts-modal{box-shadow:none!important}.arcgis-charts-modal-header{background-color:#0000000d;font-weight:700;padding:4px;align-content:center}.show{display:block}.notifyPanel{flex:0 1 auto}.disable-interactions{pointer-events:none}.dim-text{color:var(--arcgis-charts-dim-text)}.name{display:flex;flex-direction:column;margin:.5rem;padding:.5rem;background:#fff;min-width:250px}.header{margin:0;font-weight:bolder}.fab{display:flex;justify-content:stretch}.pick-list-item-label{font-size:small}.sort-icon-color{color:#000}.sort-y-axis{transform:rotate(90deg)}.sort-order-pick-list{overflow-y:hidden}.panel{max-height:60vh}`, n = {
  jsAppFlyout: "js-app-flyout",
  name: "bar-chart-data-popover",
  numericFieldPickList: "bar-chart-data-numeric-fields-pick-list",
  header: "header",
  fab: "fab",
  pickListItemLabel: "pick-list-item-label",
  rotateClass: "sort-y-axis",
  sortOrderPickList: "sort-order-pick-list",
  panel: "panel"
};
class j extends F {
  constructor() {
    super(...arguments), this._messages = I(), this.picklist = y(), this.popoverElement = y(), this.placement = "leading", this.offsetDistance = -200, this.pickListHasChanged = !1, this.aggregationLabels = [], this.layerFieldsInfo = [], this.isCustomSortDisabled = !1, this.isNoAggregationDisabled = !1, this.arcgisChartsConfigBarChartDataPopoverChange = C(), this.arcgisChartsConfigPopoverClose = C();
  }
  static {
    this.properties = { aggregationLabels: 16, messageOverrides: 0, headingTitle: 3, referenceElement: 2, open: 7, contentKind: 1, layerFieldsInfo: 0, selectedContent: 1, isCustomSortDisabled: 5, isNoAggregationDisabled: 5 };
  }
  static {
    this.styles = K;
  }
  async reposition() {
    await this.popoverElement.value?.reposition();
  }
  async willUpdate(e) {
    e.has("selectedContent") && this.selectedContentChange(this.selectedContent, e.get("selectedContent"));
    const t = await Promise.all(f.map(async (i) => await P(i)));
    k(this.aggregationLabels, t) || (this.aggregationLabels = t);
  }
  async updated() {
    setTimeout(async () => {
      await this.picklist.value?.setFocus();
    }, u.PopoverTimer), S(this.popoverElement.value, this.open), this.hasUpdated || setTimeout(() => {
      this.popoverElement.value?.reposition();
    }, u.PopoverTimer);
  }
  selectedContentChange(e, t) {
    this.pickListHasChanged = !k(e, t);
  }
  buildAggregationListItems() {
    const e = this.isNoAggregationDisabled ? f.filter((t) => t !== $.NoAggregation) : f;
    return b(e, (t) => t, (t, i) => r`<calcite-list-item .label=${this.aggregationLabels[i]} .value=${t} .selected=${h(t, this.selectedContent)}></calcite-list-item>`);
  }
  createTooltip(e) {
    if (e.target?.label === this._messages.customSort && this.isCustomSortDisabled) {
      const t = e.target, { formatLocale: i } = x(), a = new Intl.NumberFormat(i).format(u.customSortLimit), s = D(this._messages.customSortLimit ?? "", {
        totalLimit: a
      });
      this.tooltip = A(t), this.tooltip.innerHTML = s, document.body.appendChild(this.tooltip);
    }
  }
  destroyTooltip() {
    E(this.tooltip);
  }
  buildSortOrder() {
    const e = Object.keys(d);
    let t;
    return b(e, (i) => i, (i) => {
      let l = "";
      switch (i) {
        case d.xAxisAsc: {
          t = "a-z-down";
          break;
        }
        case d.xAxisDesc: {
          t = "a-z-up";
          break;
        }
        case d.yAxisAsc: {
          t = "sort-descending", l = n.rotateClass;
          break;
        }
        case d.yAxisDesc: {
          t = "sort-ascending", l = n.rotateClass;
          break;
        }
        case d.customSort: {
          t = "arrow-up-down";
          break;
        }
      }
      return r`<calcite-list-item .disabled=${this.isCustomSortDisabled && i === d.customSort} .label=${this._messages[i]} .selected=${this.selectedContent === i} .value=${d[i]} @mouseover=${this.createTooltip} @mouseout=${this.destroyTooltip}><calcite-icon class=${`sort-icon-color ${l}`} slot=content-start .icon=${t}></calcite-icon></calcite-list-item>`;
    });
  }
  buildPickListItem(e) {
    return g(e.name, r`<calcite-list-item .label=${T(this.layerFieldsInfo, e.name)} .value=${e.name} .selected=${h(e.name, this.selectedContent)}></calcite-list-item>`);
  }
  buildPickListItems(e) {
    const t = this.layerFieldsInfo?.length ?? 0, i = [];
    switch (e) {
      case o.numericFields: {
        for (let l = 0; l < t; l += 1) {
          const a = this.layerFieldsInfo?.[l];
          a && w(a) && i.push(this.buildPickListItem(a));
        }
        break;
      }
      case o.category: {
        const l = g(" ", r`<calcite-list-item label=" " value=" " .selected=${h("", this.selectedContent)}></calcite-list-item>`);
        i.push(l);
        for (let a = 0; a < t; a += 1) {
          const s = this.layerFieldsInfo?.[a];
          s && (v(s) || s.type === m.String || s.type === m.Date || s.type === m.DateOnly || s.type === m.TimestampOffset) && i.push(this.buildPickListItem(s));
        }
        break;
      }
      case o.splitByField: {
        const l = g(" ", r`<calcite-list-item label=" " value=" " .selected=${h("", this.selectedContent)}></calcite-list-item>`);
        i.push(l);
        for (let a = 0; a < t; a += 1) {
          const s = this.layerFieldsInfo?.[a];
          s && (v(s) || s.type === m.String) && i.push(this.buildPickListItem(s));
        }
        break;
      }
    }
    return i;
  }
  closePopover() {
    if (this.contentKind === o.numericFields && this.pickListHasChanged) {
      const e = {
        pickListElement: this.picklist.value,
        eventEmitter: this.arcgisChartsConfigBarChartDataPopoverChange,
        contentKind: this.contentKind
      };
      N(e);
    }
    this.open = !1, this.arcgisChartsConfigPopoverClose.emit();
  }
  onDataContentTypeChange(e) {
    this.pickListHasChanged = !0;
    const t = e.target.selectedItems;
    this.arcgisChartsConfigBarChartDataPopoverChange.emit({
      contentKind: this.contentKind,
      value: t.map((i) => i.value)
    }), this.open = !1;
  }
  onNumericFieldsChange(e) {
    e.target !== null && (this.selectedContent = e.target.selectedItems.map((t) => t.value));
  }
  onNumericFieldSelectionDone() {
    this.pickListHasChanged && this.closePopover();
  }
  renderAggregationType() {
    const e = this.buildAggregationListItems();
    return r`<calcite-list label class=${c(n.pickListItemLabel)} selection-mode=single selection-appearance=border @calciteListChange=${this.onDataContentTypeChange}>${e}</calcite-list>`;
  }
  renderCategory() {
    const e = this.buildPickListItems(o.category);
    return r`<calcite-list label class=${c(n.pickListItemLabel)} selection-mode=single selection-appearance=border filter-enabled @calciteListChange=${this.onDataContentTypeChange} ${p(this.picklist)}>${e}</calcite-list>`;
  }
  renderNumericFields() {
    const e = this.buildPickListItems(o.numericFields);
    return r`<calcite-list label class=${c(n.numericFieldPickList)} selection-mode=multiple filter-enabled @calciteListChange=${this.onNumericFieldsChange} ${p(this.picklist)}>${e}</calcite-list>`;
  }
  renderSortOrder() {
    const e = this.buildSortOrder();
    return r`<calcite-list label class=${c(n.sortOrderPickList)} selection-mode=single selection-appearance=border @calciteListChange=${this.onDataContentTypeChange}>${e}</calcite-list>`;
  }
  renderSplitByField() {
    const e = this.buildPickListItems(o.splitByField);
    return r`<calcite-list label class=${c(n.pickListItemLabel)} selection-mode=single selection-appearance=border filter-enabled @calciteListChange=${this.onDataContentTypeChange} ${p(this.picklist)}>${e}</calcite-list>`;
  }
  renderPopoverInfo() {
    let e;
    switch (this.contentKind) {
      case o.aggregation:
        e = this.renderAggregationType();
        break;
      case o.category:
        e = this.renderCategory();
        break;
      case o.numericFields:
        e = this.renderNumericFields();
        break;
      case o.splitByField:
        e = this.renderSplitByField();
        break;
      case o.sortOrder:
        e = this.renderSortOrder();
        break;
    }
    return e;
  }
  render() {
    let e;
    return this.contentKind === o.numericFields && (e = r`<div slot=footer><calcite-fab class=${c(n.fab)} appearance=outline-fill kind=neutral icon .label=${this._messages.selectionDone} scale=s text-enabled .text=${this._messages.selectionDone} @click=${this.onNumericFieldSelectionDone}></calcite-fab></div>`), r`<div class=${c(n.jsAppFlyout)}><calcite-popover class=${c(n.name)} .referenceElement=${this.referenceElement ?? document.body} .placement=${this.placement} .offsetDistance=${this.offsetDistance} .open=${this.open} @calcitePopoverClose=${this.closePopover} label ${p(this.popoverElement)}><calcite-panel class=${c(n.panel)} closable .closed=${!this.open} @calcitePanelClose=${this.closePopover}><div slot=header-content class=${c(n.header)}>${this.headingTitle ?? ""}</div>${this.renderPopoverInfo()}${e}</calcite-panel></calcite-popover></div>`;
  }
}
L("arcgis-charts-config-bar-chart-popover", j);
export {
  j as ArcgisChartsConfigBarChartPopover
};
